

<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>BCR clustering and visualization &mdash; dandelion 0.0.26 documentation</title>
  

  
  <link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />

  
  
    <link rel="shortcut icon" href="../_static/dandelion_logo.png"/>
  
  
  

  
  <!--[if lt IE 9]>
    <script src="../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="../" src="../_static/documentation_options.js"></script>
        <script src="../_static/jquery.js"></script>
        <script src="../_static/underscore.js"></script>
        <script src="../_static/doctools.js"></script>
        <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
        <script async="async" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.7/latest.js?config=TeX-AMS-MML_HTMLorMML"></script>
        <script type="text/x-mathjax-config">MathJax.Hub.Config({"tex2jax": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true, "ignoreClass": "document", "processClass": "math|output_area"}})</script>
    
    <script type="text/javascript" src="../_static/js/theme.js"></script>

    
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Running from R" href="5_dandelion_running_from_R.html" />
    <link rel="prev" title="Filtering" href="2_dandelion_filtering-10x%20data.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="../index.html">
          

          
            
            <img src="../_static/dandelion_logo.png" class="logo" alt="Logo"/>
          
          </a>

          
            
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        
        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <p class="caption"><span class="caption-text">dandelion:</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../README.html">Overview</a></li>
<li class="toctree-l1"><a class="reference internal" href="../README.html#installation">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../README.html#required-database">Required database</a></li>
<li class="toctree-l1"><a class="reference internal" href="../README.html#external-softwares">External softwares</a></li>
<li class="toctree-l1"><a class="reference internal" href="../README.html#basic-requirements">Basic Requirements</a></li>
<li class="toctree-l1"><a class="reference internal" href="../README.html#acknowledgements">Acknowledgements</a></li>
</ul>
<p class="caption"><span class="caption-text">Tutorial:</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="1_dandelion_preprocessing-10x%20data.html">Pre-processing</a></li>
<li class="toctree-l1"><a class="reference internal" href="2_dandelion_filtering-10x%20data.html">Filtering</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">BCR clustering and visualization</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#Quick-primer-to-the-Dandelion-class">Quick primer to the <em>Dandelion</em> class</a></li>
<li class="toctree-l2"><a class="reference internal" href="#Finding-clones">Finding clones</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#Running-tl.find_clones">Running <code class="docutils literal notranslate"><span class="pre">tl.find_clones</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#Alternative-:-Running-tl.define_clones">Alternative : Running <code class="docutils literal notranslate"><span class="pre">tl.define_clones</span></code></a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#Visualization-of-BCR-network">Visualization of BCR network</a></li>
<li class="toctree-l2"><a class="reference internal" href="#Integration-with-scanpy">Integration with scanpy</a></li>
<li class="toctree-l2"><a class="reference internal" href="#Calculating-mutational-load">Calculating mutational load</a></li>
<li class="toctree-l2"><a class="reference internal" href="#Calculating-size-of-clones">Calculating size of clones</a></li>
<li class="toctree-l2"><a class="reference internal" href="#Calculating-diversity">Calculating diversity</a></li>
<li class="toctree-l2"><a class="reference internal" href="#Additional-plotting-functions">Additional plotting functions</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#barplot"><em>barplot</em></a></li>
<li class="toctree-l3"><a class="reference internal" href="#stackedbarplot"><em>stackedbarplot</em></a></li>
<li class="toctree-l3"><a class="reference internal" href="#spectratype"><em>spectratype</em></a></li>
<li class="toctree-l3"><a class="reference internal" href="#clone_overlap"><em>clone_overlap</em></a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="5_dandelion_running_from_R.html">Running from R</a></li>
</ul>

            
          
        </div>
        
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">dandelion</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="../index.html" class="icon icon-home"></a> &raquo;</li>
        
      <li>BCR clustering and visualization</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            <a href="../_sources/notebooks/3_dandelion_findingclones_and_analysis-10x data.ipynb.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  
<style>
/* CSS for nbsphinx extension */

/* remove conflicting styling from Sphinx themes */
div.nbinput.container div.prompt *,
div.nboutput.container div.prompt *,
div.nbinput.container div.input_area pre,
div.nboutput.container div.output_area pre,
div.nbinput.container div.input_area .highlight,
div.nboutput.container div.output_area .highlight {
    border: none;
    padding: 0;
    margin: 0;
    box-shadow: none;
}

div.nbinput.container > div[class*=highlight],
div.nboutput.container > div[class*=highlight] {
    margin: 0;
}

div.nbinput.container div.prompt *,
div.nboutput.container div.prompt * {
    background: none;
}

div.nboutput.container div.output_area .highlight,
div.nboutput.container div.output_area pre {
    background: unset;
}

div.nboutput.container div.output_area div.highlight {
    color: unset;  /* override Pygments text color */
}

/* avoid gaps between output lines */
div.nboutput.container div[class*=highlight] pre {
    line-height: normal;
}

/* input/output containers */
div.nbinput.container,
div.nboutput.container {
    display: -webkit-flex;
    display: flex;
    align-items: flex-start;
    margin: 0;
    width: 100%;
}
@media (max-width: 540px) {
    div.nbinput.container,
    div.nboutput.container {
        flex-direction: column;
    }
}

/* input container */
div.nbinput.container {
    padding-top: 5px;
}

/* last container */
div.nblast.container {
    padding-bottom: 5px;
}

/* input prompt */
div.nbinput.container div.prompt pre {
    color: #307FC1;
}

/* output prompt */
div.nboutput.container div.prompt pre {
    color: #BF5B3D;
}

/* all prompts */
div.nbinput.container div.prompt,
div.nboutput.container div.prompt {
    width: 4.5ex;
    padding-top: 5px;
    position: relative;
    user-select: none;
}

div.nbinput.container div.prompt > div,
div.nboutput.container div.prompt > div {
    position: absolute;
    right: 0;
    margin-right: 0.3ex;
}

@media (max-width: 540px) {
    div.nbinput.container div.prompt,
    div.nboutput.container div.prompt {
        width: unset;
        text-align: left;
        padding: 0.4em;
    }
    div.nboutput.container div.prompt.empty {
        padding: 0;
    }

    div.nbinput.container div.prompt > div,
    div.nboutput.container div.prompt > div {
        position: unset;
    }
}

/* disable scrollbars on prompts */
div.nbinput.container div.prompt pre,
div.nboutput.container div.prompt pre {
    overflow: hidden;
}

/* input/output area */
div.nbinput.container div.input_area,
div.nboutput.container div.output_area {
    -webkit-flex: 1;
    flex: 1;
    overflow: auto;
}
@media (max-width: 540px) {
    div.nbinput.container div.input_area,
    div.nboutput.container div.output_area {
        width: 100%;
    }
}

/* input area */
div.nbinput.container div.input_area {
    border: 1px solid #e0e0e0;
    border-radius: 2px;
    /*background: #f5f5f5;*/
}

/* override MathJax center alignment in output cells */
div.nboutput.container div[class*=MathJax] {
    text-align: left !important;
}

/* override sphinx.ext.imgmath center alignment in output cells */
div.nboutput.container div.math p {
    text-align: left;
}

/* standard error */
div.nboutput.container div.output_area.stderr {
    background: #fdd;
}

/* ANSI colors */
.ansi-black-fg { color: #3E424D; }
.ansi-black-bg { background-color: #3E424D; }
.ansi-black-intense-fg { color: #282C36; }
.ansi-black-intense-bg { background-color: #282C36; }
.ansi-red-fg { color: #E75C58; }
.ansi-red-bg { background-color: #E75C58; }
.ansi-red-intense-fg { color: #B22B31; }
.ansi-red-intense-bg { background-color: #B22B31; }
.ansi-green-fg { color: #00A250; }
.ansi-green-bg { background-color: #00A250; }
.ansi-green-intense-fg { color: #007427; }
.ansi-green-intense-bg { background-color: #007427; }
.ansi-yellow-fg { color: #DDB62B; }
.ansi-yellow-bg { background-color: #DDB62B; }
.ansi-yellow-intense-fg { color: #B27D12; }
.ansi-yellow-intense-bg { background-color: #B27D12; }
.ansi-blue-fg { color: #208FFB; }
.ansi-blue-bg { background-color: #208FFB; }
.ansi-blue-intense-fg { color: #0065CA; }
.ansi-blue-intense-bg { background-color: #0065CA; }
.ansi-magenta-fg { color: #D160C4; }
.ansi-magenta-bg { background-color: #D160C4; }
.ansi-magenta-intense-fg { color: #A03196; }
.ansi-magenta-intense-bg { background-color: #A03196; }
.ansi-cyan-fg { color: #60C6C8; }
.ansi-cyan-bg { background-color: #60C6C8; }
.ansi-cyan-intense-fg { color: #258F8F; }
.ansi-cyan-intense-bg { background-color: #258F8F; }
.ansi-white-fg { color: #C5C1B4; }
.ansi-white-bg { background-color: #C5C1B4; }
.ansi-white-intense-fg { color: #A1A6B2; }
.ansi-white-intense-bg { background-color: #A1A6B2; }

.ansi-default-inverse-fg { color: #FFFFFF; }
.ansi-default-inverse-bg { background-color: #000000; }

.ansi-bold { font-weight: bold; }
.ansi-underline { text-decoration: underline; }


div.nbinput.container div.input_area div[class*=highlight] > pre,
div.nboutput.container div.output_area div[class*=highlight] > pre,
div.nboutput.container div.output_area div[class*=highlight].math,
div.nboutput.container div.output_area.rendered_html,
div.nboutput.container div.output_area > div.output_javascript,
div.nboutput.container div.output_area:not(.rendered_html) > img{
    padding: 5px;
    margin: 0;
}

/* fix copybtn overflow problem in chromium (needed for 'sphinx_copybutton') */
div.nbinput.container div.input_area > div[class^='highlight'],
div.nboutput.container div.output_area > div[class^='highlight']{
    overflow-y: hidden;
}

/* hide copybtn icon on prompts (needed for 'sphinx_copybutton') */
.prompt a.copybtn {
    display: none;
}

/* Some additional styling taken form the Jupyter notebook CSS */
div.rendered_html table {
  border: none;
  border-collapse: collapse;
  border-spacing: 0;
  color: black;
  font-size: 12px;
  table-layout: fixed;
}
div.rendered_html thead {
  border-bottom: 1px solid black;
  vertical-align: bottom;
}
div.rendered_html tr,
div.rendered_html th,
div.rendered_html td {
  text-align: right;
  vertical-align: middle;
  padding: 0.5em 0.5em;
  line-height: normal;
  white-space: normal;
  max-width: none;
  border: none;
}
div.rendered_html th {
  font-weight: bold;
}
div.rendered_html tbody tr:nth-child(odd) {
  background: #f5f5f5;
}
div.rendered_html tbody tr:hover {
  background: rgba(66, 165, 245, 0.2);
}

/* CSS overrides for sphinx_rtd_theme */

/* 24px margin */
.nbinput.nblast.container,
.nboutput.nblast.container {
    margin-bottom: 19px;  /* padding has already 5px */
}

/* ... except between code cells! */
.nblast.container + .nbinput.container {
    margin-top: -19px;
}

.admonition > p:before {
    margin-right: 4px;  /* make room for the exclamation icon */
}

/* Fix math alignment, see https://github.com/rtfd/sphinx_rtd_theme/pull/686 */
.math {
    text-align: unset;
}
</style>
<div class="section" id="BCR-clustering-and-visualization">
<h1>BCR clustering and visualization<a class="headerlink" href="#BCR-clustering-and-visualization" title="Permalink to this headline">¶</a></h1>
<p><img alt="dandelion_logo" src="../_images/dandelion_logo_illustration.png" /></p>
<p>Now that we have both 1) a pre-processed BCR data and 2) matching <code class="docutils literal notranslate"><span class="pre">AnnData</span></code> object, we can start finding clones and <em>‘integrate’</em> the results. All the BCR analyses files can be saved as <em>.tsv</em> format so that it can be used in other tools like <em>immcantation</em>, <em>immunoarch</em>, <em>vdjtools</em>, etc.</p>
<p>On the topic of finding clones, there are many ways used for identifying BCR clones, almost all involving some measure based on sequence similarity. There are also a lot of very well established guidelines and criterias maintained by the BCR community. For example, <em>immcantation</em> uses a number of model-based <a class="reference external" href="https://changeo.readthedocs.io/en/stable/methods/clustering.html">methods</a> to group clones based on the distribution of length-normalised junctional hamming distance while
<a class="reference external" href="https://www.well.ox.ac.uk/research/research-groups/bashford-rogers">rbr1</a> developed a method to use the whole BCR VDJ sequence to define clones as shown in this recent <a class="reference external" href="https://www.nature.com/articles/s41586-019-1595-3.pdf">paper</a>. While these methods have mainly been applied to bulk BCR-seq protocols, they are biological grounded and should be applicable to single cells.</p>
<p><strong>Import modules</strong></p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[1]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="kn">import</span> <span class="nn">os</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="n">os</span><span class="o">.</span><span class="n">chdir</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">expanduser</span><span class="p">(</span><span class="s1">&#39;/Users/kt16/Documents/Github/dandelion&#39;</span><span class="p">))</span>
<span class="kn">import</span> <span class="nn">dandelion</span> <span class="k">as</span> <span class="nn">ddl</span>
<span class="c1"># change directory to somewhere more workable</span>
<span class="n">os</span><span class="o">.</span><span class="n">chdir</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">expanduser</span><span class="p">(</span><span class="s1">&#39;/Users/kt16/Downloads/dandelion_tutorial/&#39;</span><span class="p">))</span>
<span class="c1"># I&#39;m importing scanpy here to make use of its logging module.</span>
<span class="kn">import</span> <span class="nn">scanpy</span> <span class="k">as</span> <span class="nn">sc</span>
<span class="n">sc</span><span class="o">.</span><span class="n">settings</span><span class="o">.</span><span class="n">verbosity</span> <span class="o">=</span> <span class="mi">3</span>
<span class="kn">import</span> <span class="nn">warnings</span>
<span class="n">warnings</span><span class="o">.</span><span class="n">filterwarnings</span><span class="p">(</span><span class="s1">&#39;ignore&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<p><strong>Read in the previously saved files</strong></p>
<p>I will work with the same example from the previous notebook since I have the <code class="docutils literal notranslate"><span class="pre">AnnData</span></code> object saved and vdj table filtered.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[2]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">adata</span> <span class="o">=</span> <span class="n">sc</span><span class="o">.</span><span class="n">read_h5ad</span><span class="p">(</span><span class="s1">&#39;adata.h5ad&#39;</span><span class="p">)</span>
<span class="n">adata</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[2]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
AnnData object with n_obs × n_vars = 16492 × 1497
    obs: &#39;sampleid&#39;, &#39;batch&#39;, &#39;scrublet_score&#39;, &#39;n_genes&#39;, &#39;percent_mito&#39;, &#39;n_counts&#39;, &#39;is_doublet&#39;, &#39;filter_rna&#39;, &#39;has_bcr&#39;, &#39;filter_bcr_quality&#39;, &#39;filter_bcr_heavy&#39;, &#39;filter_bcr_light&#39;, &#39;bcr_QC_pass&#39;, &#39;filter_bcr&#39;, &#39;leiden&#39;
    var: &#39;feature_types&#39;, &#39;genome&#39;, &#39;gene_ids-0&#39;, &#39;gene_ids-1&#39;, &#39;gene_ids-2&#39;, &#39;gene_ids-3&#39;, &#39;n_cells&#39;, &#39;highly_variable&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;, &#39;mean&#39;, &#39;std&#39;
    uns: &#39;bcr_QC_pass_colors&#39;, &#39;hvg&#39;, &#39;leiden&#39;, &#39;leiden_colors&#39;, &#39;neighbors&#39;, &#39;pca&#39;, &#39;umap&#39;
    obsm: &#39;X_pca&#39;, &#39;X_umap&#39;
    varm: &#39;PCs&#39;
    obsp: &#39;connectivities&#39;, &#39;distances&#39;
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[3]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">vdj</span> <span class="o">=</span> <span class="n">ddl</span><span class="o">.</span><span class="n">read_h5</span><span class="p">(</span><span class="s1">&#39;dandelion_results.h5&#39;</span><span class="p">)</span>
<span class="n">vdj</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[3]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Dandelion class object with n_obs = 838 and n_contigs = 1700
    data: &#39;sequence_id&#39;, &#39;sequence&#39;, &#39;rev_comp&#39;, &#39;productive&#39;, &#39;v_call&#39;, &#39;d_call&#39;, &#39;j_call&#39;, &#39;sequence_alignment&#39;, &#39;germline_alignment&#39;, &#39;junction&#39;, &#39;junction_aa&#39;, &#39;v_cigar&#39;, &#39;d_cigar&#39;, &#39;j_cigar&#39;, &#39;stop_codon&#39;, &#39;vj_in_frame&#39;, &#39;locus&#39;, &#39;junction_length&#39;, &#39;np1_length&#39;, &#39;np2_length&#39;, &#39;v_sequence_start&#39;, &#39;v_sequence_end&#39;, &#39;v_germline_start&#39;, &#39;v_germline_end&#39;, &#39;d_sequence_start&#39;, &#39;d_sequence_end&#39;, &#39;d_germline_start&#39;, &#39;d_germline_end&#39;, &#39;j_sequence_start&#39;, &#39;j_sequence_end&#39;, &#39;j_germline_start&#39;, &#39;j_germline_end&#39;, &#39;v_score&#39;, &#39;v_identity&#39;, &#39;v_support&#39;, &#39;d_score&#39;, &#39;d_identity&#39;, &#39;d_support&#39;, &#39;j_score&#39;, &#39;j_identity&#39;, &#39;j_support&#39;, &#39;fwr1&#39;, &#39;fwr2&#39;, &#39;fwr3&#39;, &#39;fwr4&#39;, &#39;cdr1&#39;, &#39;cdr2&#39;, &#39;cdr3&#39;, &#39;cell_id&#39;, &#39;c_call&#39;, &#39;consensus_count&#39;, &#39;umi_count&#39;, &#39;v_call_10x&#39;, &#39;d_call_10x&#39;, &#39;j_call_10x&#39;, &#39;junction_10x&#39;, &#39;junction_10x_aa&#39;, &#39;v_call_genotyped&#39;, &#39;germline_alignment_d_mask&#39;, &#39;sample_id&#39;, &#39;c_sequence_alignment&#39;, &#39;c_germline_alignment&#39;, &#39;c_sequence_start&#39;, &#39;c_sequence_end&#39;, &#39;c_score&#39;, &#39;c_identity&#39;, &#39;c_support&#39;, &#39;c_call_10x&#39;, &#39;junction_aa_length&#39;, &#39;fwr1_aa&#39;, &#39;fwr2_aa&#39;, &#39;fwr3_aa&#39;, &#39;fwr4_aa&#39;, &#39;cdr1_aa&#39;, &#39;cdr2_aa&#39;, &#39;cdr3_aa&#39;, &#39;sequence_alignment_aa&#39;, &#39;v_sequence_alignment_aa&#39;, &#39;d_sequence_alignment_aa&#39;, &#39;j_sequence_alignment_aa&#39;, &#39;mu_freq&#39;, &#39;duplicate_count&#39;
    metadata: &#39;sample_id&#39;, &#39;isotype&#39;, &#39;lightchain&#39;, &#39;status&#39;, &#39;vdj_status&#39;, &#39;productive&#39;, &#39;umi_counts_heavy&#39;, &#39;umi_counts_light&#39;, &#39;v_call_heavy&#39;, &#39;v_call_light&#39;, &#39;j_call_heavy&#39;, &#39;j_call_light&#39;, &#39;c_call_heavy&#39;, &#39;c_call_light&#39;
    distance: None
    edges: None
    layout: None
    graph: None
</pre></div></div>
</div>
<div class="section" id="Quick-primer-to-the-Dandelion-class">
<h2>Quick primer to the <em>Dandelion</em> class<a class="headerlink" href="#Quick-primer-to-the-Dandelion-class" title="Permalink to this headline">¶</a></h2>
<p>So far, we have been operating with the <code class="docutils literal notranslate"><span class="pre">Dandelion</span></code> class:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[4]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">vdj</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[4]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Dandelion class object with n_obs = 838 and n_contigs = 1700
    data: &#39;sequence_id&#39;, &#39;sequence&#39;, &#39;rev_comp&#39;, &#39;productive&#39;, &#39;v_call&#39;, &#39;d_call&#39;, &#39;j_call&#39;, &#39;sequence_alignment&#39;, &#39;germline_alignment&#39;, &#39;junction&#39;, &#39;junction_aa&#39;, &#39;v_cigar&#39;, &#39;d_cigar&#39;, &#39;j_cigar&#39;, &#39;stop_codon&#39;, &#39;vj_in_frame&#39;, &#39;locus&#39;, &#39;junction_length&#39;, &#39;np1_length&#39;, &#39;np2_length&#39;, &#39;v_sequence_start&#39;, &#39;v_sequence_end&#39;, &#39;v_germline_start&#39;, &#39;v_germline_end&#39;, &#39;d_sequence_start&#39;, &#39;d_sequence_end&#39;, &#39;d_germline_start&#39;, &#39;d_germline_end&#39;, &#39;j_sequence_start&#39;, &#39;j_sequence_end&#39;, &#39;j_germline_start&#39;, &#39;j_germline_end&#39;, &#39;v_score&#39;, &#39;v_identity&#39;, &#39;v_support&#39;, &#39;d_score&#39;, &#39;d_identity&#39;, &#39;d_support&#39;, &#39;j_score&#39;, &#39;j_identity&#39;, &#39;j_support&#39;, &#39;fwr1&#39;, &#39;fwr2&#39;, &#39;fwr3&#39;, &#39;fwr4&#39;, &#39;cdr1&#39;, &#39;cdr2&#39;, &#39;cdr3&#39;, &#39;cell_id&#39;, &#39;c_call&#39;, &#39;consensus_count&#39;, &#39;umi_count&#39;, &#39;v_call_10x&#39;, &#39;d_call_10x&#39;, &#39;j_call_10x&#39;, &#39;junction_10x&#39;, &#39;junction_10x_aa&#39;, &#39;v_call_genotyped&#39;, &#39;germline_alignment_d_mask&#39;, &#39;sample_id&#39;, &#39;c_sequence_alignment&#39;, &#39;c_germline_alignment&#39;, &#39;c_sequence_start&#39;, &#39;c_sequence_end&#39;, &#39;c_score&#39;, &#39;c_identity&#39;, &#39;c_support&#39;, &#39;c_call_10x&#39;, &#39;junction_aa_length&#39;, &#39;fwr1_aa&#39;, &#39;fwr2_aa&#39;, &#39;fwr3_aa&#39;, &#39;fwr4_aa&#39;, &#39;cdr1_aa&#39;, &#39;cdr2_aa&#39;, &#39;cdr3_aa&#39;, &#39;sequence_alignment_aa&#39;, &#39;v_sequence_alignment_aa&#39;, &#39;d_sequence_alignment_aa&#39;, &#39;j_sequence_alignment_aa&#39;, &#39;mu_freq&#39;, &#39;duplicate_count&#39;
    metadata: &#39;sample_id&#39;, &#39;isotype&#39;, &#39;lightchain&#39;, &#39;status&#39;, &#39;vdj_status&#39;, &#39;productive&#39;, &#39;umi_counts_heavy&#39;, &#39;umi_counts_light&#39;, &#39;v_call_heavy&#39;, &#39;v_call_light&#39;, &#39;j_call_heavy&#39;, &#39;j_call_light&#39;, &#39;c_call_heavy&#39;, &#39;c_call_light&#39;
    distance: None
    edges: None
    layout: None
    graph: None
</pre></div></div>
</div>
<p>Essentially, the <code class="docutils literal notranslate"><span class="pre">.data</span></code> slot holds the AIRR contig table while the <code class="docutils literal notranslate"><span class="pre">.metadata</span></code> holds a collapsed version that is compatible with combining with <code class="docutils literal notranslate"><span class="pre">AnnData</span></code>’s <code class="docutils literal notranslate"><span class="pre">.obs</span></code> slot. The other slots will be gradually filled as we go through the notebook. You can retrieve these slots like a typical class object; for example, if I want the metadata:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[5]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">vdj</span><span class="o">.</span><span class="n">metadata</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[5]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>sample_id</th>
      <th>isotype</th>
      <th>lightchain</th>
      <th>status</th>
      <th>vdj_status</th>
      <th>productive</th>
      <th>umi_counts_heavy</th>
      <th>umi_counts_light</th>
      <th>v_call_heavy</th>
      <th>v_call_light</th>
      <th>j_call_heavy</th>
      <th>j_call_light</th>
      <th>c_call_heavy</th>
      <th>c_call_light</th>
    </tr>
    <tr>
      <th>cell_id</th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>sc5p_v2_hs_PBMC_1k_AACTGGTTCTCTAAGG</th>
      <td>sc5p_v2_hs_PBMC_1k</td>
      <td>IgM</td>
      <td>IgL</td>
      <td>IGH + IGL</td>
      <td>Single</td>
      <td>T</td>
      <td>17</td>
      <td>34.0</td>
      <td>IGHV3-15</td>
      <td>IGLV3-1</td>
      <td>IGHJ4</td>
      <td>IGLJ3,IGLJ2</td>
      <td>IGHM</td>
      <td>IGLC</td>
    </tr>
    <tr>
      <th>sc5p_v2_hs_PBMC_1k_AATCCAGTCAGTTGAC</th>
      <td>sc5p_v2_hs_PBMC_1k</td>
      <td>IgM</td>
      <td>IgK</td>
      <td>IGH + IGK</td>
      <td>Single</td>
      <td>T</td>
      <td>25</td>
      <td>36.0</td>
      <td>IGHV3-30</td>
      <td>IGKV3-20</td>
      <td>IGHJ6</td>
      <td>IGKJ3</td>
      <td>IGHM</td>
      <td>IGKC</td>
    </tr>
    <tr>
      <th>sc5p_v2_hs_PBMC_1k_AATCGGTGTTAGGGTG</th>
      <td>sc5p_v2_hs_PBMC_1k</td>
      <td>IgM</td>
      <td>IgK</td>
      <td>IGH + IGK</td>
      <td>Single</td>
      <td>T</td>
      <td>41</td>
      <td>41.0</td>
      <td>IGHV1-18</td>
      <td>IGKV4-1</td>
      <td>IGHJ4</td>
      <td>IGKJ5</td>
      <td>IGHM</td>
      <td>IGKC</td>
    </tr>
    <tr>
      <th>sc5p_v2_hs_PBMC_1k_ACACCGGGTTATTCTC</th>
      <td>sc5p_v2_hs_PBMC_1k</td>
      <td>IgM</td>
      <td>IgK</td>
      <td>IGH + IGK</td>
      <td>Single</td>
      <td>T</td>
      <td>24</td>
      <td>37.0</td>
      <td>IGHV3-23</td>
      <td>IGKV1-8</td>
      <td>IGHJ4</td>
      <td>IGKJ1</td>
      <td>IGHM</td>
      <td>IGKC</td>
    </tr>
    <tr>
      <th>sc5p_v2_hs_PBMC_1k_ACCTTTAAGACAAAGG</th>
      <td>sc5p_v2_hs_PBMC_1k</td>
      <td>IgM</td>
      <td>IgK</td>
      <td>IGH + IGK</td>
      <td>Single</td>
      <td>T</td>
      <td>34</td>
      <td>53.0</td>
      <td>IGHV1-18</td>
      <td>IGKV1-39,IGKV1D-39</td>
      <td>IGHJ4</td>
      <td>IGKJ2</td>
      <td>IGHM</td>
      <td>IGKC</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>vdj_nextgem_hs_pbmc3_TTGCGTCTCGTAGGTT</th>
      <td>vdj_nextgem_hs_pbmc3</td>
      <td>IgM</td>
      <td>IgK</td>
      <td>IGH + IGK</td>
      <td>Single</td>
      <td>T</td>
      <td>14</td>
      <td>65.0</td>
      <td>IGHV4-59</td>
      <td>IGKV3-11</td>
      <td>IGHJ6</td>
      <td>IGKJ5</td>
      <td>IGHM</td>
      <td>IGKC</td>
    </tr>
    <tr>
      <th>vdj_nextgem_hs_pbmc3_TTTATGCTCCGCATAA</th>
      <td>vdj_nextgem_hs_pbmc3</td>
      <td>IgM</td>
      <td>IgL</td>
      <td>IGH + IGL</td>
      <td>Single</td>
      <td>T</td>
      <td>31</td>
      <td>38.0</td>
      <td>IGHV3-7</td>
      <td>IGLV1-51</td>
      <td>IGHJ6</td>
      <td>IGLJ3,IGLJ2</td>
      <td>IGHM</td>
      <td>IGLC</td>
    </tr>
    <tr>
      <th>vdj_nextgem_hs_pbmc3_TTTATGCTCTGATTCT</th>
      <td>vdj_nextgem_hs_pbmc3</td>
      <td>IgM</td>
      <td>IgL</td>
      <td>IGH + IGL</td>
      <td>Single</td>
      <td>T</td>
      <td>16</td>
      <td>4.0</td>
      <td>IGHV2-26</td>
      <td>IGLV9-49</td>
      <td>IGHJ4</td>
      <td>IGLJ3</td>
      <td>IGHM</td>
      <td>IGLC</td>
    </tr>
    <tr>
      <th>vdj_nextgem_hs_pbmc3_TTTGCGCCACCATGTA</th>
      <td>vdj_nextgem_hs_pbmc3</td>
      <td>IgM</td>
      <td>IgK</td>
      <td>IGH + IGK</td>
      <td>Single</td>
      <td>T</td>
      <td>30</td>
      <td>38.0</td>
      <td>IGHV1-18</td>
      <td>IGKV1-27</td>
      <td>IGHJ4</td>
      <td>IGKJ4</td>
      <td>IGHM</td>
      <td>IGKC</td>
    </tr>
    <tr>
      <th>vdj_nextgem_hs_pbmc3_TTTGCGCTCCAGTATG</th>
      <td>vdj_nextgem_hs_pbmc3</td>
      <td>IgM</td>
      <td>IgL</td>
      <td>IGH + IGL</td>
      <td>Single</td>
      <td>T</td>
      <td>40</td>
      <td>83.0</td>
      <td>IGHV2-5</td>
      <td>IGLV1-47</td>
      <td>IGHJ4</td>
      <td>IGLJ3,IGLJ2</td>
      <td>IGHM</td>
      <td>IGLC</td>
    </tr>
  </tbody>
</table>
<p>838 rows × 14 columns</p>
</div></div>
</div>
<p>You can deep copy the <code class="docutils literal notranslate"><span class="pre">Dandelion</span></code> object to another variable which will inherit all slots:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[6]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">vdj2</span> <span class="o">=</span> <span class="n">vdj</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
<span class="n">vdj2</span><span class="o">.</span><span class="n">metadata</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[6]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>sample_id</th>
      <th>isotype</th>
      <th>lightchain</th>
      <th>status</th>
      <th>vdj_status</th>
      <th>productive</th>
      <th>umi_counts_heavy</th>
      <th>umi_counts_light</th>
      <th>v_call_heavy</th>
      <th>v_call_light</th>
      <th>j_call_heavy</th>
      <th>j_call_light</th>
      <th>c_call_heavy</th>
      <th>c_call_light</th>
    </tr>
    <tr>
      <th>cell_id</th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>sc5p_v2_hs_PBMC_1k_AACTGGTTCTCTAAGG</th>
      <td>sc5p_v2_hs_PBMC_1k</td>
      <td>IgM</td>
      <td>IgL</td>
      <td>IGH + IGL</td>
      <td>Single</td>
      <td>T</td>
      <td>17</td>
      <td>34.0</td>
      <td>IGHV3-15</td>
      <td>IGLV3-1</td>
      <td>IGHJ4</td>
      <td>IGLJ3,IGLJ2</td>
      <td>IGHM</td>
      <td>IGLC</td>
    </tr>
    <tr>
      <th>sc5p_v2_hs_PBMC_1k_AATCCAGTCAGTTGAC</th>
      <td>sc5p_v2_hs_PBMC_1k</td>
      <td>IgM</td>
      <td>IgK</td>
      <td>IGH + IGK</td>
      <td>Single</td>
      <td>T</td>
      <td>25</td>
      <td>36.0</td>
      <td>IGHV3-30</td>
      <td>IGKV3-20</td>
      <td>IGHJ6</td>
      <td>IGKJ3</td>
      <td>IGHM</td>
      <td>IGKC</td>
    </tr>
    <tr>
      <th>sc5p_v2_hs_PBMC_1k_AATCGGTGTTAGGGTG</th>
      <td>sc5p_v2_hs_PBMC_1k</td>
      <td>IgM</td>
      <td>IgK</td>
      <td>IGH + IGK</td>
      <td>Single</td>
      <td>T</td>
      <td>41</td>
      <td>41.0</td>
      <td>IGHV1-18</td>
      <td>IGKV4-1</td>
      <td>IGHJ4</td>
      <td>IGKJ5</td>
      <td>IGHM</td>
      <td>IGKC</td>
    </tr>
    <tr>
      <th>sc5p_v2_hs_PBMC_1k_ACACCGGGTTATTCTC</th>
      <td>sc5p_v2_hs_PBMC_1k</td>
      <td>IgM</td>
      <td>IgK</td>
      <td>IGH + IGK</td>
      <td>Single</td>
      <td>T</td>
      <td>24</td>
      <td>37.0</td>
      <td>IGHV3-23</td>
      <td>IGKV1-8</td>
      <td>IGHJ4</td>
      <td>IGKJ1</td>
      <td>IGHM</td>
      <td>IGKC</td>
    </tr>
    <tr>
      <th>sc5p_v2_hs_PBMC_1k_ACCTTTAAGACAAAGG</th>
      <td>sc5p_v2_hs_PBMC_1k</td>
      <td>IgM</td>
      <td>IgK</td>
      <td>IGH + IGK</td>
      <td>Single</td>
      <td>T</td>
      <td>34</td>
      <td>53.0</td>
      <td>IGHV1-18</td>
      <td>IGKV1-39,IGKV1D-39</td>
      <td>IGHJ4</td>
      <td>IGKJ2</td>
      <td>IGHM</td>
      <td>IGKC</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>vdj_nextgem_hs_pbmc3_TTGCGTCTCGTAGGTT</th>
      <td>vdj_nextgem_hs_pbmc3</td>
      <td>IgM</td>
      <td>IgK</td>
      <td>IGH + IGK</td>
      <td>Single</td>
      <td>T</td>
      <td>14</td>
      <td>65.0</td>
      <td>IGHV4-59</td>
      <td>IGKV3-11</td>
      <td>IGHJ6</td>
      <td>IGKJ5</td>
      <td>IGHM</td>
      <td>IGKC</td>
    </tr>
    <tr>
      <th>vdj_nextgem_hs_pbmc3_TTTATGCTCCGCATAA</th>
      <td>vdj_nextgem_hs_pbmc3</td>
      <td>IgM</td>
      <td>IgL</td>
      <td>IGH + IGL</td>
      <td>Single</td>
      <td>T</td>
      <td>31</td>
      <td>38.0</td>
      <td>IGHV3-7</td>
      <td>IGLV1-51</td>
      <td>IGHJ6</td>
      <td>IGLJ3,IGLJ2</td>
      <td>IGHM</td>
      <td>IGLC</td>
    </tr>
    <tr>
      <th>vdj_nextgem_hs_pbmc3_TTTATGCTCTGATTCT</th>
      <td>vdj_nextgem_hs_pbmc3</td>
      <td>IgM</td>
      <td>IgL</td>
      <td>IGH + IGL</td>
      <td>Single</td>
      <td>T</td>
      <td>16</td>
      <td>4.0</td>
      <td>IGHV2-26</td>
      <td>IGLV9-49</td>
      <td>IGHJ4</td>
      <td>IGLJ3</td>
      <td>IGHM</td>
      <td>IGLC</td>
    </tr>
    <tr>
      <th>vdj_nextgem_hs_pbmc3_TTTGCGCCACCATGTA</th>
      <td>vdj_nextgem_hs_pbmc3</td>
      <td>IgM</td>
      <td>IgK</td>
      <td>IGH + IGK</td>
      <td>Single</td>
      <td>T</td>
      <td>30</td>
      <td>38.0</td>
      <td>IGHV1-18</td>
      <td>IGKV1-27</td>
      <td>IGHJ4</td>
      <td>IGKJ4</td>
      <td>IGHM</td>
      <td>IGKC</td>
    </tr>
    <tr>
      <th>vdj_nextgem_hs_pbmc3_TTTGCGCTCCAGTATG</th>
      <td>vdj_nextgem_hs_pbmc3</td>
      <td>IgM</td>
      <td>IgL</td>
      <td>IGH + IGL</td>
      <td>Single</td>
      <td>T</td>
      <td>40</td>
      <td>83.0</td>
      <td>IGHV2-5</td>
      <td>IGLV1-47</td>
      <td>IGHJ4</td>
      <td>IGLJ3,IGLJ2</td>
      <td>IGHM</td>
      <td>IGLC</td>
    </tr>
  </tbody>
</table>
<p>838 rows × 14 columns</p>
</div></div>
</div>
<p><strong>updating metadata</strong></p>
<p>The <code class="docutils literal notranslate"><span class="pre">.metadata</span></code> slot in Dandelion class automatically initializes whenever the <code class="docutils literal notranslate"><span class="pre">.data</span></code> slot is filled. However, it only returns a standard number of columns that are pre-specified. To retrieve other columns in the <code class="docutils literal notranslate"><span class="pre">.data</span></code> slot, we can update the metadata with <code class="docutils literal notranslate"><span class="pre">ddl.update_metadata</span></code> and specify the option <code class="docutils literal notranslate"><span class="pre">retrieve</span></code>. you can also pass <code class="docutils literal notranslate"><span class="pre">collapse</span> <span class="pre">=</span> <span class="pre">True</span></code> and <code class="docutils literal notranslate"><span class="pre">split_heavy_light</span> <span class="pre">=</span> <span class="pre">False</span></code> if the two columns can be reasonably combined for that cell/barcode.</p>
<p><strong>Example 1 : retrieving junction amino acid sequences</strong></p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[7]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">ddl</span><span class="o">.</span><span class="n">update_metadata</span><span class="p">(</span><span class="n">vdj</span><span class="p">,</span> <span class="n">retrieve</span> <span class="o">=</span> <span class="s1">&#39;junction_aa&#39;</span><span class="p">)</span>
<span class="n">vdj</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[7]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Dandelion class object with n_obs = 838 and n_contigs = 1700
    data: &#39;sequence_id&#39;, &#39;sequence&#39;, &#39;rev_comp&#39;, &#39;productive&#39;, &#39;v_call&#39;, &#39;d_call&#39;, &#39;j_call&#39;, &#39;sequence_alignment&#39;, &#39;germline_alignment&#39;, &#39;junction&#39;, &#39;junction_aa&#39;, &#39;v_cigar&#39;, &#39;d_cigar&#39;, &#39;j_cigar&#39;, &#39;stop_codon&#39;, &#39;vj_in_frame&#39;, &#39;locus&#39;, &#39;junction_length&#39;, &#39;np1_length&#39;, &#39;np2_length&#39;, &#39;v_sequence_start&#39;, &#39;v_sequence_end&#39;, &#39;v_germline_start&#39;, &#39;v_germline_end&#39;, &#39;d_sequence_start&#39;, &#39;d_sequence_end&#39;, &#39;d_germline_start&#39;, &#39;d_germline_end&#39;, &#39;j_sequence_start&#39;, &#39;j_sequence_end&#39;, &#39;j_germline_start&#39;, &#39;j_germline_end&#39;, &#39;v_score&#39;, &#39;v_identity&#39;, &#39;v_support&#39;, &#39;d_score&#39;, &#39;d_identity&#39;, &#39;d_support&#39;, &#39;j_score&#39;, &#39;j_identity&#39;, &#39;j_support&#39;, &#39;fwr1&#39;, &#39;fwr2&#39;, &#39;fwr3&#39;, &#39;fwr4&#39;, &#39;cdr1&#39;, &#39;cdr2&#39;, &#39;cdr3&#39;, &#39;cell_id&#39;, &#39;c_call&#39;, &#39;consensus_count&#39;, &#39;umi_count&#39;, &#39;v_call_10x&#39;, &#39;d_call_10x&#39;, &#39;j_call_10x&#39;, &#39;junction_10x&#39;, &#39;junction_10x_aa&#39;, &#39;v_call_genotyped&#39;, &#39;germline_alignment_d_mask&#39;, &#39;sample_id&#39;, &#39;c_sequence_alignment&#39;, &#39;c_germline_alignment&#39;, &#39;c_sequence_start&#39;, &#39;c_sequence_end&#39;, &#39;c_score&#39;, &#39;c_identity&#39;, &#39;c_support&#39;, &#39;c_call_10x&#39;, &#39;junction_aa_length&#39;, &#39;fwr1_aa&#39;, &#39;fwr2_aa&#39;, &#39;fwr3_aa&#39;, &#39;fwr4_aa&#39;, &#39;cdr1_aa&#39;, &#39;cdr2_aa&#39;, &#39;cdr3_aa&#39;, &#39;sequence_alignment_aa&#39;, &#39;v_sequence_alignment_aa&#39;, &#39;d_sequence_alignment_aa&#39;, &#39;j_sequence_alignment_aa&#39;, &#39;mu_freq&#39;, &#39;duplicate_count&#39;
    metadata: &#39;sample_id&#39;, &#39;isotype&#39;, &#39;lightchain&#39;, &#39;status&#39;, &#39;vdj_status&#39;, &#39;productive&#39;, &#39;umi_counts_heavy&#39;, &#39;umi_counts_light&#39;, &#39;v_call_heavy&#39;, &#39;v_call_light&#39;, &#39;j_call_heavy&#39;, &#39;j_call_light&#39;, &#39;c_call_heavy&#39;, &#39;c_call_light&#39;, &#39;junction_aa_heavy&#39;, &#39;junction_aa_light&#39;
    distance: None
    edges: None
    layout: None
    graph: None
</pre></div></div>
</div>
<p>Note the additional <code class="docutils literal notranslate"><span class="pre">junction_aa</span></code> columns in the metadata slot. There’s a second example but this comes later on.</p>
<p><strong>Saving</strong></p>
<p>And as described in the previous notebook, dandelion class can be saved using <code class="docutils literal notranslate"><span class="pre">.write_h5</span></code> and <code class="docutils literal notranslate"><span class="pre">.write_pkl</span></code> functions with accompanying compression methods.</p>
</div>
<div class="section" id="Finding-clones">
<h2>Finding clones<a class="headerlink" href="#Finding-clones" title="Permalink to this headline">¶</a></h2>
<p>The following is <em>dandelion</em>’s implementation of a rather conventional method to define clones, <code class="docutils literal notranslate"><span class="pre">tl.find_clones</span></code>.</p>
<p>Clone definition is based on the following criterias:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="p">(</span><span class="mi">1</span><span class="p">)</span> <span class="n">Identical</span> <span class="n">IGH</span> <span class="n">V</span><span class="o">-</span><span class="n">J</span> <span class="n">gene</span> <span class="n">usage</span><span class="o">.</span>

<span class="p">(</span><span class="mi">2</span><span class="p">)</span> <span class="n">Identical</span> <span class="n">CDR3</span> <span class="n">junctional</span> <span class="n">sequence</span> <span class="n">length</span><span class="o">.</span>

<span class="p">(</span><span class="mi">3</span><span class="p">)</span> <span class="n">CDR3</span> <span class="n">Junctional</span> <span class="n">sequences</span> <span class="n">attains</span> <span class="n">a</span> <span class="n">minimum</span> <span class="n">of</span> <span class="o">%</span> <span class="n">sequence</span> <span class="n">similarity</span><span class="p">,</span> <span class="n">based</span> <span class="n">on</span> <span class="n">hamming</span> <span class="n">distance</span><span class="o">.</span> <span class="n">The</span> <span class="n">similarity</span> <span class="n">cut</span><span class="o">-</span><span class="n">off</span> <span class="ow">is</span> <span class="n">tunable</span> <span class="p">(</span><span class="n">default</span> <span class="ow">is</span> <span class="mi">85</span><span class="o">%</span><span class="p">)</span><span class="o">.</span>

<span class="p">(</span><span class="mi">4</span><span class="p">)</span> <span class="n">Light</span> <span class="n">chain</span> <span class="n">usage</span><span class="o">.</span> <span class="n">If</span> <span class="n">cells</span> <span class="n">within</span> <span class="n">clones</span> <span class="n">use</span> <span class="n">different</span> <span class="n">light</span> <span class="n">chains</span><span class="p">,</span> <span class="n">the</span> <span class="n">clone</span> <span class="n">will</span> <span class="n">be</span> <span class="n">splitted</span> <span class="n">following</span> <span class="n">the</span> <span class="n">same</span> <span class="n">conditions</span> <span class="k">for</span> <span class="n">heavy</span> <span class="n">chains</span> <span class="ow">in</span> <span class="p">(</span><span class="mi">1</span><span class="o">-</span><span class="mi">3</span><span class="p">)</span> <span class="k">as</span> <span class="n">above</span><span class="o">.</span>
</pre></div>
</div>
<p>The ‘clone_id’ name follows a <code class="docutils literal notranslate"><span class="pre">{A}_{B}_{C}_{D}</span></code> format and largely reflects the conditions above where:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="p">{</span><span class="n">A</span><span class="p">}</span> <span class="n">indicates</span> <span class="k">if</span> <span class="n">the</span> <span class="n">contigs</span> <span class="n">use</span> <span class="n">the</span> <span class="n">same</span> <span class="n">IGH</span> <span class="n">V</span><span class="o">/</span><span class="n">J</span> <span class="n">genes</span><span class="o">.</span>

<span class="p">{</span><span class="n">B</span><span class="p">}</span> <span class="n">indicates</span> <span class="k">if</span> <span class="n">IGH</span> <span class="n">junctional</span> <span class="n">sequences</span> <span class="n">are</span> <span class="n">equal</span> <span class="ow">in</span> <span class="n">length</span><span class="o">.</span>

<span class="p">{</span><span class="n">C</span><span class="p">}</span> <span class="n">indicates</span> <span class="k">if</span> <span class="n">clones</span> <span class="n">are</span> <span class="n">splitted</span> <span class="n">based</span> <span class="n">on</span> <span class="n">junctional</span> <span class="n">hamming</span> <span class="n">distance</span> <span class="n">threshold</span>

<span class="p">{</span><span class="n">D</span><span class="p">}</span> <span class="n">indicates</span> <span class="n">light</span> <span class="n">chain</span> <span class="n">pairing</span><span class="o">.</span>
</pre></div>
</div>
<p>The last position will not be annotated if there’s only one group of light chains usage detected in the clone.</p>
<div class="section" id="Running-tl.find_clones">
<h3>Running <code class="docutils literal notranslate"><span class="pre">tl.find_clones</span></code><a class="headerlink" href="#Running-tl.find_clones" title="Permalink to this headline">¶</a></h3>
<p>The function will take a file path, a pandas <code class="docutils literal notranslate"><span class="pre">DataFrame</span></code> (for example if you’ve used pandas to read in the filtered file already), or a <code class="docutils literal notranslate"><span class="pre">Dandelion</span></code> class object (described in next section). The default mode for calculation of junctional hamming distance is to use the junction amino acid sequences. If you want to do it via nucleotide, you can specify the option:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">clustering_by</span> <span class="o">=</span> <span class="s1">&#39;nt&#39;</span>
</pre></div>
</div>
<p>If you want to use the alleles for defining V-J gene usuage, specify:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">by_alleles</span> <span class="o">=</span> <span class="kc">True</span>
</pre></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[8]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">ddl</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">find_clones</span><span class="p">(</span><span class="n">vdj</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Finding clones
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
Finding clones based on heavy chains : 100%|██████████| 176/176 [00:00&lt;00:00, 2060.53it/s]
Refining clone assignment based on light chain pairing : 100%|██████████| 819/819 [00:00&lt;00:00, 2126.98it/s]
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
 finished: Updated Dandelion object:
   &#39;data&#39;, contig-indexed clone table
   &#39;metadata&#39;, cell-indexed clone table
 (0:00:01)
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[9]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">vdj</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[9]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Dandelion class object with n_obs = 838 and n_contigs = 1700
    data: &#39;sequence_id&#39;, &#39;sequence&#39;, &#39;rev_comp&#39;, &#39;productive&#39;, &#39;v_call&#39;, &#39;d_call&#39;, &#39;j_call&#39;, &#39;sequence_alignment&#39;, &#39;germline_alignment&#39;, &#39;junction&#39;, &#39;junction_aa&#39;, &#39;v_cigar&#39;, &#39;d_cigar&#39;, &#39;j_cigar&#39;, &#39;stop_codon&#39;, &#39;vj_in_frame&#39;, &#39;locus&#39;, &#39;junction_length&#39;, &#39;np1_length&#39;, &#39;np2_length&#39;, &#39;v_sequence_start&#39;, &#39;v_sequence_end&#39;, &#39;v_germline_start&#39;, &#39;v_germline_end&#39;, &#39;d_sequence_start&#39;, &#39;d_sequence_end&#39;, &#39;d_germline_start&#39;, &#39;d_germline_end&#39;, &#39;j_sequence_start&#39;, &#39;j_sequence_end&#39;, &#39;j_germline_start&#39;, &#39;j_germline_end&#39;, &#39;v_score&#39;, &#39;v_identity&#39;, &#39;v_support&#39;, &#39;d_score&#39;, &#39;d_identity&#39;, &#39;d_support&#39;, &#39;j_score&#39;, &#39;j_identity&#39;, &#39;j_support&#39;, &#39;fwr1&#39;, &#39;fwr2&#39;, &#39;fwr3&#39;, &#39;fwr4&#39;, &#39;cdr1&#39;, &#39;cdr2&#39;, &#39;cdr3&#39;, &#39;cell_id&#39;, &#39;c_call&#39;, &#39;consensus_count&#39;, &#39;umi_count&#39;, &#39;v_call_10x&#39;, &#39;d_call_10x&#39;, &#39;j_call_10x&#39;, &#39;junction_10x&#39;, &#39;junction_10x_aa&#39;, &#39;v_call_genotyped&#39;, &#39;germline_alignment_d_mask&#39;, &#39;sample_id&#39;, &#39;c_sequence_alignment&#39;, &#39;c_germline_alignment&#39;, &#39;c_sequence_start&#39;, &#39;c_sequence_end&#39;, &#39;c_score&#39;, &#39;c_identity&#39;, &#39;c_support&#39;, &#39;c_call_10x&#39;, &#39;junction_aa_length&#39;, &#39;fwr1_aa&#39;, &#39;fwr2_aa&#39;, &#39;fwr3_aa&#39;, &#39;fwr4_aa&#39;, &#39;cdr1_aa&#39;, &#39;cdr2_aa&#39;, &#39;cdr3_aa&#39;, &#39;sequence_alignment_aa&#39;, &#39;v_sequence_alignment_aa&#39;, &#39;d_sequence_alignment_aa&#39;, &#39;j_sequence_alignment_aa&#39;, &#39;mu_freq&#39;, &#39;duplicate_count&#39;, &#39;clone_id&#39;
    metadata: &#39;sample_id&#39;, &#39;clone_id&#39;, &#39;clone_id_by_size&#39;, &#39;isotype&#39;, &#39;lightchain&#39;, &#39;status&#39;, &#39;vdj_status&#39;, &#39;productive&#39;, &#39;umi_counts_heavy&#39;, &#39;umi_counts_light&#39;, &#39;v_call_heavy&#39;, &#39;v_call_light&#39;, &#39;j_call_heavy&#39;, &#39;j_call_light&#39;, &#39;c_call_heavy&#39;, &#39;c_call_light&#39;
    distance: None
    edges: None
    layout: None
    graph: None
</pre></div></div>
</div>
<p>This will return a new column with the column name <code class="docutils literal notranslate"><span class="pre">'clone_id'</span></code> as per convention. If a filepath is provided as input, it will also save the file automatically into the base directory of the file name. Otherwise, if a <code class="docutils literal notranslate"><span class="pre">Dandelion</span></code> object or <code class="docutils literal notranslate"><span class="pre">pandas.DataFrame</span></code> is provided it will just be returned as a <code class="docutils literal notranslate"><span class="pre">Dandelion</span></code> object.</p>
<p><strong>Updating metadata example 2 : editing clone_id column</strong></p>
<p>Perhaps you want to have a bit more control with how clones are called. We can edit this directly from the <code class="docutils literal notranslate"><span class="pre">.data</span></code> slot and retrieve accordingly.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[10]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># using a list comprehension to remove the light chain clone assignment</span>
<span class="n">vdj</span><span class="o">.</span><span class="n">data</span><span class="p">[</span><span class="s1">&#39;clone_id_heavyonly&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="n">x</span><span class="o">.</span><span class="n">rsplit</span><span class="p">(</span><span class="s1">&#39;_&#39;</span><span class="p">,</span> <span class="mi">1</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span> <span class="k">if</span> <span class="n">x</span><span class="o">.</span><span class="n">count</span><span class="p">(</span><span class="s1">&#39;_&#39;</span><span class="p">)</span> <span class="o">==</span> <span class="mi">3</span> <span class="k">else</span> <span class="n">x</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">vdj</span><span class="o">.</span><span class="n">data</span><span class="p">[</span><span class="s1">&#39;clone_id&#39;</span><span class="p">]]</span>
<span class="n">ddl</span><span class="o">.</span><span class="n">update_metadata</span><span class="p">(</span><span class="n">vdj</span><span class="p">,</span> <span class="n">retrieve</span> <span class="o">=</span> <span class="s1">&#39;clone_id_heavyonly&#39;</span><span class="p">,</span> <span class="n">split_heavy_light</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span> <span class="n">collapse</span> <span class="o">=</span> <span class="kc">True</span><span class="p">)</span>
<span class="n">vdj</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[10]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Dandelion class object with n_obs = 838 and n_contigs = 1700
    data: &#39;sequence_id&#39;, &#39;sequence&#39;, &#39;rev_comp&#39;, &#39;productive&#39;, &#39;v_call&#39;, &#39;d_call&#39;, &#39;j_call&#39;, &#39;sequence_alignment&#39;, &#39;germline_alignment&#39;, &#39;junction&#39;, &#39;junction_aa&#39;, &#39;v_cigar&#39;, &#39;d_cigar&#39;, &#39;j_cigar&#39;, &#39;stop_codon&#39;, &#39;vj_in_frame&#39;, &#39;locus&#39;, &#39;junction_length&#39;, &#39;np1_length&#39;, &#39;np2_length&#39;, &#39;v_sequence_start&#39;, &#39;v_sequence_end&#39;, &#39;v_germline_start&#39;, &#39;v_germline_end&#39;, &#39;d_sequence_start&#39;, &#39;d_sequence_end&#39;, &#39;d_germline_start&#39;, &#39;d_germline_end&#39;, &#39;j_sequence_start&#39;, &#39;j_sequence_end&#39;, &#39;j_germline_start&#39;, &#39;j_germline_end&#39;, &#39;v_score&#39;, &#39;v_identity&#39;, &#39;v_support&#39;, &#39;d_score&#39;, &#39;d_identity&#39;, &#39;d_support&#39;, &#39;j_score&#39;, &#39;j_identity&#39;, &#39;j_support&#39;, &#39;fwr1&#39;, &#39;fwr2&#39;, &#39;fwr3&#39;, &#39;fwr4&#39;, &#39;cdr1&#39;, &#39;cdr2&#39;, &#39;cdr3&#39;, &#39;cell_id&#39;, &#39;c_call&#39;, &#39;consensus_count&#39;, &#39;umi_count&#39;, &#39;v_call_10x&#39;, &#39;d_call_10x&#39;, &#39;j_call_10x&#39;, &#39;junction_10x&#39;, &#39;junction_10x_aa&#39;, &#39;v_call_genotyped&#39;, &#39;germline_alignment_d_mask&#39;, &#39;sample_id&#39;, &#39;c_sequence_alignment&#39;, &#39;c_germline_alignment&#39;, &#39;c_sequence_start&#39;, &#39;c_sequence_end&#39;, &#39;c_score&#39;, &#39;c_identity&#39;, &#39;c_support&#39;, &#39;c_call_10x&#39;, &#39;junction_aa_length&#39;, &#39;fwr1_aa&#39;, &#39;fwr2_aa&#39;, &#39;fwr3_aa&#39;, &#39;fwr4_aa&#39;, &#39;cdr1_aa&#39;, &#39;cdr2_aa&#39;, &#39;cdr3_aa&#39;, &#39;sequence_alignment_aa&#39;, &#39;v_sequence_alignment_aa&#39;, &#39;d_sequence_alignment_aa&#39;, &#39;j_sequence_alignment_aa&#39;, &#39;mu_freq&#39;, &#39;duplicate_count&#39;, &#39;clone_id&#39;, &#39;clone_id_heavyonly&#39;
    metadata: &#39;sample_id&#39;, &#39;clone_id&#39;, &#39;clone_id_by_size&#39;, &#39;isotype&#39;, &#39;lightchain&#39;, &#39;status&#39;, &#39;vdj_status&#39;, &#39;productive&#39;, &#39;umi_counts_heavy&#39;, &#39;umi_counts_light&#39;, &#39;v_call_heavy&#39;, &#39;v_call_light&#39;, &#39;j_call_heavy&#39;, &#39;j_call_light&#39;, &#39;c_call_heavy&#39;, &#39;c_call_light&#39;, &#39;clone_id_heavyonly&#39;
    distance: None
    edges: None
    layout: None
    graph: None
</pre></div></div>
</div>
</div>
<div class="section" id="Alternative-:-Running-tl.define_clones">
<h3>Alternative : Running <code class="docutils literal notranslate"><span class="pre">tl.define_clones</span></code><a class="headerlink" href="#Alternative-:-Running-tl.define_clones" title="Permalink to this headline">¶</a></h3>
<p>Alternatively, a wrapper to call <em>changeo’s</em> <code class="docutils literal notranslate"><span class="pre">DefineClones.py</span></code> is also included. To run it, you need to choose the distance threshold for clonal assignment. To facilitate this, the function <code class="docutils literal notranslate"><span class="pre">pp.calculate_threshold</span></code> will run <a class="reference external" href="https://shazam.readthedocs.io/en/stable/topics/distToNearest/">shazam’s distToNearest</a> function and return a plot showing the length normalized hamming distance distribution and automated threshold value. Again, <code class="docutils literal notranslate"><span class="pre">pp.calculate_threshold</span></code> will take a file path, pandas
<code class="docutils literal notranslate"><span class="pre">DataFrame</span></code> or <code class="docutils literal notranslate"><span class="pre">Dandelion</span></code> object as input. If a dandelion object is provided, the threshold value will be inserted into the <code class="docutils literal notranslate"><span class="pre">.threshold</span></code> slot. For more fine control, please use the <code class="docutils literal notranslate"><span class="pre">DefineClones.py</span></code> function <a class="reference external" href="https://immcantation.readthedocs.io/en/stable/tutorials/10x_tutorial.html">directly</a>.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[11]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">ddl</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">calculate_threshold</span><span class="p">(</span><span class="n">vdj</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Calculating threshold
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_3_dandelion_findingclones_and_analysis-10x_data_26_1.png" src="../_images/notebooks_3_dandelion_findingclones_and_analysis-10x_data_26_1.png" />
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
&lt;ggplot: (371799273)&gt;
 finished: Updated Dandelion object:
   &#39;threshold&#39;, threshold value for tuning clonal assignment
 (0:00:32)
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[12]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># see the actual value in .threshold slot</span>
<span class="n">vdj</span><span class="o">.</span><span class="n">threshold</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[12]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
0.21946248392024584
</pre></div></div>
</div>
<p>You can also manually select a value as the threshold if you wish.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[13]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">ddl</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">calculate_threshold</span><span class="p">(</span><span class="n">vdj</span><span class="p">,</span> <span class="n">manual_threshold</span> <span class="o">=</span> <span class="mf">0.1</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Calculating threshold
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_3_dandelion_findingclones_and_analysis-10x_data_29_1.png" src="../_images/notebooks_3_dandelion_findingclones_and_analysis-10x_data_29_1.png" />
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
&lt;ggplot: (371001341)&gt;
 finished: Updated Dandelion object:
   &#39;threshold&#39;, threshold value for tuning clonal assignment
 (0:00:34)
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[14]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># see the updated .threshold slot</span>
<span class="n">vdj</span><span class="o">.</span><span class="n">threshold</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[14]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
0.1
</pre></div></div>
</div>
<p>We can run <code class="docutils literal notranslate"><span class="pre">tl.define_clones</span></code> to call <em>changeo’s</em> <code class="docutils literal notranslate"><span class="pre">DefineClones.py</span></code>; see <a class="reference external" href="https://changeo.readthedocs.io/en/stable/methods/clustering.html">here</a> for more info. Note, if a <code class="docutils literal notranslate"><span class="pre">pandas.DataFrame</span></code> or file path is provided as the input, the value in dist option (corresponds to threshold value) needs to be manually supplied. If a <code class="docutils literal notranslate"><span class="pre">Dandelion</span></code> object is provided, it will automatically retrieve it from the threshold slot.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[15]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">ddl</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">define_clones</span><span class="p">(</span><span class="n">vdj</span><span class="p">,</span> <span class="n">key_added</span> <span class="o">=</span> <span class="s1">&#39;changeo_clone_id&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Finding clones
 finished: Updated Dandelion object:
   &#39;data&#39;, contig-indexed clone table
   &#39;metadata&#39;, cell-indexed clone table
 (0:00:05)
</pre></div></div>
</div>
<p>Note that I specified the option <code class="docutils literal notranslate"><span class="pre">key_added</span></code> and this adds the output from <code class="docutils literal notranslate"><span class="pre">tl.define_clones</span></code> into a separate column. If left as default (<code class="docutils literal notranslate"><span class="pre">None</span></code>), it will write into <code class="docutils literal notranslate"><span class="pre">clone_id</span></code> column. The same option can be specified in <code class="docutils literal notranslate"><span class="pre">tl.find_clones</span></code> earlier. If there is an existing column for <code class="docutils literal notranslate"><span class="pre">clone_id</span></code>, the group column for the new key would not be computed; If you want to extract the grouped column, simply run define clones without key_added to replace the existing <code class="docutils literal notranslate"><span class="pre">clone_id</span></code> and
<code class="docutils literal notranslate"><span class="pre">clone_id_group</span></code> columns.</p>
</div>
</div>
<div class="section" id="Visualization-of-BCR-network">
<h2>Visualization of BCR network<a class="headerlink" href="#Visualization-of-BCR-network" title="Permalink to this headline">¶</a></h2>
<p><em>dandelion</em> generates a network to facilitate visualisation of results. This uses the full V(D)J contig sequences instead of just the junctional sequences to chart a tree-like network for each clone. The actual visualization will be achieved through <code class="docutils literal notranslate"><span class="pre">scanpy</span></code> later.</p>
<p><strong>tl.generate_network</strong></p>
<p>First we need to generate the network. The tool function <code class="docutils literal notranslate"><span class="pre">tl.generate_network</span></code> will take a V(D)J table that has clones defined, specifically under the <code class="docutils literal notranslate"><span class="pre">'clone_id'</span></code> column. The default mode is to use amino acid sequences for constructing Levenshtein distance matrices.</p>
<p>If you have a pre-processed table parsed from immcantation’s <a class="reference external" href="https://changeo.readthedocs.io/en/stable/examples/10x.html">method</a>, or any other method as long as it’s in a <em>AIRR</em> format, the table can be used as well.</p>
<p>You can specify the <code class="docutils literal notranslate"><span class="pre">clone_key</span></code> option for generating the network for the clone id definition of choice as long as it exists as a column in the <code class="docutils literal notranslate"><span class="pre">.data</span></code> slot.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[16]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">ddl</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">generate_network</span><span class="p">(</span><span class="n">vdj</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Generating network
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
Calculating distances... : 100%|██████████| 4/4 [00:04&lt;00:00,  1.05s/it]
Generating edge list : 100%|██████████| 7/7 [00:00&lt;00:00, 653.23it/s]
Linking edges : 100%|██████████| 821/821 [00:00&lt;00:00, 4375.56it/s]
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
generating network layout
 finished: Updated Dandelion object:
   &#39;data&#39;, contig-indexed clone table
   &#39;metadata&#39;, cell-indexed clone table
   &#39;distance&#39;, heavy and light chain distance matrices
   &#39;edges&#39;, network edges
   &#39;layout&#39;, network layout
   &#39;graph&#39;, network (0:00:11)
</pre></div></div>
</div>
<p>This step works reasonably fast here but will take quite a while when a lot of contigs are provided.</p>
<p>You can also downsample the number of cells. This will return a new object as a downsampled copy of the original with it’s own distance matrix.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[17]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">vdj_downsample</span> <span class="o">=</span> <span class="n">ddl</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">generate_network</span><span class="p">(</span><span class="n">vdj</span><span class="p">,</span> <span class="n">downsample</span> <span class="o">=</span> <span class="mi">500</span><span class="p">)</span>
<span class="n">vdj_downsample</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Generating network
Downsampling to 500 cells.
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
Calculating distances... : 100%|██████████| 4/4 [00:01&lt;00:00,  2.09it/s]
Generating edge list : 100%|██████████| 4/4 [00:00&lt;00:00, 526.16it/s]
Linking edges : 100%|██████████| 491/491 [00:00&lt;00:00, 3719.38it/s]
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
generating network layout
 finished: Updated Dandelion object:
   &#39;data&#39;, contig-indexed clone table
   &#39;metadata&#39;, cell-indexed clone table
   &#39;distance&#39;, heavy and light chain distance matrices
   &#39;edges&#39;, network edges
   &#39;layout&#39;, network layout
   &#39;graph&#39;, network (0:00:07)
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[17]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Dandelion class object with n_obs = 500 and n_contigs = 1015
    data: &#39;sequence_id&#39;, &#39;sequence&#39;, &#39;rev_comp&#39;, &#39;productive&#39;, &#39;v_call&#39;, &#39;d_call&#39;, &#39;j_call&#39;, &#39;sequence_alignment&#39;, &#39;germline_alignment&#39;, &#39;junction&#39;, &#39;junction_aa&#39;, &#39;v_cigar&#39;, &#39;d_cigar&#39;, &#39;j_cigar&#39;, &#39;stop_codon&#39;, &#39;vj_in_frame&#39;, &#39;locus&#39;, &#39;junction_length&#39;, &#39;np1_length&#39;, &#39;np2_length&#39;, &#39;v_sequence_start&#39;, &#39;v_sequence_end&#39;, &#39;v_germline_start&#39;, &#39;v_germline_end&#39;, &#39;d_sequence_start&#39;, &#39;d_sequence_end&#39;, &#39;d_germline_start&#39;, &#39;d_germline_end&#39;, &#39;j_sequence_start&#39;, &#39;j_sequence_end&#39;, &#39;j_germline_start&#39;, &#39;j_germline_end&#39;, &#39;v_score&#39;, &#39;v_identity&#39;, &#39;v_support&#39;, &#39;d_score&#39;, &#39;d_identity&#39;, &#39;d_support&#39;, &#39;j_score&#39;, &#39;j_identity&#39;, &#39;j_support&#39;, &#39;fwr1&#39;, &#39;fwr2&#39;, &#39;fwr3&#39;, &#39;fwr4&#39;, &#39;cdr1&#39;, &#39;cdr2&#39;, &#39;cdr3&#39;, &#39;cell_id&#39;, &#39;c_call&#39;, &#39;consensus_count&#39;, &#39;umi_count&#39;, &#39;v_call_10x&#39;, &#39;d_call_10x&#39;, &#39;j_call_10x&#39;, &#39;junction_10x&#39;, &#39;junction_10x_aa&#39;, &#39;v_call_genotyped&#39;, &#39;germline_alignment_d_mask&#39;, &#39;sample_id&#39;, &#39;c_sequence_alignment&#39;, &#39;c_germline_alignment&#39;, &#39;c_sequence_start&#39;, &#39;c_sequence_end&#39;, &#39;c_score&#39;, &#39;c_identity&#39;, &#39;c_support&#39;, &#39;c_call_10x&#39;, &#39;junction_aa_length&#39;, &#39;fwr1_aa&#39;, &#39;fwr2_aa&#39;, &#39;fwr3_aa&#39;, &#39;fwr4_aa&#39;, &#39;cdr1_aa&#39;, &#39;cdr2_aa&#39;, &#39;cdr3_aa&#39;, &#39;sequence_alignment_aa&#39;, &#39;v_sequence_alignment_aa&#39;, &#39;d_sequence_alignment_aa&#39;, &#39;j_sequence_alignment_aa&#39;, &#39;mu_freq&#39;, &#39;duplicate_count&#39;, &#39;clone_id&#39;, &#39;clone_id_heavyonly&#39;, &#39;changeo_clone_id&#39;
    metadata: &#39;sample_id&#39;, &#39;clone_id&#39;, &#39;clone_id_by_size&#39;, &#39;isotype&#39;, &#39;lightchain&#39;, &#39;status&#39;, &#39;vdj_status&#39;, &#39;productive&#39;, &#39;umi_counts_heavy&#39;, &#39;umi_counts_light&#39;, &#39;v_call_heavy&#39;, &#39;v_call_light&#39;, &#39;j_call_heavy&#39;, &#39;j_call_light&#39;, &#39;c_call_heavy&#39;, &#39;c_call_light&#39;, &#39;changeo_clone_id&#39;
    distance: &#39;heavy&#39;, &#39;light_0&#39;, &#39;light_1&#39;, &#39;light_2&#39;
    edges: &#39;source&#39;, &#39;target&#39;, &#39;weight&#39;
    layout: layout for 500 vertices, layout for 13 vertices
    graph: networkx graph of 500 vertices, networkx graph of 13 vertices
</pre></div></div>
</div>
<p><strong>check the newly re-initialized Dandelion object</strong></p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[18]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">vdj</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[18]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Dandelion class object with n_obs = 838 and n_contigs = 1700
    data: &#39;sequence_id&#39;, &#39;sequence&#39;, &#39;rev_comp&#39;, &#39;productive&#39;, &#39;v_call&#39;, &#39;d_call&#39;, &#39;j_call&#39;, &#39;sequence_alignment&#39;, &#39;germline_alignment&#39;, &#39;junction&#39;, &#39;junction_aa&#39;, &#39;v_cigar&#39;, &#39;d_cigar&#39;, &#39;j_cigar&#39;, &#39;stop_codon&#39;, &#39;vj_in_frame&#39;, &#39;locus&#39;, &#39;junction_length&#39;, &#39;np1_length&#39;, &#39;np2_length&#39;, &#39;v_sequence_start&#39;, &#39;v_sequence_end&#39;, &#39;v_germline_start&#39;, &#39;v_germline_end&#39;, &#39;d_sequence_start&#39;, &#39;d_sequence_end&#39;, &#39;d_germline_start&#39;, &#39;d_germline_end&#39;, &#39;j_sequence_start&#39;, &#39;j_sequence_end&#39;, &#39;j_germline_start&#39;, &#39;j_germline_end&#39;, &#39;v_score&#39;, &#39;v_identity&#39;, &#39;v_support&#39;, &#39;d_score&#39;, &#39;d_identity&#39;, &#39;d_support&#39;, &#39;j_score&#39;, &#39;j_identity&#39;, &#39;j_support&#39;, &#39;fwr1&#39;, &#39;fwr2&#39;, &#39;fwr3&#39;, &#39;fwr4&#39;, &#39;cdr1&#39;, &#39;cdr2&#39;, &#39;cdr3&#39;, &#39;cell_id&#39;, &#39;c_call&#39;, &#39;consensus_count&#39;, &#39;umi_count&#39;, &#39;v_call_10x&#39;, &#39;d_call_10x&#39;, &#39;j_call_10x&#39;, &#39;junction_10x&#39;, &#39;junction_10x_aa&#39;, &#39;v_call_genotyped&#39;, &#39;germline_alignment_d_mask&#39;, &#39;sample_id&#39;, &#39;c_sequence_alignment&#39;, &#39;c_germline_alignment&#39;, &#39;c_sequence_start&#39;, &#39;c_sequence_end&#39;, &#39;c_score&#39;, &#39;c_identity&#39;, &#39;c_support&#39;, &#39;c_call_10x&#39;, &#39;junction_aa_length&#39;, &#39;fwr1_aa&#39;, &#39;fwr2_aa&#39;, &#39;fwr3_aa&#39;, &#39;fwr4_aa&#39;, &#39;cdr1_aa&#39;, &#39;cdr2_aa&#39;, &#39;cdr3_aa&#39;, &#39;sequence_alignment_aa&#39;, &#39;v_sequence_alignment_aa&#39;, &#39;d_sequence_alignment_aa&#39;, &#39;j_sequence_alignment_aa&#39;, &#39;mu_freq&#39;, &#39;duplicate_count&#39;, &#39;clone_id&#39;, &#39;clone_id_heavyonly&#39;, &#39;changeo_clone_id&#39;
    metadata: &#39;sample_id&#39;, &#39;clone_id&#39;, &#39;clone_id_by_size&#39;, &#39;isotype&#39;, &#39;lightchain&#39;, &#39;status&#39;, &#39;vdj_status&#39;, &#39;productive&#39;, &#39;umi_counts_heavy&#39;, &#39;umi_counts_light&#39;, &#39;v_call_heavy&#39;, &#39;v_call_light&#39;, &#39;j_call_heavy&#39;, &#39;j_call_light&#39;, &#39;c_call_heavy&#39;, &#39;c_call_light&#39;, &#39;changeo_clone_id&#39;
    distance: &#39;heavy&#39;, &#39;light_0&#39;, &#39;light_1&#39;, &#39;light_2&#39;
    edges: &#39;source&#39;, &#39;target&#39;, &#39;weight&#39;
    layout: layout for 838 vertices, layout for 24 vertices
    graph: networkx graph of 838 vertices, networkx graph of 24 vertices
</pre></div></div>
</div>
<p>The graph/networks can be accessed through the <code class="docutils literal notranslate"><span class="pre">.graph</span></code> slot as an <code class="docutils literal notranslate"><span class="pre">networkx</span></code> graph object if you want to extract the data for network statistics or make any changes to the network.</p>
</div>
<div class="section" id="Integration-with-scanpy">
<h2>Integration with scanpy<a class="headerlink" href="#Integration-with-scanpy" title="Permalink to this headline">¶</a></h2>
<p>The results can also be ported into the <code class="docutils literal notranslate"><span class="pre">AnnData</span></code> object for access to more plotting functions provided through <code class="docutils literal notranslate"><span class="pre">scanpy</span></code>.</p>
<p>To proceed, we first need to initialise the <code class="docutils literal notranslate"><span class="pre">AnnData</span></code> object with our network. This is done by using the tool function <code class="docutils literal notranslate"><span class="pre">tl.transfer</span></code>.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[19]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">ddl</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">transfer</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">vdj</span><span class="p">)</span> <span class="c1"># this will include singletons. To show only expanded clones, specify expanded_only=True</span>
<span class="n">adata</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Transferring network
converting matrices
Updating anndata slots
 finished: updated `.obs` with `.metadata`
added to `.uns[&#39;neighbors&#39;]` and `.obsp`
   &#39;distances&#39;, cluster-weighted adjacency matrix
   &#39;connectivities&#39;, cluster-weighted adjacency matrix (0:00:21)
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[19]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
AnnData object with n_obs × n_vars = 16492 × 1497
    obs: &#39;sampleid&#39;, &#39;batch&#39;, &#39;scrublet_score&#39;, &#39;n_genes&#39;, &#39;percent_mito&#39;, &#39;n_counts&#39;, &#39;is_doublet&#39;, &#39;filter_rna&#39;, &#39;has_bcr&#39;, &#39;filter_bcr_quality&#39;, &#39;filter_bcr_heavy&#39;, &#39;filter_bcr_light&#39;, &#39;bcr_QC_pass&#39;, &#39;filter_bcr&#39;, &#39;leiden&#39;, &#39;sample_id&#39;, &#39;clone_id&#39;, &#39;clone_id_by_size&#39;, &#39;isotype&#39;, &#39;lightchain&#39;, &#39;status&#39;, &#39;vdj_status&#39;, &#39;productive&#39;, &#39;umi_counts_heavy&#39;, &#39;umi_counts_light&#39;, &#39;v_call_heavy&#39;, &#39;v_call_light&#39;, &#39;j_call_heavy&#39;, &#39;j_call_light&#39;, &#39;c_call_heavy&#39;, &#39;c_call_light&#39;, &#39;changeo_clone_id&#39;
    var: &#39;feature_types&#39;, &#39;genome&#39;, &#39;gene_ids-0&#39;, &#39;gene_ids-1&#39;, &#39;gene_ids-2&#39;, &#39;gene_ids-3&#39;, &#39;n_cells&#39;, &#39;highly_variable&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;, &#39;mean&#39;, &#39;std&#39;
    uns: &#39;bcr_QC_pass_colors&#39;, &#39;hvg&#39;, &#39;leiden&#39;, &#39;leiden_colors&#39;, &#39;neighbors&#39;, &#39;pca&#39;, &#39;umap&#39;, &#39;rna_neighbors&#39;
    obsm: &#39;X_pca&#39;, &#39;X_umap&#39;, &#39;X_bcr&#39;
    varm: &#39;PCs&#39;
    obsp: &#39;connectivities&#39;, &#39;distances&#39;, &#39;rna_connectivities&#39;, &#39;rna_distances&#39;, &#39;bcr_connectivities&#39;, &#39;bcr_distances&#39;
</pre></div></div>
</div>
<p>You can see that <code class="docutils literal notranslate"><span class="pre">AnnData</span></code> object now contains a couple more columns in the <code class="docutils literal notranslate"><span class="pre">.obs</span></code> slot, corresponding to the metadata that is returned after <code class="docutils literal notranslate"><span class="pre">tl.generate_network</span></code>, and newly populated <code class="docutils literal notranslate"><span class="pre">.obsm</span></code> and <code class="docutils literal notranslate"><span class="pre">.obsp</span></code> slots. The original RNA connectivities and distances are now added into the <code class="docutils literal notranslate"><span class="pre">.obsp</span></code> slot as well.</p>
<p><strong>Plotting in scanpy</strong></p>
<p><strong>pl.clone_network</strong></p>
<p>So now, basically we can plot in <code class="docutils literal notranslate"><span class="pre">scanpy</span></code> with their plotting modules. I’ve included a plotting function in <strong>dandelion</strong>, <code class="docutils literal notranslate"><span class="pre">pl.clone_network</span></code>, which is really just a wrapper of their <code class="docutils literal notranslate"><span class="pre">pl.embedding</span></code> module.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[20]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">sc</span><span class="o">.</span><span class="n">set_figure_params</span><span class="p">(</span><span class="n">figsize</span> <span class="o">=</span> <span class="p">[</span><span class="mi">4</span><span class="p">,</span><span class="mi">4</span><span class="p">])</span>
<span class="n">ddl</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">clone_network</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span>
                     <span class="n">color</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;sampleid&#39;</span><span class="p">],</span>
                     <span class="n">edges_width</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span>
                     <span class="n">size</span> <span class="o">=</span> <span class="mi">50</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
... storing &#39;sample_id&#39; as categorical
... storing &#39;clone_id&#39; as categorical
... storing &#39;isotype&#39; as categorical
... storing &#39;lightchain&#39; as categorical
... storing &#39;status&#39; as categorical
... storing &#39;vdj_status&#39; as categorical
... storing &#39;productive&#39; as categorical
... storing &#39;umi_counts_light&#39; as categorical
... storing &#39;v_call_heavy&#39; as categorical
... storing &#39;v_call_light&#39; as categorical
... storing &#39;j_call_heavy&#39; as categorical
... storing &#39;j_call_light&#39; as categorical
... storing &#39;c_call_heavy&#39; as categorical
... storing &#39;c_call_light&#39; as categorical
... storing &#39;changeo_clone_id&#39; as categorical
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_3_dandelion_findingclones_and_analysis-10x_data_46_1.png" class="no-scaled-link" src="../_images/notebooks_3_dandelion_findingclones_and_analysis-10x_data_46_1.png" style="width: 472px; height: 296px;" />
</div>
</div>
<p><strong>tl.extract_edge_weights</strong></p>
<p><strong>dandelion</strong> provides an edge weight extractor tool <code class="docutils literal notranslate"><span class="pre">tl.extract_edge_weights</span></code> to retrieve the edge weights that can be used to specify the edge widths according to weight/distance.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[21]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">edgeweights</span> <span class="o">=</span> <span class="p">[</span><span class="mi">1</span><span class="o">/</span><span class="p">(</span><span class="n">e</span><span class="o">+</span><span class="mi">1</span><span class="p">)</span> <span class="k">for</span> <span class="n">e</span> <span class="ow">in</span> <span class="n">ddl</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">extract_edge_weights</span><span class="p">(</span><span class="n">vdj</span><span class="p">)]</span> <span class="c1"># add 1 to each edge weight (e) so that distance of 0 becomes the thickest edge</span>
<span class="n">ddl</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">clone_network</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span>
                     <span class="n">color</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;isotype&#39;</span><span class="p">],</span>
                     <span class="n">legend_fontoutline</span><span class="o">=</span><span class="mi">3</span><span class="p">,</span>
                     <span class="n">edges_width</span> <span class="o">=</span> <span class="n">edgeweights</span><span class="p">,</span>
                     <span class="n">size</span> <span class="o">=</span> <span class="mi">50</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_3_dandelion_findingclones_and_analysis-10x_data_49_0.png" class="no-scaled-link" src="../_images/notebooks_3_dandelion_findingclones_and_analysis-10x_data_49_0.png" style="width: 339px; height: 296px;" />
</div>
</div>
<p>You can interact with <code class="docutils literal notranslate"><span class="pre">pl.clone_network</span></code> just as how you interact with the rest of the scatterplot modules in <code class="docutils literal notranslate"><span class="pre">scanpy</span></code>.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[22]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="kn">from</span> <span class="nn">scanpy.plotting.palettes</span> <span class="kn">import</span> <span class="n">default_28</span><span class="p">,</span> <span class="n">default_102</span>
<span class="n">sc</span><span class="o">.</span><span class="n">set_figure_params</span><span class="p">(</span><span class="n">figsize</span> <span class="o">=</span> <span class="p">[</span><span class="mi">4</span><span class="p">,</span><span class="mi">4</span><span class="p">])</span>
<span class="c1"># plot the 3 largest clones by size</span>
<span class="n">ddl</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">clone_network</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">color</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;clone_id_by_size&#39;</span><span class="p">],</span> <span class="n">groups</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;1&#39;</span><span class="p">,</span> <span class="s1">&#39;2&#39;</span><span class="p">,</span> <span class="s1">&#39;3&#39;</span><span class="p">],</span> <span class="n">ncols</span> <span class="o">=</span> <span class="mi">2</span><span class="p">,</span> <span class="n">legend_loc</span> <span class="o">=</span> <span class="s1">&#39;on data&#39;</span><span class="p">,</span> <span class="n">legend_fontoutline</span><span class="o">=</span><span class="mi">3</span><span class="p">,</span> <span class="n">edges_width</span> <span class="o">=</span> <span class="n">edgeweights</span><span class="p">,</span> <span class="n">size</span> <span class="o">=</span> <span class="mi">50</span><span class="p">,</span> <span class="n">palette</span> <span class="o">=</span> <span class="n">default_28</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
WARNING: Length of palette colors is smaller than the number of categories (palette length: 28, categories length: 821. Some categories will have the same color.
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_3_dandelion_findingclones_and_analysis-10x_data_51_1.png" class="no-scaled-link" src="../_images/notebooks_3_dandelion_findingclones_and_analysis-10x_data_51_1.png" style="width: 284px; height: 296px;" />
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[23]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">ddl</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">clone_network</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">color</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;status&#39;</span><span class="p">,</span> <span class="s1">&#39;vdj_status&#39;</span><span class="p">],</span> <span class="n">ncols</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="n">legend_fontoutline</span><span class="o">=</span><span class="mi">3</span><span class="p">,</span> <span class="n">edges_width</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="n">size</span> <span class="o">=</span> <span class="mi">50</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_3_dandelion_findingclones_and_analysis-10x_data_52_0.png" class="no-scaled-link" src="../_images/notebooks_3_dandelion_findingclones_and_analysis-10x_data_52_0.png" style="width: 502px; height: 577px;" />
</div>
</div>
<p>By specifying <code class="docutils literal notranslate"><span class="pre">expanded_only=True</span></code>, you will transfer the graph of only expanded clones (&gt; 1 cell in a clone).</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[24]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">ddl</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">transfer</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">vdj</span><span class="p">)</span>
<span class="n">edgeweights</span> <span class="o">=</span> <span class="p">[</span><span class="mi">1</span><span class="o">/</span><span class="p">(</span><span class="n">e</span><span class="o">+</span><span class="mi">1</span><span class="p">)</span> <span class="k">for</span> <span class="n">e</span> <span class="ow">in</span> <span class="n">ddl</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">extract_edge_weights</span><span class="p">(</span><span class="n">vdj</span><span class="p">)]</span> <span class="c1"># add 1 to each edge weight (e) so that distance of 0 becomes the thickest edge</span>
<span class="n">ddl</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">clone_network</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">color</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;clone_id_by_size&#39;</span><span class="p">],</span> <span class="n">groups</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;1&#39;</span><span class="p">,</span> <span class="s1">&#39;2&#39;</span><span class="p">,</span> <span class="s1">&#39;3&#39;</span><span class="p">,</span> <span class="s1">&#39;4&#39;</span><span class="p">,</span> <span class="s1">&#39;5&#39;</span><span class="p">,</span> <span class="s1">&#39;6&#39;</span><span class="p">,</span> <span class="s1">&#39;7&#39;</span><span class="p">],</span> <span class="n">legend_loc</span> <span class="o">=</span> <span class="s1">&#39;on data&#39;</span><span class="p">,</span> <span class="n">legend_fontoutline</span><span class="o">=</span><span class="mi">3</span><span class="p">,</span> <span class="n">edges_width</span> <span class="o">=</span> <span class="n">edgeweights</span><span class="p">,</span> <span class="n">size</span> <span class="o">=</span> <span class="mi">50</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Transferring network
converting matrices
Updating anndata slots
 finished: updated `.obs` with `.metadata`
added to `.uns[&#39;neighbors&#39;]` and `.obsp`
   &#39;distances&#39;, cluster-weighted adjacency matrix
   &#39;connectivities&#39;, cluster-weighted adjacency matrix (0:00:18)
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_3_dandelion_findingclones_and_analysis-10x_data_54_1.png" class="no-scaled-link" src="../_images/notebooks_3_dandelion_findingclones_and_analysis-10x_data_54_1.png" style="width: 284px; height: 296px;" />
</div>
</div>
</div>
<div class="section" id="Calculating-mutational-load">
<h2>Calculating mutational load<a class="headerlink" href="#Calculating-mutational-load" title="Permalink to this headline">¶</a></h2>
<p>To calculate mutational load, I’ve ported the functions from immcantation suite’s changeo and shazam to work with the <code class="docutils literal notranslate"><span class="pre">dandelion</span></code> class object.</p>
<p>This can be run immediately after <code class="docutils literal notranslate"><span class="pre">pp.reassign_alleles</span></code> in during pre-processing because the required germline columns should be present in the genotyped <code class="docutils literal notranslate"><span class="pre">.tsv</span></code> file. I would reccomend to run this after TIgGER, after the v_calls were corrected.</p>
<p>But say you want to run it now instead, this can be achieved:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[26]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># let&#39;s recreate the vdj object with only the first two samples</span>
<span class="n">subset_data</span> <span class="o">=</span> <span class="n">vdj</span><span class="o">.</span><span class="n">data</span><span class="p">[</span><span class="n">vdj</span><span class="o">.</span><span class="n">data</span><span class="p">[</span><span class="s1">&#39;sample_id&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">isin</span><span class="p">([</span><span class="s1">&#39;sc5p_v2_hs_PBMC_1k&#39;</span><span class="p">,</span> <span class="s1">&#39;sc5p_v2_hs_PBMC_10k&#39;</span><span class="p">])]</span>
<span class="n">subset_data</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[26]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>sequence_id</th>
      <th>sequence</th>
      <th>rev_comp</th>
      <th>productive</th>
      <th>v_call</th>
      <th>d_call</th>
      <th>j_call</th>
      <th>sequence_alignment</th>
      <th>germline_alignment</th>
      <th>junction</th>
      <th>...</th>
      <th>cdr2_aa</th>
      <th>cdr3_aa</th>
      <th>sequence_alignment_aa</th>
      <th>v_sequence_alignment_aa</th>
      <th>d_sequence_alignment_aa</th>
      <th>j_sequence_alignment_aa</th>
      <th>mu_freq</th>
      <th>duplicate_count</th>
      <th>clone_id</th>
      <th>changeo_clone_id</th>
    </tr>
    <tr>
      <th>sequence_id</th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>sc5p_v2_hs_PBMC_1k_AACTGGTTCTCTAAGG_contig_1</th>
      <td>sc5p_v2_hs_PBMC_1k_AACTGGTTCTCTAAGG_contig_1</td>
      <td>CTGGGCCTCAGGAAGCAGCATCGGAGGTGCCTCAGCCATGGCATGG...</td>
      <td>F</td>
      <td>T</td>
      <td>IGLV3-1*01</td>
      <td></td>
      <td>IGLJ2*01,IGLJ3*01</td>
      <td>TCCTATGAGCTGACTCAGCCACCCTCA...GTGTCCGTGTCCCCAG...</td>
      <td>TCCTATGAGCTGACTCAGCCACCCTCA...GTGTCCGTGTCCCCAG...</td>
      <td>TGTCAGGCGTGGGACAGCAGCAATGTGGTATTC</td>
      <td>...</td>
      <td>QDN</td>
      <td>QAWDSSNVV</td>
      <td>SYELTQPPSVSVSPGQTASITCSGDKLGHKYACWYQQKPGQSPVLV...</td>
      <td>SYELTQPPSVSVSPGQTASITCSGDKLGHKYACWYQQKPGQSPVLV...</td>
      <td></td>
      <td>VFGGGTKLTVL</td>
      <td>0.009434</td>
      <td>0</td>
      <td>121_8_3</td>
      <td>57_0</td>
    </tr>
    <tr>
      <th>sc5p_v2_hs_PBMC_1k_AACTGGTTCTCTAAGG_contig_2</th>
      <td>sc5p_v2_hs_PBMC_1k_AACTGGTTCTCTAAGG_contig_2</td>
      <td>AGCTCTGGGAGAGGAGCCCCAGCCTTGGGATTCCCAAGTGTTTTCA...</td>
      <td>F</td>
      <td>T</td>
      <td>IGHV3-15*01</td>
      <td>IGHD4-23*01</td>
      <td>IGHJ4*02</td>
      <td>GAGGTGCAGCTGGTGGAGTCTGGGGGA...GGCTTGGTAAAGCCTG...</td>
      <td>GAGGTGCAGCTGGTGGAGTCTGGGGGA...GGCTTGGTAAAGCCTG...</td>
      <td>TGTACCACAGAAGCTTTAGACTACGGTGCTAACTCGCGATCCCCGA...</td>
      <td>...</td>
      <td>IKSNTDGATT</td>
      <td>TTEALDYGANSRSPNFDY</td>
      <td>EVQLVESGGGLVKPGGSLRLSCAASGFIFSNAWMSWVRQAPGKGLE...</td>
      <td>EVQLVESGGGLVKPGGSLRLSCAASGFIFSNAWMSWVRQAPGKGLE...</td>
      <td>DYGANS</td>
      <td>FDYWGQGTLVTVSS</td>
      <td>0.008671</td>
      <td>0</td>
      <td>121_8_3</td>
      <td>57_0</td>
    </tr>
    <tr>
      <th>sc5p_v2_hs_PBMC_1k_AATCCAGTCAGTTGAC_contig_1</th>
      <td>sc5p_v2_hs_PBMC_1k_AATCCAGTCAGTTGAC_contig_1</td>
      <td>AGAGCTCTGGAGAAGAGCTGCTCAGTTAGGACCCAGAGGGAACCAT...</td>
      <td>F</td>
      <td>T</td>
      <td>IGKV3-20*01</td>
      <td></td>
      <td>IGKJ3*01</td>
      <td>GAAATTGTGTTGACGCAGTCTCCAGGCACCCTGTCTTTGTCTCCAG...</td>
      <td>GAAATTGTGTTGACGCAGTCTCCAGGCACCCTGTCTTTGTCTCCAG...</td>
      <td>TGTCAGCAGTATGGTAGCTCACCTCCATTCACTTTC</td>
      <td>...</td>
      <td>GAS</td>
      <td>QQYGSSPPFT</td>
      <td>EIVLTQSPGTLSLSPGERATLSCRASQSVSSSYLAWYQQKPGQAPR...</td>
      <td>EIVLTQSPGTLSLSPGERATLSCRASQSVSSSYLAWYQQKPGQAPR...</td>
      <td></td>
      <td>FTFGPGTKVDIK</td>
      <td>0.000000</td>
      <td>0</td>
      <td>23_6_2</td>
      <td>76_1</td>
    </tr>
    <tr>
      <th>sc5p_v2_hs_PBMC_1k_AATCCAGTCAGTTGAC_contig_2</th>
      <td>sc5p_v2_hs_PBMC_1k_AATCCAGTCAGTTGAC_contig_2</td>
      <td>GAGCTCTGGGAGAGGAGCCCAGCACTAGAAGTCGGCGGTGTTTCCA...</td>
      <td>F</td>
      <td>T</td>
      <td>IGHV3-30*02,IGHV3-30-5*02</td>
      <td>IGHD1-26*01</td>
      <td>IGHJ6*02</td>
      <td>CAGGTGCAGCTGGTGGAGTCTGGGGGA...GGCGTGGTCCAGCCTG...</td>
      <td>CAGGTGCAGCTGGTGGAGTCTGGGGGA...GGCGTGGTCCAGCCTG...</td>
      <td>TGTGCGAAAGATACTGAAGTGGGAGCGAGCCGATACTACTACTACT...</td>
      <td>...</td>
      <td>IRYDGSNK</td>
      <td>AKDTEVGASRYYYYYGMDV</td>
      <td>QVQLVESGGGVVQPGGSLRLSCAASGFTFSSYGMHWVRQAPGKGLE...</td>
      <td>QVQLVESGGGVVQPGGSLRLSCAASGFTFSSYGMHWVRQAPGKGLE...</td>
      <td>VGA</td>
      <td>YYYYYGMDVWGQGTTVTVSS</td>
      <td>0.000000</td>
      <td>0</td>
      <td>23_6_2</td>
      <td>76_1</td>
    </tr>
    <tr>
      <th>sc5p_v2_hs_PBMC_1k_AATCGGTGTTAGGGTG_contig_1</th>
      <td>sc5p_v2_hs_PBMC_1k_AATCGGTGTTAGGGTG_contig_1</td>
      <td>GAGCTACAACAGGCAGGCAGGGGCAGCAAGATGGTGTTGCAGACCC...</td>
      <td>F</td>
      <td>T</td>
      <td>IGKV4-1*01</td>
      <td></td>
      <td>IGKJ5*01</td>
      <td>GACATCGTGATGACCCAGTCTCCAGACTCCCTGGCTGTGTCTCTGG...</td>
      <td>GACATCGTGATGACCCAGTCTCCAGACTCCCTGGCTGTGTCTCTGG...</td>
      <td>TGTCAGCAATATTATAGTACTCCGATCACCTTC</td>
      <td>...</td>
      <td>WAS</td>
      <td>QQYYSTPIT</td>
      <td>DIVMTQSPDSLAVSLGERATINCKSSQSVLYSSNNKNYLAWYQQKP...</td>
      <td>DIVMTQSPDSLAVSLGERATINCKSSQSVLYSSNNKNYLAWYQQKP...</td>
      <td></td>
      <td>ITFGQGTRLEIK</td>
      <td>0.000000</td>
      <td>0</td>
      <td>84_3_2</td>
      <td>184_2</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>sc5p_v2_hs_PBMC_10k_TTTCCTCTCTTCAACT_contig_1</th>
      <td>sc5p_v2_hs_PBMC_10k_TTTCCTCTCTTCAACT_contig_1</td>
      <td>TTTCCTCTCTTCAACTGCGAACCGACTTTCTGCGATGGGGACTCAA...</td>
      <td>F</td>
      <td>T</td>
      <td>IGHV1-8*01</td>
      <td>IGHD2-2*02</td>
      <td>IGHJ6*02</td>
      <td>CAGGTGCAGCTGGTGCAGTCTGGGGCT...GAGGTGAAGAAGCCTG...</td>
      <td>CAGGTGCAGCTGGTGCAGTCTGGGGCT...GAGGTGAAGAAGCCTG...</td>
      <td>TGTGCGAGATATTGTAGTAGTACCAGCTGCTATACGACCTATTACT...</td>
      <td>...</td>
      <td>MNPNSGNT</td>
      <td>ARYCSSTSCYTTYYYYYYGMDV</td>
      <td>QVQLVQSGAEVKKPGASVKVSCKASGYTFTSYDINWVRQATGQGLE...</td>
      <td>QVQLVQSGAEVKKPGASVKVSCKASGYTFTSYDINWVRQATGQGLE...</td>
      <td>YCSSTSCYT</td>
      <td>YYYYYGMDVWGQGTTVTVSS</td>
      <td>0.000000</td>
      <td>0</td>
      <td>28_6_2</td>
      <td>757_448</td>
    </tr>
    <tr>
      <th>sc5p_v2_hs_PBMC_10k_TTTGGTTTCAGAGCTT_contig_2</th>
      <td>sc5p_v2_hs_PBMC_10k_TTTGGTTTCAGAGCTT_contig_2</td>
      <td>GGGAGAGCCCTGGGGAGGAACTGCTCAGTTAGGACCCAGAGGGAAC...</td>
      <td>F</td>
      <td>T</td>
      <td>IGKV3-11*01</td>
      <td></td>
      <td>IGKJ5*01</td>
      <td>GAAATTGTGTTGACACAGTCTCCAGCCACCCTGTCTTTGTCTCCAG...</td>
      <td>GAAATTGTGTTGACACAGTCTCCAGCCACCCTGTCTTTGTCTCCAG...</td>
      <td>TGTCAGCAGCGTAGCAACTGGCTCACCTTC</td>
      <td>...</td>
      <td>DAS</td>
      <td>QQRSNWLT</td>
      <td>EIVLTQSPATLSLSPGERATLSCRASQSVSSYLAWYQQKPGQAPRL...</td>
      <td>EIVLTQSPATLSLSPGERATLSCRASQSVSSYLAWYQQKPGQAPRL...</td>
      <td></td>
      <td>TFGQGTRLEIK</td>
      <td>0.000000</td>
      <td>0</td>
      <td>159_2_1</td>
      <td>425_449</td>
    </tr>
    <tr>
      <th>sc5p_v2_hs_PBMC_10k_TTTGGTTTCAGAGCTT_contig_1</th>
      <td>sc5p_v2_hs_PBMC_10k_TTTGGTTTCAGAGCTT_contig_1</td>
      <td>ACAACCACACCCCTCCTAAGAAGAAGCCCCTAGACCACAGCTCCAC...</td>
      <td>F</td>
      <td>T</td>
      <td>IGHV7-4-1*02</td>
      <td>IGHD3-10*01</td>
      <td>IGHJ5*02</td>
      <td>CAGGTGCAGCTGGTGCAATCTGGGTCT...GAGTTGAAGAAGCCTG...</td>
      <td>CAGGTGCAGCTGGTGCAATCTGGGTCT...GAGTTGAAGAAGCCTG...</td>
      <td>TGTGCGAGAGTTTTTAGACGCTATGGTTCGGGGAGTTATTATAACC...</td>
      <td>...</td>
      <td>INTNTGNP</td>
      <td>ARVFRRYGSGSYYNL</td>
      <td>QVQLVQSGSELKKPGASVKVSCKASGYTFTSYAMNWVRQAPGQGLE...</td>
      <td>QVQLVQSGSELKKPGASVKVSCKASGYTFTSYAMNWVRQAPGQGLE...</td>
      <td>YGSGSYY</td>
      <td>LWGQGTLVTVSS</td>
      <td>0.003003</td>
      <td>0</td>
      <td>159_2_1</td>
      <td>425_449</td>
    </tr>
    <tr>
      <th>sc5p_v2_hs_PBMC_10k_TTTGGTTTCGGTGTCG_contig_1</th>
      <td>sc5p_v2_hs_PBMC_10k_TTTGGTTTCGGTGTCG_contig_1</td>
      <td>GAGAGAGGAGCCTTAGCCCTGGATTCCAAGGCCTATCCACTTGGTG...</td>
      <td>F</td>
      <td>T</td>
      <td>IGHV3-21*01</td>
      <td>IGHD4-17*01</td>
      <td>IGHJ2*01</td>
      <td>GAGGTGCAGCTGGTGGAGTCTGGGGGA...GGCCTGGTCAAGCCTG...</td>
      <td>GAGGTGCAGCTGGTGGAGTCTGGGGGA...GGCCTGGTCAAGCCTG...</td>
      <td>TGTGCGAGAGATCCCGGTGACTACGTAGAAATTGAGTGGTACTTCG...</td>
      <td>...</td>
      <td>ISSSSSYI</td>
      <td>ARDPGDYVEIEWYFDL</td>
      <td>EVQLVESGGGLVKPGGSLRLSCAASGFTFSSYSMNWVRQAPGKGLE...</td>
      <td>EVQLVESGGGLVKPGGSLRLSCAASGFTFSSYSMNWVRQAPGKGLE...</td>
      <td>GDY</td>
      <td>WYFDLWGRGTLVTVSS</td>
      <td>0.000000</td>
      <td>0</td>
      <td>10_1_1</td>
      <td>357_450</td>
    </tr>
    <tr>
      <th>sc5p_v2_hs_PBMC_10k_TTTGGTTTCGGTGTCG_contig_2</th>
      <td>sc5p_v2_hs_PBMC_10k_TTTGGTTTCGGTGTCG_contig_2</td>
      <td>GGGAGAGCCCTGGGGAGGAACTGCTCAGTTAGGACCCAGAGGGAAC...</td>
      <td>F</td>
      <td>T</td>
      <td>IGKV3-11*01</td>
      <td></td>
      <td>IGKJ4*01</td>
      <td>GAAATTGTGTTGACACAGTCTCCAGCCACCCTGTCTTTGTCTCCAG...</td>
      <td>GAAATTGTGTTGACACAGTCTCCAGCCACCCTGTCTTTGTCTCCAG...</td>
      <td>TGTCAGCAGCGTAGCAACTGGCCTAGGCTCACTTTC</td>
      <td>...</td>
      <td>DAS</td>
      <td>QQRSNWPRLT</td>
      <td>EIVLTQSPATLSLSPGERATLSCRASQSVSSYLAWYQQKPGQAPRL...</td>
      <td>EIVLTQSPATLSLSPGERATLSCRASQSVSSYLAWYQQKPGQAPRL...</td>
      <td></td>
      <td>LTFGGGTKVEIK</td>
      <td>0.000000</td>
      <td>0</td>
      <td>10_1_1</td>
      <td>357_450</td>
    </tr>
  </tbody>
</table>
<p>926 rows × 84 columns</p>
</div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[27]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># create a new Dandelion class with this subset</span>
<span class="n">vdj2</span> <span class="o">=</span> <span class="n">ddl</span><span class="o">.</span><span class="n">Dandelion</span><span class="p">(</span><span class="n">subset_data</span><span class="p">)</span>
<span class="n">vdj2</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[27]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Dandelion class object with n_obs = 454 and n_contigs = 926
    data: &#39;sequence_id&#39;, &#39;sequence&#39;, &#39;rev_comp&#39;, &#39;productive&#39;, &#39;v_call&#39;, &#39;d_call&#39;, &#39;j_call&#39;, &#39;sequence_alignment&#39;, &#39;germline_alignment&#39;, &#39;junction&#39;, &#39;junction_aa&#39;, &#39;v_cigar&#39;, &#39;d_cigar&#39;, &#39;j_cigar&#39;, &#39;stop_codon&#39;, &#39;vj_in_frame&#39;, &#39;locus&#39;, &#39;junction_length&#39;, &#39;np1_length&#39;, &#39;np2_length&#39;, &#39;v_sequence_start&#39;, &#39;v_sequence_end&#39;, &#39;v_germline_start&#39;, &#39;v_germline_end&#39;, &#39;d_sequence_start&#39;, &#39;d_sequence_end&#39;, &#39;d_germline_start&#39;, &#39;d_germline_end&#39;, &#39;j_sequence_start&#39;, &#39;j_sequence_end&#39;, &#39;j_germline_start&#39;, &#39;j_germline_end&#39;, &#39;v_score&#39;, &#39;v_identity&#39;, &#39;v_support&#39;, &#39;d_score&#39;, &#39;d_identity&#39;, &#39;d_support&#39;, &#39;j_score&#39;, &#39;j_identity&#39;, &#39;j_support&#39;, &#39;fwr1&#39;, &#39;fwr2&#39;, &#39;fwr3&#39;, &#39;fwr4&#39;, &#39;cdr1&#39;, &#39;cdr2&#39;, &#39;cdr3&#39;, &#39;cell_id&#39;, &#39;c_call&#39;, &#39;consensus_count&#39;, &#39;umi_count&#39;, &#39;v_call_10x&#39;, &#39;d_call_10x&#39;, &#39;j_call_10x&#39;, &#39;junction_10x&#39;, &#39;junction_10x_aa&#39;, &#39;v_call_genotyped&#39;, &#39;germline_alignment_d_mask&#39;, &#39;sample_id&#39;, &#39;c_sequence_alignment&#39;, &#39;c_germline_alignment&#39;, &#39;c_sequence_start&#39;, &#39;c_sequence_end&#39;, &#39;c_score&#39;, &#39;c_identity&#39;, &#39;c_support&#39;, &#39;c_call_10x&#39;, &#39;junction_aa_length&#39;, &#39;fwr1_aa&#39;, &#39;fwr2_aa&#39;, &#39;fwr3_aa&#39;, &#39;fwr4_aa&#39;, &#39;cdr1_aa&#39;, &#39;cdr2_aa&#39;, &#39;cdr3_aa&#39;, &#39;sequence_alignment_aa&#39;, &#39;v_sequence_alignment_aa&#39;, &#39;d_sequence_alignment_aa&#39;, &#39;j_sequence_alignment_aa&#39;, &#39;mu_freq&#39;, &#39;duplicate_count&#39;, &#39;clone_id&#39;, &#39;changeo_clone_id&#39;
    metadata: &#39;sample_id&#39;, &#39;clone_id&#39;, &#39;clone_id_by_size&#39;, &#39;isotype&#39;, &#39;lightchain&#39;, &#39;status&#39;, &#39;vdj_status&#39;, &#39;productive&#39;, &#39;umi_counts_heavy&#39;, &#39;umi_counts_light&#39;, &#39;v_call_heavy&#39;, &#39;v_call_light&#39;, &#39;j_call_heavy&#39;, &#39;j_call_light&#39;, &#39;c_call_heavy&#39;, &#39;c_call_light&#39;
    distance: None
    edges: None
    layout: None
    graph: None
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[28]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># update the germline using the corrected files after tigger</span>
<span class="n">vdj2</span><span class="o">.</span><span class="n">update_germline</span><span class="p">(</span><span class="n">corrected</span> <span class="o">=</span> <span class="s1">&#39;tutorial_scgp1/tutorial_scgp1_heavy_igblast_db-pass_genotype.fasta&#39;</span><span class="p">,</span> <span class="n">germline</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">org</span> <span class="o">=</span> <span class="s1">&#39;human&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Updating germline reference
 finished: Updated Dandelion object:
   &#39;germline&#39;, updated germline reference
 (0:00:00)
</pre></div></div>
</div>
<p><strong>pp.create_germlines</strong></p>
<p>Then we run <code class="docutils literal notranslate"><span class="pre">pp.create_germline</span></code> to (re)create the <code class="docutils literal notranslate"><span class="pre">germline_alignment_d_mask</span></code> column in the data. If <code class="docutils literal notranslate"><span class="pre">update_germline</span></code> was run like above, there’s no need to specify the <code class="docutils literal notranslate"><span class="pre">germline</span></code> option as the function will simply retrieve it from the <code class="docutils literal notranslate"><span class="pre">Dandelion</span></code> object.</p>
<p><em>Note: the ability to run the original CreateGermlines.py with –cloned option is not currently possible through pp.create_germlines(). This is possible with pp.external.creategermlines but requires a physical file for CreateGermlines.py to work on. Thus, I would reccomend for you to run CreateGermlines.py separately if you intend to use the –cloned option. See</em><a class="reference external" href="https://changeo.readthedocs.io/en/stable/examples/germlines.html">here</a><em>for more info.</em></p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[29]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">ddl</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">create_germlines</span><span class="p">(</span><span class="n">vdj2</span><span class="p">,</span> <span class="n">v_field</span> <span class="o">=</span> <span class="s1">&#39;v_call_genotyped&#39;</span><span class="p">,</span> <span class="n">germ_types</span><span class="o">=</span><span class="s1">&#39;dmask&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Reconstructing germline sequences
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
   Building dmask germline sequences: 926it [00:00, 967.80it/s]
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
 finished: Updated Dandelion object:
   &#39;data&#39;, updated germline alignment in contig-indexed clone table
   &#39;germline&#39;, updated germline reference
 (0:00:01)
</pre></div></div>
</div>
<p>Ensure that the germline_alignment_d_mask column is populated or subsequent steps will fail.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[30]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">vdj2</span><span class="o">.</span><span class="n">data</span><span class="p">[[</span><span class="s1">&#39;v_call_genotyped&#39;</span><span class="p">,</span> <span class="s1">&#39;germline_alignment_d_mask&#39;</span><span class="p">]]</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[30]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>v_call_genotyped</th>
      <th>germline_alignment_d_mask</th>
    </tr>
    <tr>
      <th>sequence_id</th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>sc5p_v2_hs_PBMC_1k_AACTGGTTCTCTAAGG_contig_1</th>
      <td>IGLV3-1*01</td>
      <td>TCCTATGAGCTGACTCAGCCACCCTCA...GTGTCCGTGTCCCCAG...</td>
    </tr>
    <tr>
      <th>sc5p_v2_hs_PBMC_1k_AACTGGTTCTCTAAGG_contig_2</th>
      <td>IGHV3-15*01</td>
      <td>GAGGTGCAGCTGGTGGAGTCTGGGGGA...GGCTTGGTAAAGCCTG...</td>
    </tr>
    <tr>
      <th>sc5p_v2_hs_PBMC_1k_AATCCAGTCAGTTGAC_contig_1</th>
      <td>IGKV3-20*01</td>
      <td>GAAATTGTGTTGACGCAGTCTCCAGGCACCCTGTCTTTGTCTCCAG...</td>
    </tr>
    <tr>
      <th>sc5p_v2_hs_PBMC_1k_AATCCAGTCAGTTGAC_contig_2</th>
      <td>IGHV3-30*02</td>
      <td>CAGGTGCAGCTGGTGGAGTCTGGGGGA...GGCGTGGTCCAGCCTG...</td>
    </tr>
    <tr>
      <th>sc5p_v2_hs_PBMC_1k_AATCGGTGTTAGGGTG_contig_1</th>
      <td>IGKV4-1*01</td>
      <td>GACATCGTGATGACCCAGTCTCCAGACTCCCTGGCTGTGTCTCTGG...</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>sc5p_v2_hs_PBMC_10k_TTTCCTCTCTTCAACT_contig_1</th>
      <td>IGHV1-8*01</td>
      <td>CAGGTGCAGCTGGTGCAGTCTGGGGCT...GAGGTGAAGAAGCCTG...</td>
    </tr>
    <tr>
      <th>sc5p_v2_hs_PBMC_10k_TTTGGTTTCAGAGCTT_contig_2</th>
      <td>IGKV3-11*01</td>
      <td>GAAATTGTGTTGACACAGTCTCCAGCCACCCTGTCTTTGTCTCCAG...</td>
    </tr>
    <tr>
      <th>sc5p_v2_hs_PBMC_10k_TTTGGTTTCAGAGCTT_contig_1</th>
      <td>IGHV7-4-1*02</td>
      <td>CAGGTGCAGCTGGTGCAATCTGGGTCT...GAGTTGAAGAAGCCTG...</td>
    </tr>
    <tr>
      <th>sc5p_v2_hs_PBMC_10k_TTTGGTTTCGGTGTCG_contig_1</th>
      <td>IGHV3-21*01</td>
      <td>GAGGTGCAGCTGGTGGAGTCTGGGGGA...GGCCTGGTCAAGCCTG...</td>
    </tr>
    <tr>
      <th>sc5p_v2_hs_PBMC_10k_TTTGGTTTCGGTGTCG_contig_2</th>
      <td>IGKV3-11*01</td>
      <td>GAAATTGTGTTGACACAGTCTCCAGCCACCCTGTCTTTGTCTCCAG...</td>
    </tr>
  </tbody>
</table>
<p>926 rows × 2 columns</p>
</div></div>
</div>
<p>The default behaviour is to mask the D region with Ns with option <code class="docutils literal notranslate"><span class="pre">germ_types</span> <span class="pre">=</span> <span class="pre">'dmask'</span></code>. See <a class="reference external" href="https://changeo.readthedocs.io/en/stable/methods/germlines.html">here</a> for more info.</p>
<p><strong>``pp.quantify_mutations``</strong></p>
<p>The options for <code class="docutils literal notranslate"><span class="pre">pp.quantify_mutations</span></code> are the same as the basic mutational load analysis <a class="reference external" href="https://shazam.readthedocs.io/en/version-0.1.8---mutation-profiling-enhancements/vignettes/Mutation-Vignette/">vignette</a>. The default behavior is to sum all mutations scores (heavy and light chains, silent and replacement mutations) for the same cell.</p>
<p>Again, this function can be run immediately after <code class="docutils literal notranslate"><span class="pre">pp.reassign_alleles</span></code> on the genotyped <code class="docutils literal notranslate"><span class="pre">.tsv</span></code> files (without loading into <code class="docutils literal notranslate"><span class="pre">pandas</span></code> or <code class="docutils literal notranslate"><span class="pre">Dandelion</span></code>). Here I’m illustrating a few other options that may be useful.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[31]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># switching back to using the full vdj object</span>
<span class="n">ddl</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">quantify_mutations</span><span class="p">(</span><span class="n">vdj</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Quantifying mutations
 finished: Updated Dandelion object:
   &#39;data&#39;, contig-indexed clone table
   &#39;metadata&#39;, cell-indexed clone table
 (0:00:05)
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[32]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">ddl</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">quantify_mutations</span><span class="p">(</span><span class="n">vdj</span><span class="p">,</span> <span class="n">combine</span> <span class="o">=</span> <span class="kc">False</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Quantifying mutations
 finished: Updated Dandelion object:
   &#39;data&#39;, contig-indexed clone table
   &#39;metadata&#39;, cell-indexed clone table
 (0:00:04)
</pre></div></div>
</div>
<p>Specifying <code class="docutils literal notranslate"><span class="pre">split_locus</span> <span class="pre">=</span> <span class="pre">True</span></code> will split up the results for the different chains.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[33]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">ddl</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">quantify_mutations</span><span class="p">(</span><span class="n">vdj</span><span class="p">,</span> <span class="n">split_locus</span> <span class="o">=</span> <span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Quantifying mutations
 finished: Updated Dandelion object:
   &#39;data&#39;, contig-indexed clone table
   &#39;metadata&#39;, cell-indexed clone table
 (0:00:06)
</pre></div></div>
</div>
<p>To update the <code class="docutils literal notranslate"><span class="pre">AnnData</span></code> object, simply rerun <code class="docutils literal notranslate"><span class="pre">tl.transfer_network</span></code> with keep_raw option set to <code class="docutils literal notranslate"><span class="pre">False</span></code> to avoid overwriting the stashed neighborhood graph.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[34]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">ddl</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">transfer</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">vdj</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Transferring network
converting matrices
Updating anndata slots
 finished: updated `.obs` with `.metadata`
added to `.uns[&#39;neighbors&#39;]` and `.obsp`
   &#39;distances&#39;, cluster-weighted adjacency matrix
   &#39;connectivities&#39;, cluster-weighted adjacency matrix (0:00:18)
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[35]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">sc</span><span class="o">.</span><span class="n">set_figure_params</span><span class="p">(</span><span class="n">figsize</span> <span class="o">=</span> <span class="p">[</span><span class="mi">4</span><span class="p">,</span><span class="mi">4</span><span class="p">])</span>
<span class="n">ddl</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">clone_network</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">color</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;clone_id&#39;</span><span class="p">,</span> <span class="s1">&#39;mu_freq&#39;</span><span class="p">,</span> <span class="s1">&#39;mu_freq_seq_r&#39;</span><span class="p">,</span> <span class="s1">&#39;mu_freq_seq_s&#39;</span><span class="p">,</span> <span class="s1">&#39;mu_freq_IGH&#39;</span><span class="p">,</span> <span class="s1">&#39;mu_freq_IGL&#39;</span><span class="p">],</span> <span class="n">ncols</span> <span class="o">=</span> <span class="mi">2</span><span class="p">,</span> <span class="n">legend_loc</span> <span class="o">=</span> <span class="s1">&#39;none&#39;</span><span class="p">,</span> <span class="n">legend_fontoutline</span><span class="o">=</span><span class="mi">3</span><span class="p">,</span> <span class="n">edges_width</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="n">palette</span> <span class="o">=</span> <span class="n">default_28</span> <span class="o">+</span> <span class="n">default_102</span><span class="p">,</span> <span class="n">color_map</span> <span class="o">=</span> <span class="s1">&#39;viridis&#39;</span><span class="p">,</span> <span class="n">size</span> <span class="o">=</span> <span class="mi">50</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
WARNING: Length of palette colors is smaller than the number of categories (palette length: 130, categories length: 822. Some categories will have the same color.
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_3_dandelion_findingclones_and_analysis-10x_data_72_1.png" class="no-scaled-link" src="../_images/notebooks_3_dandelion_findingclones_and_analysis-10x_data_72_1.png" style="width: 681px; height: 856px;" />
</div>
</div>
</div>
<div class="section" id="Calculating-size-of-clones">
<h2>Calculating size of clones<a class="headerlink" href="#Calculating-size-of-clones" title="Permalink to this headline">¶</a></h2>
<p><strong>``tl.clone_size``</strong></p>
<p>Sometimes it’s useful to evaluate the size of the clone. Here <code class="docutils literal notranslate"><span class="pre">tl.quantify_clone_size</span></code> does a simple calculation to enable that.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[36]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">ddl</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">clone_size</span><span class="p">(</span><span class="n">vdj</span><span class="p">)</span>
<span class="n">ddl</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">transfer</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">vdj</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Quantifying clone sizes
 finished: Updated Dandelion object:
   &#39;metadata&#39;, cell-indexed clone table (0:00:00)
Transferring network
converting matrices
Updating anndata slots
 finished: updated `.obs` with `.metadata`
added to `.uns[&#39;neighbors&#39;]` and `.obsp`
   &#39;distances&#39;, cluster-weighted adjacency matrix
   &#39;connectivities&#39;, cluster-weighted adjacency matrix (0:00:18)
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[37]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">ddl</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">clone_network</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">color</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;clone_id_size&#39;</span><span class="p">],</span> <span class="n">legend_loc</span> <span class="o">=</span> <span class="s1">&#39;none&#39;</span><span class="p">,</span> <span class="n">legend_fontoutline</span><span class="o">=</span><span class="mi">3</span><span class="p">,</span> <span class="n">edges_width</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="n">size</span> <span class="o">=</span> <span class="mi">50</span><span class="p">)</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">color</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;clone_id_size&#39;</span><span class="p">])</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_3_dandelion_findingclones_and_analysis-10x_data_75_0.png" class="no-scaled-link" src="../_images/notebooks_3_dandelion_findingclones_and_analysis-10x_data_75_0.png" style="width: 297px; height: 296px;" />
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_3_dandelion_findingclones_and_analysis-10x_data_75_1.png" class="no-scaled-link" src="../_images/notebooks_3_dandelion_findingclones_and_analysis-10x_data_75_1.png" style="width: 297px; height: 296px;" />
</div>
</div>
<p>You can also specify <code class="docutils literal notranslate"><span class="pre">max_size</span></code> to clip off the calculation at a fixed value.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[38]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">ddl</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">clone_size</span><span class="p">(</span><span class="n">vdj</span><span class="p">,</span> <span class="n">max_size</span> <span class="o">=</span> <span class="mi">3</span><span class="p">)</span>
<span class="n">ddl</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">transfer</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">vdj</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Quantifying clone sizes
 finished: Updated Dandelion object:
   &#39;metadata&#39;, cell-indexed clone table (0:00:00)
Transferring network
converting matrices
Updating anndata slots
 finished: updated `.obs` with `.metadata`
added to `.uns[&#39;neighbors&#39;]` and `.obsp`
   &#39;distances&#39;, cluster-weighted adjacency matrix
   &#39;connectivities&#39;, cluster-weighted adjacency matrix (0:00:19)
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[39]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">ddl</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">clone_network</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">color</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;clone_id_size&#39;</span><span class="p">],</span> <span class="n">ncols</span> <span class="o">=</span> <span class="mi">2</span><span class="p">,</span> <span class="n">legend_fontoutline</span><span class="o">=</span><span class="mi">3</span><span class="p">,</span> <span class="n">edges_width</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="n">palette</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;grey&#39;</span><span class="p">,</span> <span class="s1">&#39;red&#39;</span><span class="p">,</span> <span class="s1">&#39;blue&#39;</span><span class="p">,</span> <span class="s1">&#39;white&#39;</span><span class="p">],</span> <span class="n">size</span> <span class="o">=</span> <span class="mi">50</span><span class="p">)</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="n">adata</span><span class="p">[</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;has_bcr&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="s1">&#39;True&#39;</span><span class="p">],</span> <span class="n">color</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;clone_id_size&#39;</span><span class="p">],</span> <span class="n">palette</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;grey&#39;</span><span class="p">,</span> <span class="s1">&#39;red&#39;</span><span class="p">,</span> <span class="s1">&#39;blue&#39;</span><span class="p">,</span> <span class="s1">&#39;white&#39;</span><span class="p">])</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_3_dandelion_findingclones_and_analysis-10x_data_78_0.png" class="no-scaled-link" src="../_images/notebooks_3_dandelion_findingclones_and_analysis-10x_data_78_0.png" style="width: 297px; height: 296px;" />
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_3_dandelion_findingclones_and_analysis-10x_data_78_1.png" class="no-scaled-link" src="../_images/notebooks_3_dandelion_findingclones_and_analysis-10x_data_78_1.png" style="width: 297px; height: 296px;" />
</div>
</div>
</div>
<div class="section" id="Calculating-diversity">
<h2>Calculating diversity<a class="headerlink" href="#Calculating-diversity" title="Permalink to this headline">¶</a></h2>
<p><em>disclaimer: the functions here are not complete yet. Please look to other sources/methods for doing this properly. Also, would appreciate any help to help me finalise this!</em></p>
<p><strong>``tl.clone_rarefaction/pl.clone_rarefaction``</strong></p>
<p>We can use <code class="docutils literal notranslate"><span class="pre">pl.clone_rarefaction</span></code> to generate rarefaction curves for the clones. <code class="docutils literal notranslate"><span class="pre">tl.clone_rarefaction</span></code> will populate the <code class="docutils literal notranslate"><span class="pre">.uns</span></code> slot with the results. <code class="docutils literal notranslate"><span class="pre">groupby</span></code> option must be specified. In this case, I decided to group by sample. The function will only work on an <code class="docutils literal notranslate"><span class="pre">AnnData</span></code> object and not a <code class="docutils literal notranslate"><span class="pre">Dandelion</span></code> object.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[40]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">ddl</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">clone_rarefaction</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">groupby</span> <span class="o">=</span> <span class="s1">&#39;sampleid&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
removing due to zero counts:
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
Calculating rarefaction curve : 100%|██████████| 4/4 [00:00&lt;00:00, 14.54it/s]
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_3_dandelion_findingclones_and_analysis-10x_data_80_2.png" class="no-scaled-link" src="../_images/notebooks_3_dandelion_findingclones_and_analysis-10x_data_80_2.png" style="width: 753px; height: 393px;" />
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[40]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
&lt;ggplot: (349039913)&gt;
</pre></div></div>
</div>
<p><strong>``tl.clone_diversity``</strong></p>
<p><code class="docutils literal notranslate"><span class="pre">tl.clone_diversity</span></code> allows for calculation of diversity measures such as <strong>Chao1</strong>, <strong>Shannon Entropy</strong> and <strong>Gini indices</strong>.</p>
<p>While the function can work on both <code class="docutils literal notranslate"><span class="pre">AnnData</span></code> and <code class="docutils literal notranslate"><span class="pre">Dandelion</span></code> objects, the methods for gini index calculation will only work on a <code class="docutils literal notranslate"><span class="pre">Dandelion</span></code> object as it requires access to the network.</p>
<p>For Gini indices, we provide several types of measures, inspired by bulk BCRseq analysis methods from Rachael Bashford-Rogers’ <a class="reference external" href="https://genome.cshlp.org/content/23/11/1874.full">works</a>:</p>
<p><strong>Default</strong></p>
<p><strong>i) network cluster/clone size Gini index - ``metric = clone_network``</strong></p>
<p>In a contracted BCR network (where identical BCRs are collapsed into the same node/vertex), disparity in the distribution should be correlated to the amount of mutation events i.e. larger networks should indicate more mutation events and smaller networks should indicate lesser mutation events.</p>
<p><strong>ii) network vertex/node size Gini index - ``metric = clone_network``</strong></p>
<p>In the same contracted network, we can count the number of merged/contracted nodes; nodes with higher count numbers indicate more clonal expansion. Thus, disparity in the distribution of count numbers (referred to as vertex size) should be correlated to the overall clonality i.e. clones with larger vertex sizes are more monoclonal and clones with smaller vertex sizes are more polyclonal.</p>
<p>Therefore, a Gini index of 1 on either measures repesents perfect inequality (i.e. monoclonal and highly mutated) and a value of 0 represents perfect equality (i.e. polyclonal and unmutated).</p>
<p>However, there are a few limitations/challenges that comes with single-cell data:</p>
<ol class="upperalpha simple">
<li><p>In the process of contracting the network, we discard the single-cell level information.</p></li>
<li><p>Contraction of network is very slow, particularly when there is a lot of clonally-related cells.</p></li>
<li><p>For the full implementation and interpretation of both measures, although more evident with cluster/clone size, it requires the BCR repertoire to be reasonably/deeply sampled and we know that this is currently limited by the low recovery from single cell data with current technologies.</p></li>
</ol>
<p>Therefore, we implement a few work arounds, and ‘experimental’ options below, to try and circumvent these issues.</p>
<p>Firstly, as a work around for (C), the cluster size gini index can be calculated before or after network contraction. If performing before network contraction (default), it will be calculated based on the size of subgraphs of connected components in the main graph. This will retain the single-cell information and should appropriately show the distribution of the data. If performing after network contraction, the calculation is performed after network contraction, achieving the same effect as the
method for bulk BCR-seq as described above. This option can be toggled by <code class="docutils literal notranslate"><span class="pre">use_contracted</span></code> and only applies to network cluster size gini index calculation.</p>
<p><strong>Alternative</strong></p>
<p><strong>iii) clone centrality Gini index - ``metric = clone_centrality``</strong></p>
<p>Node/vertex closeness centrality indicates how tightly packed clones are (more clonally related) and thus the distribution of the number of cells connected in each clone informs on whether clones in general are more monoclonal or polyclonal.</p>
<p><strong>iv) clone degree Gini index - ``metric = clone_degree``</strong></p>
<p>Node/vertex degree indicates how many cells are connected to an individual cell, another indication of how clonally related cells are. However, this would also highlight cells that are in the middle of large networks but are not necessarily within clonally expanded regions (e.g. intermediate connecting cells within the minimum spanning tree)</p>
<p><strong>v) clone size Gini index - ``metric = clone_size``</strong></p>
<p>This is not to be confused with the network cluster size gini index calculation above as this doesn’t rely on the network, although the values should be similar. This is just a simple implementation based on the data frame for the relevant <code class="docutils literal notranslate"><span class="pre">clone_id</span></code> column. By default, this metric is also returned when running <code class="docutils literal notranslate"><span class="pre">metric</span> <span class="pre">=</span> <span class="pre">clone_centrality</span></code> or <code class="docutils literal notranslate"><span class="pre">metric</span> <span class="pre">=</span> <span class="pre">clone_degree</span></code>.</p>
<p>For (i) and (ii), we can specify <code class="docutils literal notranslate"><span class="pre">expanded_only</span></code> option to compute the statistic for all clones or expanded only clones. Unlike options (i) and (ii), the current calculation for (iii) and (iv) is largely influenced by the amount of expanded clones i.e. clones with at least 2 cells, and not affected by the number of singleton clones because singleton clones will have a value of 0 regardless.</p>
<p>The diversity functions also have the option to perform downsampling to a fixed number of cells, or to the smallest sample size specified via <code class="docutils literal notranslate"><span class="pre">groupby</span></code> (default) so that sample sizes are even when comparing between groups.</p>
<p>if <code class="docutils literal notranslate"><span class="pre">update_obs_meta=False</span></code>, a data frame is returned; otherwise, the value gets added to the <code class="docutils literal notranslate"><span class="pre">AnnData.obs</span></code> or <code class="docutils literal notranslate"><span class="pre">Dandelion.metadata</span></code> accordingly.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[41]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">ddl</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">clone_diversity</span><span class="p">(</span><span class="n">vdj</span><span class="p">,</span> <span class="n">groupby</span> <span class="o">=</span> <span class="s1">&#39;sample_id&#39;</span><span class="p">,</span> <span class="n">method</span> <span class="o">=</span> <span class="s1">&#39;gini&#39;</span><span class="p">,</span> <span class="n">metric</span> <span class="o">=</span> <span class="s1">&#39;clone_network&#39;</span><span class="p">)</span>
<span class="n">ddl</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">clone_diversity</span><span class="p">(</span><span class="n">vdj</span><span class="p">,</span> <span class="n">groupby</span> <span class="o">=</span> <span class="s1">&#39;sample_id&#39;</span><span class="p">,</span> <span class="n">method</span> <span class="o">=</span> <span class="s1">&#39;gini&#39;</span><span class="p">,</span> <span class="n">metric</span> <span class="o">=</span> <span class="s1">&#39;clone_centrality&#39;</span><span class="p">)</span>
<span class="n">ddl</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">transfer</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">vdj</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Calculating Gini indices
Computing Gini indices for cluster and vertex size using network.
 finished: updated `.metadata` with Gini indices.
 (0:00:12)
Calculating Gini indices
Computing gini indices for clone size using metadata and node closeness centrality using network.
Calculating node closeness centrality
 finished: Updated Dandelion metadata
 (0:00:00)
 finished: updated `.metadata` with Gini indices.
 (0:00:01)
Transferring network
converting matrices
Updating anndata slots
 finished: updated `.obs` with `.metadata`
added to `.uns[&#39;neighbors&#39;]` and `.obsp`
   &#39;distances&#39;, cluster-weighted adjacency matrix
   &#39;connectivities&#39;, cluster-weighted adjacency matrix (0:00:20)
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[42]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">ddl</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">clone_network</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">color</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;clone_network_cluster_size_gini&#39;</span><span class="p">,</span> <span class="s1">&#39;clone_network_vertex_size_gini&#39;</span><span class="p">,</span> <span class="s1">&#39;clone_size_gini&#39;</span><span class="p">,</span> <span class="s1">&#39;clone_centrality_gini&#39;</span><span class="p">],</span> <span class="n">ncols</span> <span class="o">=</span> <span class="mi">2</span><span class="p">,</span> <span class="n">size</span> <span class="o">=</span> <span class="mi">50</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_3_dandelion_findingclones_and_analysis-10x_data_84_0.png" class="no-scaled-link" src="../_images/notebooks_3_dandelion_findingclones_and_analysis-10x_data_84_0.png" style="width: 681px; height: 577px;" />
</div>
</div>
<p>With these particular samples, because there is not many expanded clones in general, the gini indices are quite low when calculated within each sample. We can re-run it by specifying <code class="docutils literal notranslate"><span class="pre">expanded_only</span> <span class="pre">=</span> <span class="pre">True</span></code> to only factor in expanded_clones. We also specify the key_added option to create a new column instead of writing over the original columns.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[43]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">ddl</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">clone_diversity</span><span class="p">(</span><span class="n">vdj</span><span class="p">,</span> <span class="n">groupby</span> <span class="o">=</span> <span class="s1">&#39;sample_id&#39;</span><span class="p">,</span> <span class="n">method</span> <span class="o">=</span> <span class="s1">&#39;gini&#39;</span><span class="p">,</span> <span class="n">metric</span> <span class="o">=</span> <span class="s1">&#39;clone_network&#39;</span><span class="p">,</span> <span class="n">expanded_only</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span> <span class="n">key_added</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;clone_network_cluster_size_gini_expanded&#39;</span><span class="p">,</span> <span class="s1">&#39;clone_network_vertex_size_gini_expanded&#39;</span><span class="p">])</span>
<span class="n">ddl</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">transfer</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">vdj</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Calculating Gini indices
Computing Gini indices for cluster and vertex size using network.
 finished: updated `.metadata` with Gini indices.
 (0:00:12)
Transferring network
converting matrices
Updating anndata slots
 finished: updated `.obs` with `.metadata`
added to `.uns[&#39;neighbors&#39;]` and `.obsp`
   &#39;distances&#39;, cluster-weighted adjacency matrix
   &#39;connectivities&#39;, cluster-weighted adjacency matrix (0:00:19)
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[44]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">ddl</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">clone_network</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">color</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;clone_network_cluster_size_gini_expanded&#39;</span><span class="p">,</span> <span class="s1">&#39;clone_network_vertex_size_gini_expanded&#39;</span><span class="p">],</span> <span class="n">ncols</span> <span class="o">=</span> <span class="mi">2</span><span class="p">,</span> <span class="n">size</span> <span class="o">=</span> <span class="mi">50</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_3_dandelion_findingclones_and_analysis-10x_data_87_0.png" class="no-scaled-link" src="../_images/notebooks_3_dandelion_findingclones_and_analysis-10x_data_87_0.png" style="width: 678px; height: 299px;" />
</div>
</div>
<p>We can also choose not to update the metadata to return a pandas dataframe.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[45]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">gini</span> <span class="o">=</span> <span class="n">ddl</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">clone_diversity</span><span class="p">(</span><span class="n">vdj</span><span class="p">,</span> <span class="n">groupby</span> <span class="o">=</span> <span class="s1">&#39;sample_id&#39;</span><span class="p">,</span> <span class="n">method</span> <span class="o">=</span> <span class="s1">&#39;gini&#39;</span><span class="p">,</span> <span class="n">update_obs_meta</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<span class="n">gini</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Calculating Gini indices
Computing Gini indices for cluster and vertex size using network.
 finished (0:00:12)
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[45]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>clone_network_cluster_size_gini</th>
      <th>clone_network_vertex_size_gini</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>vdj_v1_hs_pbmc3</th>
      <td>0.026316</td>
      <td>0.000000</td>
    </tr>
    <tr>
      <th>sc5p_v2_hs_PBMC_1k</th>
      <td>0.029412</td>
      <td>0.000000</td>
    </tr>
    <tr>
      <th>sc5p_v2_hs_PBMC_10k</th>
      <td>0.004739</td>
      <td>0.001584</td>
    </tr>
    <tr>
      <th>vdj_nextgem_hs_pbmc3</th>
      <td>0.048583</td>
      <td>0.021134</td>
    </tr>
  </tbody>
</table>
</div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[46]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">gini2</span> <span class="o">=</span> <span class="n">ddl</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">clone_diversity</span><span class="p">(</span><span class="n">vdj</span><span class="p">,</span> <span class="n">groupby</span> <span class="o">=</span> <span class="s1">&#39;sample_id&#39;</span><span class="p">,</span> <span class="n">method</span> <span class="o">=</span> <span class="s1">&#39;gini&#39;</span><span class="p">,</span> <span class="n">update_obs_meta</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">expanded_only</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span> <span class="n">key_added</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;clone_network_cluster_size_gini_expanded&#39;</span><span class="p">,</span> <span class="s1">&#39;clone_network_vertex_size_gini_expanded&#39;</span><span class="p">])</span>
<span class="n">gini2</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Calculating Gini indices
Computing Gini indices for cluster and vertex size using network.
 finished (0:00:12)
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[46]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>clone_network_cluster_size_gini_expanded</th>
      <th>clone_network_vertex_size_gini_expanded</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>vdj_v1_hs_pbmc3</th>
      <td>0.026316</td>
      <td>0.000000</td>
    </tr>
    <tr>
      <th>sc5p_v2_hs_PBMC_1k</th>
      <td>0.029412</td>
      <td>0.000000</td>
    </tr>
    <tr>
      <th>sc5p_v2_hs_PBMC_10k</th>
      <td>0.000000</td>
      <td>0.333333</td>
    </tr>
    <tr>
      <th>vdj_nextgem_hs_pbmc3</th>
      <td>0.467532</td>
      <td>0.333333</td>
    </tr>
  </tbody>
</table>
</div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[47]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="kn">import</span> <span class="nn">seaborn</span> <span class="k">as</span> <span class="nn">sns</span>
<span class="n">p</span> <span class="o">=</span> <span class="n">sns</span><span class="o">.</span><span class="n">scatterplot</span><span class="p">(</span><span class="n">x</span> <span class="o">=</span> <span class="s1">&#39;clone_network_cluster_size_gini&#39;</span><span class="p">,</span> <span class="n">y</span> <span class="o">=</span> <span class="s1">&#39;clone_network_vertex_size_gini&#39;</span><span class="p">,</span> <span class="n">data</span> <span class="o">=</span> <span class="n">gini</span><span class="p">,</span> <span class="n">hue</span> <span class="o">=</span> <span class="n">gini</span><span class="o">.</span><span class="n">index</span><span class="p">,</span> <span class="n">palette</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">(</span><span class="nb">zip</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;sampleid&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">cat</span><span class="o">.</span><span class="n">categories</span><span class="p">,</span> <span class="n">adata</span><span class="o">.</span><span class="n">uns</span><span class="p">[</span><span class="s1">&#39;sampleid_colors&#39;</span><span class="p">])))</span>
<span class="n">p</span><span class="o">.</span><span class="n">set</span><span class="p">(</span><span class="n">ylim</span><span class="o">=</span><span class="p">(</span><span class="o">-</span><span class="mf">0.1</span><span class="p">,</span><span class="mi">1</span><span class="p">),</span> <span class="n">xlim</span> <span class="o">=</span> <span class="p">(</span><span class="o">-</span><span class="mf">0.1</span><span class="p">,</span><span class="mi">1</span><span class="p">))</span>
<span class="n">p</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[47]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
&lt;AxesSubplot:xlabel=&#39;clone_network_cluster_size_gini&#39;, ylabel=&#39;clone_network_vertex_size_gini&#39;&gt;
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_3_dandelion_findingclones_and_analysis-10x_data_91_1.png" class="no-scaled-link" src="../_images/notebooks_3_dandelion_findingclones_and_analysis-10x_data_91_1.png" style="width: 329px; height: 306px;" />
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[48]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">p2</span> <span class="o">=</span> <span class="n">sns</span><span class="o">.</span><span class="n">scatterplot</span><span class="p">(</span><span class="n">x</span> <span class="o">=</span> <span class="s1">&#39;clone_network_cluster_size_gini_expanded&#39;</span><span class="p">,</span> <span class="n">y</span> <span class="o">=</span> <span class="s1">&#39;clone_network_vertex_size_gini_expanded&#39;</span><span class="p">,</span> <span class="n">data</span> <span class="o">=</span> <span class="n">gini2</span><span class="p">,</span> <span class="n">hue</span> <span class="o">=</span> <span class="n">gini2</span><span class="o">.</span><span class="n">index</span><span class="p">,</span> <span class="n">palette</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">(</span><span class="nb">zip</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;sampleid&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">cat</span><span class="o">.</span><span class="n">categories</span><span class="p">,</span> <span class="n">adata</span><span class="o">.</span><span class="n">uns</span><span class="p">[</span><span class="s1">&#39;sampleid_colors&#39;</span><span class="p">])))</span>
<span class="n">p2</span><span class="o">.</span><span class="n">set</span><span class="p">(</span><span class="n">ylim</span><span class="o">=</span><span class="p">(</span><span class="o">-</span><span class="mf">0.1</span><span class="p">,</span><span class="mi">1</span><span class="p">),</span> <span class="n">xlim</span> <span class="o">=</span> <span class="p">(</span><span class="o">-</span><span class="mf">0.1</span><span class="p">,</span><span class="mi">1</span><span class="p">))</span>
<span class="n">p2</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[48]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
&lt;AxesSubplot:xlabel=&#39;clone_network_cluster_size_gini_expanded&#39;, ylabel=&#39;clone_network_vertex_size_gini_expanded&#39;&gt;
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_3_dandelion_findingclones_and_analysis-10x_data_92_1.png" class="no-scaled-link" src="../_images/notebooks_3_dandelion_findingclones_and_analysis-10x_data_92_1.png" style="width: 339px; height: 327px;" />
</div>
</div>
<p>We can also visualise what the results for the clone centrality gini indices.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[49]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">gini</span> <span class="o">=</span> <span class="n">ddl</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">clone_diversity</span><span class="p">(</span><span class="n">vdj</span><span class="p">,</span> <span class="n">groupby</span> <span class="o">=</span> <span class="s1">&#39;sample_id&#39;</span><span class="p">,</span> <span class="n">method</span> <span class="o">=</span> <span class="s1">&#39;gini&#39;</span><span class="p">,</span> <span class="n">metric</span> <span class="o">=</span> <span class="s1">&#39;clone_centrality&#39;</span><span class="p">,</span> <span class="n">update_obs_meta</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<span class="n">gini</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Calculating Gini indices
Computing gini indices for clone size using metadata and node closeness centrality using network.
Calculating node closeness centrality
 finished: Updated Dandelion metadata
 (0:00:00)
 finished (0:00:01)
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[49]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>clone_size_gini</th>
      <th>clone_centrality_gini</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>vdj_v1_hs_pbmc3</th>
      <td>0.026316</td>
      <td>0.000000</td>
    </tr>
    <tr>
      <th>sc5p_v2_hs_PBMC_1k</th>
      <td>0.029412</td>
      <td>0.000000</td>
    </tr>
    <tr>
      <th>sc5p_v2_hs_PBMC_10k</th>
      <td>0.004739</td>
      <td>0.000000</td>
    </tr>
    <tr>
      <th>vdj_nextgem_hs_pbmc3</th>
      <td>0.048583</td>
      <td>0.045455</td>
    </tr>
  </tbody>
</table>
</div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[50]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># not a great example because there&#39;s only 1 big clone in 1 sample.</span>
<span class="n">p</span> <span class="o">=</span> <span class="n">sns</span><span class="o">.</span><span class="n">scatterplot</span><span class="p">(</span><span class="n">x</span> <span class="o">=</span> <span class="s1">&#39;clone_size_gini&#39;</span><span class="p">,</span> <span class="n">y</span> <span class="o">=</span> <span class="s1">&#39;clone_centrality_gini&#39;</span><span class="p">,</span> <span class="n">data</span> <span class="o">=</span> <span class="n">gini</span><span class="p">,</span> <span class="n">hue</span> <span class="o">=</span> <span class="n">gini</span><span class="o">.</span><span class="n">index</span><span class="p">,</span> <span class="n">palette</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">(</span><span class="nb">zip</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;sampleid&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">cat</span><span class="o">.</span><span class="n">categories</span><span class="p">,</span> <span class="n">adata</span><span class="o">.</span><span class="n">uns</span><span class="p">[</span><span class="s1">&#39;sampleid_colors&#39;</span><span class="p">])))</span>
<span class="n">p</span><span class="o">.</span><span class="n">set</span><span class="p">(</span><span class="n">ylim</span><span class="o">=</span><span class="p">(</span><span class="o">-</span><span class="mf">0.1</span><span class="p">,</span><span class="mi">1</span><span class="p">),</span> <span class="n">xlim</span> <span class="o">=</span> <span class="p">(</span><span class="o">-</span><span class="mf">0.1</span><span class="p">,</span><span class="mi">1</span><span class="p">))</span>
<span class="n">p</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[50]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
&lt;AxesSubplot:xlabel=&#39;clone_size_gini&#39;, ylabel=&#39;clone_centrality_gini&#39;&gt;
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_3_dandelion_findingclones_and_analysis-10x_data_95_1.png" class="no-scaled-link" src="../_images/notebooks_3_dandelion_findingclones_and_analysis-10x_data_95_1.png" style="width: 329px; height: 306px;" />
</div>
</div>
<p>Chao1 is an estimator based on abundance</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[51]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">ddl</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">clone_diversity</span><span class="p">(</span><span class="n">vdj</span><span class="p">,</span> <span class="n">groupby</span> <span class="o">=</span> <span class="s1">&#39;sample_id&#39;</span><span class="p">,</span> <span class="n">method</span> <span class="o">=</span> <span class="s1">&#39;chao1&#39;</span><span class="p">,</span> <span class="n">update_obs_meta</span> <span class="o">=</span> <span class="kc">False</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Calculating Chao1 estimates
 finished (0:00:01)
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[51]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>clone_size_chao1</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>vdj_v1_hs_pbmc3</th>
      <td>703.0</td>
    </tr>
    <tr>
      <th>sc5p_v2_hs_PBMC_1k</th>
      <td>561.0</td>
    </tr>
    <tr>
      <th>sc5p_v2_hs_PBMC_10k</th>
      <td>44205.5</td>
    </tr>
    <tr>
      <th>vdj_nextgem_hs_pbmc3</th>
      <td>9106.0</td>
    </tr>
  </tbody>
</table>
</div></div>
</div>
<p>For Shannon Entropy, we can calculate a normalized (inspired by <code class="docutils literal notranslate"><span class="pre">scirpy</span></code>’s <a class="reference external" href="https://icbi-lab.github.io/scirpy/generated/scirpy.tl.alpha_diversity.html?highlight=diversity#scirpy.tl.alpha_diversity">function</a>) and non-normalized value.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[52]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">ddl</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">clone_diversity</span><span class="p">(</span><span class="n">vdj</span><span class="p">,</span> <span class="n">groupby</span> <span class="o">=</span> <span class="s1">&#39;sample_id&#39;</span><span class="p">,</span> <span class="n">method</span> <span class="o">=</span> <span class="s1">&#39;shannon&#39;</span><span class="p">,</span> <span class="n">update_obs_meta</span> <span class="o">=</span> <span class="kc">False</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Calculating Shannon entropy
 finished (0:00:01)
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[52]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>clone_size_normalized_shannon</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>vdj_v1_hs_pbmc3</th>
      <td>1.000000</td>
    </tr>
    <tr>
      <th>sc5p_v2_hs_PBMC_1k</th>
      <td>1.000000</td>
    </tr>
    <tr>
      <th>sc5p_v2_hs_PBMC_10k</th>
      <td>0.999849</td>
    </tr>
    <tr>
      <th>vdj_nextgem_hs_pbmc3</th>
      <td>0.989883</td>
    </tr>
  </tbody>
</table>
</div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[53]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">ddl</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">clone_diversity</span><span class="p">(</span><span class="n">vdj</span><span class="p">,</span> <span class="n">groupby</span> <span class="o">=</span> <span class="s1">&#39;sample_id&#39;</span><span class="p">,</span> <span class="n">method</span> <span class="o">=</span> <span class="s1">&#39;shannon&#39;</span><span class="p">,</span> <span class="n">update_obs_meta</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span> <span class="n">normalize</span> <span class="o">=</span> <span class="kc">False</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Calculating Shannon entropy
 finished (0:00:01)
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[53]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>clone_size_shannon</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>vdj_v1_hs_pbmc3</th>
      <td>5.209453</td>
    </tr>
    <tr>
      <th>sc5p_v2_hs_PBMC_1k</th>
      <td>5.044394</td>
    </tr>
    <tr>
      <th>sc5p_v2_hs_PBMC_10k</th>
      <td>8.712926</td>
    </tr>
    <tr>
      <th>vdj_nextgem_hs_pbmc3</th>
      <td>8.285998</td>
    </tr>
  </tbody>
</table>
</div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[54]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">adata</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[54]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
AnnData object with n_obs × n_vars = 16492 × 1497
    obs: &#39;sampleid&#39;, &#39;batch&#39;, &#39;scrublet_score&#39;, &#39;n_genes&#39;, &#39;percent_mito&#39;, &#39;n_counts&#39;, &#39;is_doublet&#39;, &#39;filter_rna&#39;, &#39;has_bcr&#39;, &#39;filter_bcr_quality&#39;, &#39;filter_bcr_heavy&#39;, &#39;filter_bcr_light&#39;, &#39;bcr_QC_pass&#39;, &#39;filter_bcr&#39;, &#39;leiden&#39;, &#39;sample_id&#39;, &#39;clone_id&#39;, &#39;clone_id_by_size&#39;, &#39;isotype&#39;, &#39;lightchain&#39;, &#39;status&#39;, &#39;vdj_status&#39;, &#39;productive&#39;, &#39;umi_counts_heavy&#39;, &#39;umi_counts_light&#39;, &#39;v_call_heavy&#39;, &#39;v_call_light&#39;, &#39;j_call_heavy&#39;, &#39;j_call_light&#39;, &#39;c_call_heavy&#39;, &#39;c_call_light&#39;, &#39;changeo_clone_id&#39;, &#39;mu_freq&#39;, &#39;mu_freq_seq_r&#39;, &#39;mu_freq_seq_s&#39;, &#39;mu_freq_seq_r_IGK&#39;, &#39;mu_freq_seq_s_IGK&#39;, &#39;mu_freq_IGK&#39;, &#39;mu_freq_seq_r_IGH&#39;, &#39;mu_freq_seq_s_IGH&#39;, &#39;mu_freq_IGH&#39;, &#39;mu_freq_seq_r_IGL&#39;, &#39;mu_freq_seq_s_IGL&#39;, &#39;mu_freq_IGL&#39;, &#39;clone_id_size&#39;, &#39;clone_id_size_max_3&#39;, &#39;clone_network_cluster_size_gini&#39;, &#39;clone_network_vertex_size_gini&#39;, &#39;clone_centrality&#39;, &#39;clone_size_gini&#39;, &#39;clone_centrality_gini&#39;, &#39;clone_network_cluster_size_gini_expanded&#39;, &#39;clone_network_vertex_size_gini_expanded&#39;
    var: &#39;feature_types&#39;, &#39;genome&#39;, &#39;gene_ids-0&#39;, &#39;gene_ids-1&#39;, &#39;gene_ids-2&#39;, &#39;gene_ids-3&#39;, &#39;n_cells&#39;, &#39;highly_variable&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;, &#39;mean&#39;, &#39;std&#39;
    uns: &#39;bcr_QC_pass_colors&#39;, &#39;hvg&#39;, &#39;leiden&#39;, &#39;leiden_colors&#39;, &#39;neighbors&#39;, &#39;pca&#39;, &#39;umap&#39;, &#39;rna_neighbors&#39;, &#39;sampleid_colors&#39;, &#39;isotype_colors&#39;, &#39;clone_id_by_size_colors&#39;, &#39;status_colors&#39;, &#39;vdj_status_colors&#39;, &#39;clone_id_colors&#39;
    obsm: &#39;X_pca&#39;, &#39;X_umap&#39;, &#39;X_bcr&#39;
    varm: &#39;PCs&#39;
    obsp: &#39;connectivities&#39;, &#39;distances&#39;, &#39;rna_connectivities&#39;, &#39;rna_distances&#39;, &#39;bcr_connectivities&#39;, &#39;bcr_distances&#39;
</pre></div></div>
</div>
</div>
<div class="section" id="Additional-plotting-functions">
<h2>Additional plotting functions<a class="headerlink" href="#Additional-plotting-functions" title="Permalink to this headline">¶</a></h2>
<div class="section" id="barplot">
<h3><em>barplot</em><a class="headerlink" href="#barplot" title="Permalink to this headline">¶</a></h3>
<p><code class="docutils literal notranslate"><span class="pre">pl.barplot</span></code> is a generic barplot function that will plot items in the metadata slot as a bar plot. This function will also interact with <code class="docutils literal notranslate"><span class="pre">.obs</span></code> slot if a <code class="docutils literal notranslate"><span class="pre">scanpy</span></code> object is used in place of <code class="docutils literal notranslate"><span class="pre">Dandelion</span></code> object. However, if your <code class="docutils literal notranslate"><span class="pre">scanpy</span></code> object holds a lot of non-B cells, then the plotting will be just be saturated with nan values.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[55]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="kn">import</span> <span class="nn">matplotlib</span> <span class="k">as</span> <span class="nn">mpl</span>
<span class="n">mpl</span><span class="o">.</span><span class="n">rcParams</span><span class="o">.</span><span class="n">update</span><span class="p">(</span><span class="n">mpl</span><span class="o">.</span><span class="n">rcParamsDefault</span><span class="p">)</span>
<span class="n">ddl</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">barplot</span><span class="p">(</span><span class="n">vdj</span><span class="p">,</span> <span class="n">variable</span> <span class="o">=</span> <span class="s1">&#39;v_call_heavy&#39;</span><span class="p">,</span> <span class="n">figsize</span> <span class="o">=</span> <span class="p">(</span><span class="mi">12</span><span class="p">,</span> <span class="mi">4</span><span class="p">))</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[55]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
(&lt;Figure size 1200x400 with 1 Axes&gt;,
 &lt;AxesSubplot:title={&#39;center&#39;:&#39;v call heavy usage&#39;}, ylabel=&#39;proportion&#39;&gt;)
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_3_dandelion_findingclones_and_analysis-10x_data_103_1.png" class="no-scaled-link" src="../_images/notebooks_3_dandelion_findingclones_and_analysis-10x_data_103_1.png" style="width: 1005px; height: 432px;" />
</div>
</div>
<p>You can prevent it from sorting by specifying <code class="docutils literal notranslate"><span class="pre">sort_descending</span> <span class="pre">=</span> <span class="pre">None</span></code>. Colours can be changed with <code class="docutils literal notranslate"><span class="pre">palette</span></code> option.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[56]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">ddl</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">barplot</span><span class="p">(</span><span class="n">vdj</span><span class="p">,</span> <span class="n">variable</span> <span class="o">=</span> <span class="s1">&#39;v_call_heavy&#39;</span><span class="p">,</span> <span class="n">figsize</span> <span class="o">=</span> <span class="p">(</span><span class="mi">12</span><span class="p">,</span> <span class="mi">4</span><span class="p">),</span> <span class="n">sort_descending</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">palette</span> <span class="o">=</span> <span class="s1">&#39;tab20&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[56]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
(&lt;Figure size 1200x400 with 1 Axes&gt;,
 &lt;AxesSubplot:title={&#39;center&#39;:&#39;v call heavy usage&#39;}, ylabel=&#39;proportion&#39;&gt;)
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_3_dandelion_findingclones_and_analysis-10x_data_105_1.png" class="no-scaled-link" src="../_images/notebooks_3_dandelion_findingclones_and_analysis-10x_data_105_1.png" style="width: 1005px; height: 432px;" />
</div>
</div>
<p>Specifying <code class="docutils literal notranslate"><span class="pre">normalize</span> <span class="pre">=</span> <span class="pre">False</span></code> will change the y-axis to counts.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[57]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">ddl</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">barplot</span><span class="p">(</span><span class="n">vdj</span><span class="p">,</span> <span class="n">variable</span> <span class="o">=</span> <span class="s1">&#39;v_call_heavy&#39;</span><span class="p">,</span> <span class="n">normalize</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span> <span class="n">figsize</span> <span class="o">=</span> <span class="p">(</span><span class="mi">12</span><span class="p">,</span> <span class="mi">4</span><span class="p">),</span> <span class="n">sort_descending</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">palette</span> <span class="o">=</span> <span class="s1">&#39;tab20&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[57]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
(&lt;Figure size 1200x400 with 1 Axes&gt;,
 &lt;AxesSubplot:title={&#39;center&#39;:&#39;v call heavy usage&#39;}, ylabel=&#39;count&#39;&gt;)
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_3_dandelion_findingclones_and_analysis-10x_data_107_1.png" class="no-scaled-link" src="../_images/notebooks_3_dandelion_findingclones_and_analysis-10x_data_107_1.png" style="width: 993px; height: 432px;" />
</div>
</div>
</div>
<div class="section" id="stackedbarplot">
<h3><em>stackedbarplot</em><a class="headerlink" href="#stackedbarplot" title="Permalink to this headline">¶</a></h3>
<p><code class="docutils literal notranslate"><span class="pre">pl.stackedbarplot</span></code> is similar to above but can split between specified groups. Some examples below:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[58]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="kn">import</span> <span class="nn">matplotlib.pyplot</span> <span class="k">as</span> <span class="nn">plt</span>
<span class="n">ddl</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">stackedbarplot</span><span class="p">(</span><span class="n">vdj</span><span class="p">,</span> <span class="n">variable</span> <span class="o">=</span> <span class="s1">&#39;isotype&#39;</span><span class="p">,</span> <span class="n">groupby</span> <span class="o">=</span> <span class="s1">&#39;status&#39;</span><span class="p">,</span> <span class="n">xtick_rotation</span> <span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">figsize</span> <span class="o">=</span> <span class="p">(</span><span class="mi">4</span><span class="p">,</span><span class="mi">4</span><span class="p">))</span>
<span class="n">plt</span><span class="o">.</span><span class="n">legend</span><span class="p">(</span><span class="n">bbox_to_anchor</span> <span class="o">=</span> <span class="p">(</span><span class="mi">1</span><span class="p">,</span><span class="mi">1</span><span class="p">),</span> <span class="n">loc</span><span class="o">=</span><span class="s1">&#39;upper left&#39;</span><span class="p">,</span> <span class="n">frameon</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[58]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
&lt;matplotlib.legend.Legend at 0x169239d50&gt;
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_3_dandelion_findingclones_and_analysis-10x_data_109_1.png" class="no-scaled-link" src="../_images/notebooks_3_dandelion_findingclones_and_analysis-10x_data_109_1.png" style="width: 554px; height: 371px;" />
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[59]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">ddl</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">stackedbarplot</span><span class="p">(</span><span class="n">vdj</span><span class="p">,</span> <span class="n">variable</span> <span class="o">=</span> <span class="s1">&#39;v_call_heavy&#39;</span><span class="p">,</span> <span class="n">groupby</span> <span class="o">=</span> <span class="s1">&#39;isotype&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">legend</span><span class="p">(</span><span class="n">bbox_to_anchor</span> <span class="o">=</span> <span class="p">(</span><span class="mi">1</span><span class="p">,</span><span class="mi">1</span><span class="p">),</span> <span class="n">loc</span><span class="o">=</span><span class="s1">&#39;upper left&#39;</span><span class="p">,</span> <span class="n">frameon</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[59]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
&lt;matplotlib.legend.Legend at 0x16c216750&gt;
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_3_dandelion_findingclones_and_analysis-10x_data_110_1.png" class="no-scaled-link" src="../_images/notebooks_3_dandelion_findingclones_and_analysis-10x_data_110_1.png" style="width: 1073px; height: 432px;" />
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[60]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">ddl</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">stackedbarplot</span><span class="p">(</span><span class="n">vdj</span><span class="p">,</span> <span class="n">variable</span> <span class="o">=</span> <span class="s1">&#39;v_call_heavy&#39;</span><span class="p">,</span> <span class="n">groupby</span> <span class="o">=</span> <span class="s1">&#39;isotype&#39;</span><span class="p">,</span> <span class="n">normalize</span> <span class="o">=</span> <span class="kc">True</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">legend</span><span class="p">(</span><span class="n">bbox_to_anchor</span> <span class="o">=</span> <span class="p">(</span><span class="mi">1</span><span class="p">,</span><span class="mi">1</span><span class="p">),</span> <span class="n">loc</span><span class="o">=</span><span class="s1">&#39;upper left&#39;</span><span class="p">,</span> <span class="n">frameon</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[60]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
&lt;matplotlib.legend.Legend at 0x16adeda90&gt;
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_3_dandelion_findingclones_and_analysis-10x_data_111_1.png" class="no-scaled-link" src="../_images/notebooks_3_dandelion_findingclones_and_analysis-10x_data_111_1.png" style="width: 1077px; height: 432px;" />
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[61]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">ddl</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">stackedbarplot</span><span class="p">(</span><span class="n">vdj</span><span class="p">,</span> <span class="n">variable</span> <span class="o">=</span> <span class="s1">&#39;v_call_heavy&#39;</span><span class="p">,</span> <span class="n">groupby</span> <span class="o">=</span> <span class="s1">&#39;vdj_status&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">legend</span><span class="p">(</span><span class="n">bbox_to_anchor</span> <span class="o">=</span> <span class="p">(</span><span class="mi">1</span><span class="p">,</span><span class="mi">1</span><span class="p">),</span> <span class="n">loc</span><span class="o">=</span><span class="s1">&#39;upper left&#39;</span><span class="p">,</span> <span class="n">frameon</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[61]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
&lt;matplotlib.legend.Legend at 0x16aade450&gt;
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_3_dandelion_findingclones_and_analysis-10x_data_112_1.png" class="no-scaled-link" src="../_images/notebooks_3_dandelion_findingclones_and_analysis-10x_data_112_1.png" style="width: 1204px; height: 432px;" />
</div>
</div>
<p>It’s obviously more useful if you don’t have too many groups, but you could try and plot everything and jiggle the legend options and color.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[62]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">ddl</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">stackedbarplot</span><span class="p">(</span><span class="n">vdj</span><span class="p">,</span> <span class="n">variable</span> <span class="o">=</span> <span class="s1">&#39;v_call_heavy&#39;</span><span class="p">,</span> <span class="n">groupby</span> <span class="o">=</span> <span class="s1">&#39;sample_id&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">legend</span><span class="p">(</span><span class="n">bbox_to_anchor</span> <span class="o">=</span> <span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mf">0.5</span><span class="p">),</span> <span class="n">loc</span><span class="o">=</span><span class="s1">&#39;center left&#39;</span><span class="p">,</span> <span class="n">frameon</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[62]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
&lt;matplotlib.legend.Legend at 0x1698e8250&gt;
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_3_dandelion_findingclones_and_analysis-10x_data_114_1.png" class="no-scaled-link" src="../_images/notebooks_3_dandelion_findingclones_and_analysis-10x_data_114_1.png" style="width: 1203px; height: 432px;" />
</div>
</div>
</div>
<div class="section" id="spectratype">
<h3><em>spectratype</em><a class="headerlink" href="#spectratype" title="Permalink to this headline">¶</a></h3>
<p>Spectratype plots contain info displaying CDR3 length distribution for specified groups. For this function, the current method only works for <code class="docutils literal notranslate"><span class="pre">dandelion</span></code> objects as it requires access to the contig-indexed <em>.data</em> slot.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[63]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">ddl</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">spectratype</span><span class="p">(</span><span class="n">vdj</span><span class="p">,</span> <span class="n">variable</span> <span class="o">=</span> <span class="s1">&#39;junction_length&#39;</span><span class="p">,</span> <span class="n">groupby</span> <span class="o">=</span> <span class="s1">&#39;c_call&#39;</span><span class="p">,</span> <span class="n">locus</span><span class="o">=</span><span class="s1">&#39;IGH&#39;</span><span class="p">,</span> <span class="n">width</span> <span class="o">=</span> <span class="mf">2.3</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">legend</span><span class="p">(</span><span class="n">bbox_to_anchor</span> <span class="o">=</span> <span class="p">(</span><span class="mi">1</span><span class="p">,</span><span class="mi">1</span><span class="p">),</span> <span class="n">loc</span><span class="o">=</span><span class="s1">&#39;upper left&#39;</span><span class="p">,</span> <span class="n">frameon</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[63]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
&lt;matplotlib.legend.Legend at 0x1690f9590&gt;
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_3_dandelion_findingclones_and_analysis-10x_data_116_1.png" class="no-scaled-link" src="../_images/notebooks_3_dandelion_findingclones_and_analysis-10x_data_116_1.png" style="width: 629px; height: 369px;" />
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[64]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">ddl</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">spectratype</span><span class="p">(</span><span class="n">vdj</span><span class="p">,</span> <span class="n">variable</span> <span class="o">=</span> <span class="s1">&#39;junction_aa_length&#39;</span><span class="p">,</span> <span class="n">groupby</span> <span class="o">=</span> <span class="s1">&#39;c_call&#39;</span><span class="p">,</span> <span class="n">locus</span><span class="o">=</span><span class="s1">&#39;IGH&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">legend</span><span class="p">(</span><span class="n">bbox_to_anchor</span> <span class="o">=</span> <span class="p">(</span><span class="mi">1</span><span class="p">,</span><span class="mi">1</span><span class="p">),</span> <span class="n">loc</span><span class="o">=</span><span class="s1">&#39;upper left&#39;</span><span class="p">,</span> <span class="n">frameon</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[64]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
&lt;matplotlib.legend.Legend at 0x16b8d9d10&gt;
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_3_dandelion_findingclones_and_analysis-10x_data_117_1.png" class="no-scaled-link" src="../_images/notebooks_3_dandelion_findingclones_and_analysis-10x_data_117_1.png" style="width: 629px; height: 369px;" />
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[65]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">ddl</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">spectratype</span><span class="p">(</span><span class="n">vdj</span><span class="p">,</span> <span class="n">variable</span> <span class="o">=</span> <span class="s1">&#39;junction_aa_length&#39;</span><span class="p">,</span> <span class="n">groupby</span> <span class="o">=</span> <span class="s1">&#39;c_call&#39;</span><span class="p">,</span> <span class="n">locus</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;IGK&#39;</span><span class="p">,</span><span class="s1">&#39;IGL&#39;</span><span class="p">])</span>
<span class="n">plt</span><span class="o">.</span><span class="n">legend</span><span class="p">(</span><span class="n">bbox_to_anchor</span> <span class="o">=</span> <span class="p">(</span><span class="mi">1</span><span class="p">,</span><span class="mi">1</span><span class="p">),</span> <span class="n">loc</span><span class="o">=</span><span class="s1">&#39;upper left&#39;</span><span class="p">,</span> <span class="n">frameon</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[65]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
&lt;matplotlib.legend.Legend at 0x16b793e50&gt;
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_3_dandelion_findingclones_and_analysis-10x_data_118_1.png" class="no-scaled-link" src="../_images/notebooks_3_dandelion_findingclones_and_analysis-10x_data_118_1.png" style="width: 669px; height: 369px;" />
</div>
</div>
</div>
<div class="section" id="clone_overlap">
<h3><em>clone_overlap</em><a class="headerlink" href="#clone_overlap" title="Permalink to this headline">¶</a></h3>
<p>There is now a circos-style clone overlap function where it looks for whather different samples share a clone. If they do, an arc/connection will be drawn between them. This requires the python module <code class="docutils literal notranslate"><span class="pre">nxviz</span></code> to be installed; at the writing of this notebook, there are some dependencies issues with <code class="docutils literal notranslate"><span class="pre">pip</span> <span class="pre">install</span> <span class="pre">nxviz</span></code>, therefore I’ve adjusted the requirements in a forked repository which you can install via: <code class="docutils literal notranslate"><span class="pre">pip</span> <span class="pre">install</span> <span class="pre">git+https://github.com/zktuong/nxviz.git</span></code></p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[66]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">ddl</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">clone_overlap</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">groupby</span> <span class="o">=</span> <span class="s1">&#39;leiden&#39;</span><span class="p">,</span> <span class="n">colorby</span> <span class="o">=</span> <span class="s1">&#39;leiden&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Finding clones
 finished: Updated AnnData:
   &#39;uns&#39;, clone overlap table (0:00:00)
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[67]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">ddl</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">clone_overlap</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">groupby</span> <span class="o">=</span> <span class="s1">&#39;leiden&#39;</span><span class="p">,</span> <span class="n">colorby</span> <span class="o">=</span> <span class="s1">&#39;leiden&#39;</span><span class="p">,</span> <span class="n">return_graph</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">group_label_offset</span><span class="o">=.</span><span class="mi">5</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[67]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
&lt;nxviz.plots.CircosPlot at 0x16c3e4f90&gt;
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_3_dandelion_findingclones_and_analysis-10x_data_121_1.png" class="no-scaled-link" src="../_images/notebooks_3_dandelion_findingclones_and_analysis-10x_data_121_1.png" style="width: 667px; height: 687px;" />
</div>
</div>
<p>Other use cases for this would be, for example, to plot nodes as individual samples and the colors as group classifications of the samples. As long as this information is found in the <code class="docutils literal notranslate"><span class="pre">.obs</span></code> column in the <code class="docutils literal notranslate"><span class="pre">AnnData</span></code>, or even <code class="docutils literal notranslate"><span class="pre">Dandelion.metadata</span></code>, this will work.</p>
<p>That sums it up for now! Let me know if you have any ideas at [<a class="reference external" href="mailto:kt16&#37;&#52;&#48;sanger&#46;ac&#46;uk">kt16<span>&#64;</span>sanger<span>&#46;</span>ac<span>&#46;</span>uk</a>] and I can try and see if i can implement it or we can work something out to collaborate on!</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span>
</pre></div>
</div>
</div>
</div>
</div>
</div>


           </div>
           
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="5_dandelion_running_from_R.html" class="btn btn-neutral float-right" title="Running from R" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="2_dandelion_filtering-10x%20data.html" class="btn btn-neutral float-left" title="Filtering" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        
        &copy; Copyright 2020, zktuong

    </p>
  </div>
    
    
    
    Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a
    
    <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a>
    
    provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  

  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>