

<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>dandelion Notebook-5 &mdash; dandelion 0.0.26 documentation</title>
  

  
  <link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />

  
  
    <link rel="shortcut icon" href="../_static/dandelion_logo.png"/>
  
  
  

  
  <!--[if lt IE 9]>
    <script src="../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="../" src="../_static/documentation_options.js"></script>
        <script src="../_static/jquery.js"></script>
        <script src="../_static/underscore.js"></script>
        <script src="../_static/doctools.js"></script>
        <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
        <script async="async" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.7/latest.js?config=TeX-AMS-MML_HTMLorMML"></script>
        <script type="text/x-mathjax-config">MathJax.Hub.Config({"tex2jax": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true, "ignoreClass": "document", "processClass": "math|output_area"}})</script>
    
    <script type="text/javascript" src="../_static/js/theme.js"></script>

    
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="prev" title="dandelion Notebook-3" href="3_dandelion_findingclones_and_analysis-10x%20data.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="../index.html">
          

          
            
            <img src="../_static/dandelion_logo.png" class="logo" alt="Logo"/>
          
          </a>

          
            
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        
        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <p class="caption"><span class="caption-text">Tutorials:</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="1_dandelion_preprocessing-10x%20data.html"><em>dandelion</em> Notebook-1</a></li>
<li class="toctree-l1"><a class="reference internal" href="2_dandelion_filtering-10x%20data.html"><em>dandelion</em> Notebook-2</a></li>
<li class="toctree-l1"><a class="reference internal" href="3_dandelion_findingclones_and_analysis-10x%20data.html"><em>dandelion</em> Notebook-3</a></li>
<li class="toctree-l1"><a class="reference internal" href="3_dandelion_findingclones_and_analysis-10x%20data.html#BCR-clustering-and-visualization">BCR clustering and visualization</a></li>
<li class="toctree-l1"><a class="reference internal" href="3_dandelion_findingclones_and_analysis-10x%20data.html#Finding-clones">Finding clones</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#"><em>dandelion</em> Notebook-5</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#Pre-processing">Pre-processing</a></li>
<li class="toctree-l2"><a class="reference internal" href="#Step-1:">Step 1:</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#Formatting-the-headers-of-the-cellranger-fasta-file">Formatting the headers of the cellranger fasta file</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#Step-2:">Step 2:</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#Reannotate-the-V/D/J-genes-with-igblastn.">Reannotate the V/D/J genes with <em>igblastn</em>.</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#Step-3-(optional-but-recommended):">Step 3 <em>(optional but recommended)</em>:</a></li>
<li class="toctree-l2"><a class="reference internal" href="#Step-4:">Step 4:</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#Assigning-constant-region-calls">Assigning constant region calls</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#Step-5:-Quantify-mutations-(optional).">Step 5: Quantify mutations <em>(optional)</em>.</a></li>
<li class="toctree-l2"><a class="reference internal" href="#Step-6:-Marking-BCR-doublets-and-poor-quality-BCRs">Step 6: Marking BCR doublets and poor quality BCRs</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#Create-a-Seurat-object-from-the-transcriptome-data">Create a Seurat object from the transcriptome data</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#Standard-Seurat-pre-processing-workflow">Standard Seurat pre-processing workflow</a></li>
<li class="toctree-l2"><a class="reference internal" href="#Convert-to-AnnData">Convert to AnnData</a></li>
<li class="toctree-l2"><a class="reference internal" href="#Read-in-the-BCR-files-and-merge-them">Read in the BCR files and merge them</a></li>
<li class="toctree-l2"><a class="reference internal" href="#Step-7:-Run-ddl$pp$filter_bcr">Step 7: Run <code class="docutils literal notranslate"><span class="pre">ddl$pp$filter_bcr</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="#Step-8-:-Finding-clones">Step 8 : Finding clones</a></li>
<li class="toctree-l2"><a class="reference internal" href="#Calculating-size-of-clones">Calculating size of clones</a></li>
<li class="toctree-l2"><a class="reference internal" href="#Step-9-:-Generate-BCR-network-visualization">Step 9 : Generate BCR network visualization</a></li>
<li class="toctree-l2"><a class="reference internal" href="#Step-10-:-Integrating-with-Seurat">Step 10 : Integrating with Seurat</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#Transfer-Dandelion-to-AnnData">Transfer <code class="docutils literal notranslate"><span class="pre">Dandelion</span></code> to <code class="docutils literal notranslate"><span class="pre">AnnData</span></code></a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#Save">Save</a></li>
<li class="toctree-l2"><a class="reference internal" href="#New-Session:-Transfer-AnnData-to-Seurat">New Session: Transfer <code class="docutils literal notranslate"><span class="pre">AnnData</span></code> to <code class="docutils literal notranslate"><span class="pre">Seurat</span></code></a><ul>
<li class="toctree-l3"><a class="reference internal" href="#Extract-the-coordinates-from-AnnData-to-a-R-friendly-version">Extract the coordinates from <code class="docutils literal notranslate"><span class="pre">AnnData</span></code> to a R friendly version</a></li>
</ul>
</li>
</ul>
</li>
</ul>

            
          
        </div>
        
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">dandelion</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="../index.html" class="icon icon-home"></a> &raquo;</li>
        
      <li><em>dandelion</em> Notebook-5</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            <a href="../_sources/notebooks/5_dandelion_running_from_R.ipynb.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  
<style>
/* CSS for nbsphinx extension */

/* remove conflicting styling from Sphinx themes */
div.nbinput.container div.prompt *,
div.nboutput.container div.prompt *,
div.nbinput.container div.input_area pre,
div.nboutput.container div.output_area pre,
div.nbinput.container div.input_area .highlight,
div.nboutput.container div.output_area .highlight {
    border: none;
    padding: 0;
    margin: 0;
    box-shadow: none;
}

div.nbinput.container > div[class*=highlight],
div.nboutput.container > div[class*=highlight] {
    margin: 0;
}

div.nbinput.container div.prompt *,
div.nboutput.container div.prompt * {
    background: none;
}

div.nboutput.container div.output_area .highlight,
div.nboutput.container div.output_area pre {
    background: unset;
}

div.nboutput.container div.output_area div.highlight {
    color: unset;  /* override Pygments text color */
}

/* avoid gaps between output lines */
div.nboutput.container div[class*=highlight] pre {
    line-height: normal;
}

/* input/output containers */
div.nbinput.container,
div.nboutput.container {
    display: -webkit-flex;
    display: flex;
    align-items: flex-start;
    margin: 0;
    width: 100%;
}
@media (max-width: 540px) {
    div.nbinput.container,
    div.nboutput.container {
        flex-direction: column;
    }
}

/* input container */
div.nbinput.container {
    padding-top: 5px;
}

/* last container */
div.nblast.container {
    padding-bottom: 5px;
}

/* input prompt */
div.nbinput.container div.prompt pre {
    color: #307FC1;
}

/* output prompt */
div.nboutput.container div.prompt pre {
    color: #BF5B3D;
}

/* all prompts */
div.nbinput.container div.prompt,
div.nboutput.container div.prompt {
    width: 4.5ex;
    padding-top: 5px;
    position: relative;
    user-select: none;
}

div.nbinput.container div.prompt > div,
div.nboutput.container div.prompt > div {
    position: absolute;
    right: 0;
    margin-right: 0.3ex;
}

@media (max-width: 540px) {
    div.nbinput.container div.prompt,
    div.nboutput.container div.prompt {
        width: unset;
        text-align: left;
        padding: 0.4em;
    }
    div.nboutput.container div.prompt.empty {
        padding: 0;
    }

    div.nbinput.container div.prompt > div,
    div.nboutput.container div.prompt > div {
        position: unset;
    }
}

/* disable scrollbars on prompts */
div.nbinput.container div.prompt pre,
div.nboutput.container div.prompt pre {
    overflow: hidden;
}

/* input/output area */
div.nbinput.container div.input_area,
div.nboutput.container div.output_area {
    -webkit-flex: 1;
    flex: 1;
    overflow: auto;
}
@media (max-width: 540px) {
    div.nbinput.container div.input_area,
    div.nboutput.container div.output_area {
        width: 100%;
    }
}

/* input area */
div.nbinput.container div.input_area {
    border: 1px solid #e0e0e0;
    border-radius: 2px;
    /*background: #f5f5f5;*/
}

/* override MathJax center alignment in output cells */
div.nboutput.container div[class*=MathJax] {
    text-align: left !important;
}

/* override sphinx.ext.imgmath center alignment in output cells */
div.nboutput.container div.math p {
    text-align: left;
}

/* standard error */
div.nboutput.container div.output_area.stderr {
    background: #fdd;
}

/* ANSI colors */
.ansi-black-fg { color: #3E424D; }
.ansi-black-bg { background-color: #3E424D; }
.ansi-black-intense-fg { color: #282C36; }
.ansi-black-intense-bg { background-color: #282C36; }
.ansi-red-fg { color: #E75C58; }
.ansi-red-bg { background-color: #E75C58; }
.ansi-red-intense-fg { color: #B22B31; }
.ansi-red-intense-bg { background-color: #B22B31; }
.ansi-green-fg { color: #00A250; }
.ansi-green-bg { background-color: #00A250; }
.ansi-green-intense-fg { color: #007427; }
.ansi-green-intense-bg { background-color: #007427; }
.ansi-yellow-fg { color: #DDB62B; }
.ansi-yellow-bg { background-color: #DDB62B; }
.ansi-yellow-intense-fg { color: #B27D12; }
.ansi-yellow-intense-bg { background-color: #B27D12; }
.ansi-blue-fg { color: #208FFB; }
.ansi-blue-bg { background-color: #208FFB; }
.ansi-blue-intense-fg { color: #0065CA; }
.ansi-blue-intense-bg { background-color: #0065CA; }
.ansi-magenta-fg { color: #D160C4; }
.ansi-magenta-bg { background-color: #D160C4; }
.ansi-magenta-intense-fg { color: #A03196; }
.ansi-magenta-intense-bg { background-color: #A03196; }
.ansi-cyan-fg { color: #60C6C8; }
.ansi-cyan-bg { background-color: #60C6C8; }
.ansi-cyan-intense-fg { color: #258F8F; }
.ansi-cyan-intense-bg { background-color: #258F8F; }
.ansi-white-fg { color: #C5C1B4; }
.ansi-white-bg { background-color: #C5C1B4; }
.ansi-white-intense-fg { color: #A1A6B2; }
.ansi-white-intense-bg { background-color: #A1A6B2; }

.ansi-default-inverse-fg { color: #FFFFFF; }
.ansi-default-inverse-bg { background-color: #000000; }

.ansi-bold { font-weight: bold; }
.ansi-underline { text-decoration: underline; }


div.nbinput.container div.input_area div[class*=highlight] > pre,
div.nboutput.container div.output_area div[class*=highlight] > pre,
div.nboutput.container div.output_area div[class*=highlight].math,
div.nboutput.container div.output_area.rendered_html,
div.nboutput.container div.output_area > div.output_javascript,
div.nboutput.container div.output_area:not(.rendered_html) > img{
    padding: 5px;
    margin: 0;
}

/* fix copybtn overflow problem in chromium (needed for 'sphinx_copybutton') */
div.nbinput.container div.input_area > div[class^='highlight'],
div.nboutput.container div.output_area > div[class^='highlight']{
    overflow-y: hidden;
}

/* hide copybtn icon on prompts (needed for 'sphinx_copybutton') */
.prompt a.copybtn {
    display: none;
}

/* Some additional styling taken form the Jupyter notebook CSS */
div.rendered_html table {
  border: none;
  border-collapse: collapse;
  border-spacing: 0;
  color: black;
  font-size: 12px;
  table-layout: fixed;
}
div.rendered_html thead {
  border-bottom: 1px solid black;
  vertical-align: bottom;
}
div.rendered_html tr,
div.rendered_html th,
div.rendered_html td {
  text-align: right;
  vertical-align: middle;
  padding: 0.5em 0.5em;
  line-height: normal;
  white-space: normal;
  max-width: none;
  border: none;
}
div.rendered_html th {
  font-weight: bold;
}
div.rendered_html tbody tr:nth-child(odd) {
  background: #f5f5f5;
}
div.rendered_html tbody tr:hover {
  background: rgba(66, 165, 245, 0.2);
}

/* CSS overrides for sphinx_rtd_theme */

/* 24px margin */
.nbinput.nblast.container,
.nboutput.nblast.container {
    margin-bottom: 19px;  /* padding has already 5px */
}

/* ... except between code cells! */
.nblast.container + .nbinput.container {
    margin-top: -19px;
}

.admonition > p:before {
    margin-right: 4px;  /* make room for the exclamation icon */
}

/* Fix math alignment, see https://github.com/rtfd/sphinx_rtd_theme/pull/686 */
.math {
    text-align: unset;
}
</style>
<div class="section" id="dandelion-Notebook-5">
<h1><em>dandelion</em> Notebook-5<a class="headerlink" href="#dandelion-Notebook-5" title="Permalink to this headline">¶</a></h1>
<p><img alt="dandelion_logo" src="../_images/dandelion_logo.png" /> ## Foreword <strong>dandelion</strong> is written in <code class="docutils literal notranslate"><span class="pre">python==3.7.6</span></code> and it is primarily a single-cell BCR-seq analysis package. It makes use of some tools from the fantastic <a class="reference external" href="https://immcantation.readthedocs.io/">immcantation suite</a> and the main idea is that it implements a workflow for the pre-processing and exploratory stages with integrated use of tools from <em>immcantation</em> for the BCR side of things and analysis tools from <a class="reference external" href="https://scanpy.readthedocs.io/">scanpy</a> for the RNA-seq
side of things. I hope to be able to introduce some new single-cell BCR-seq exploratory tools down the road through <em>dandelion</em>.</p>
<p><strong>dandelion</strong> can be run in <code class="docutils literal notranslate"><span class="pre">R</span></code> through <code class="docutils literal notranslate"><span class="pre">reticulate</span></code>. This notebook will try to replicate the examples in notebooks 1-3 entirely in R.</p>
<p>There are some issues with the conversion of dataframes between python and R so I would <strong>not</strong> recommend saving the final <code class="docutils literal notranslate"><span class="pre">AnnData</span></code> object as a final out file, but only use this to help generate the intermediate files from the BCR processing and the plots. I would also skip quantify mutation step due to conflicts between rpy2 and reticulate.</p>
<p>For more details, please refer to the original notebooks 1-3.</p>
<p>Let’s start!</p>
<p>First, install reticulate via if you don’t already have it:</p>
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="nf">install.packages</span><span class="p">(</span><span class="s">&#39;reticulate&#39;</span><span class="p">)</span>
</pre></div>
</div>
<p>Because we are managing the packages through a conda virtual environment, we will need to point reticulate to the right python paths.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[1]:
</pre></div>
</div>
<div class="input_area highlight-r notranslate"><div class="highlight"><pre>
<span></span><span class="nf">library</span><span class="p">(</span><span class="n">reticulate</span><span class="p">)</span>
<span class="nf">use_condaenv</span><span class="p">(</span><span class="s">&#39;dandelion&#39;</span><span class="p">)</span>
<span class="c1"># or use Sys.setenv(RETICULATE_PYTHON = conda_python(envname=&#39;dandelion&#39;))</span>
</pre></div>
</div>
</div>
<p>You can check if the python config is set up properly with <code class="docutils literal notranslate"><span class="pre">py_config()</span></code></p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[2]:
</pre></div>
</div>
<div class="input_area highlight-r notranslate"><div class="highlight"><pre>
<span></span><span class="nf">py_config</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
python:         /Users/kt16/miniconda3/envs/dandelion/bin/python
libpython:      /Users/kt16/miniconda3/envs/dandelion/lib/libpython3.7m.dylib
pythonhome:     /Users/kt16/miniconda3/envs/dandelion:/Users/kt16/miniconda3/envs/dandelion
version:        3.7.6 | packaged by conda-forge | (default, Jun  1 2020, 18:33:30)  [Clang 9.0.1 ]
numpy:          /Users/kt16/miniconda3/envs/dandelion/lib/python3.7/site-packages/numpy
numpy_version:  1.17.5

NOTE: Python version was forced by RETICULATE_PYTHON
</pre></div></div>
</div>
<p>To proceed with the analyses, we first change the working directory and also import the dandelion module.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[3]:
</pre></div>
</div>
<div class="input_area highlight-r notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># change directory to somewhere more workable</span>
<span class="nf">setwd</span><span class="p">(</span><span class="s">&#39;/Users/kt16/Downloads/dandelion_tutorial_R/&#39;</span><span class="p">)</span>
<span class="c1"># ddl = import_from_path(&#39;dandelion&#39;, path = &#39;/Users/kt16/Documents/Github/dandelion&#39;)</span>
<span class="n">ddl</span> <span class="o">=</span> <span class="nf">import</span><span class="p">(</span><span class="s">&#39;dandelion&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<p>As per reticulate convention, python <code class="docutils literal notranslate"><span class="pre">.</span></code> operators are to be swaped with <code class="docutils literal notranslate"><span class="pre">$</span></code> in R.</p>
<div class="section" id="Pre-processing">
<h2>Pre-processing<a class="headerlink" href="#Pre-processing" title="Permalink to this headline">¶</a></h2>
<p>To being the BCR preprocessing, the minimum files you require are the following:</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="go">filtered_contig.fasta</span>
<span class="go">filtered_contig_annotations.csv</span>
</pre></div>
</div>
</div>
<div class="section" id="Step-1:">
<h2>Step 1:<a class="headerlink" href="#Step-1:" title="Permalink to this headline">¶</a></h2>
<div class="section" id="Formatting-the-headers-of-the-cellranger-fasta-file">
<h3>Formatting the headers of the cellranger fasta file<a class="headerlink" href="#Formatting-the-headers-of-the-cellranger-fasta-file" title="Permalink to this headline">¶</a></h3>
<p>This will do the following: 1) add the prefix provided to every sequence header</p>
<ol class="arabic simple" start="2">
<li><p>add the prefix provided to every contig_id in the annotation.csv file</p></li>
<li><p>create a folder called <code class="docutils literal notranslate"><span class="pre">dandelion/data</span></code> (if left as default) and saves a copy of these files in that directory.</p></li>
</ol>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[4]:
</pre></div>
</div>
<div class="input_area highlight-r notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># the first option is a list of fasta files to format and the second option is the list of prefix to add to each file.</span>
<span class="n">samples</span> <span class="o">=</span> <span class="nf">c</span><span class="p">(</span><span class="s">&#39;sc5p_v2_hs_PBMC_1k&#39;</span><span class="p">,</span> <span class="s">&#39;sc5p_v2_hs_PBMC_10k&#39;</span><span class="p">,</span> <span class="s">&#39;vdj_v1_hs_pbmc3&#39;</span><span class="p">,</span> <span class="s">&#39;vdj_nextgem_hs_pbmc3&#39;</span><span class="p">)</span>
<span class="n">ddl</span><span class="o">$</span><span class="n">pp</span><span class="o">$</span><span class="nf">format_fastas</span><span class="p">(</span><span class="n">samples</span><span class="p">,</span> <span class="n">prefix</span> <span class="o">=</span> <span class="n">samples</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="Step-2:">
<h2>Step 2:<a class="headerlink" href="#Step-2:" title="Permalink to this headline">¶</a></h2>
<div class="section" id="Reannotate-the-V/D/J-genes-with-igblastn.">
<h3>Reannotate the V/D/J genes with <em>igblastn</em>.<a class="headerlink" href="#Reannotate-the-V/D/J-genes-with-igblastn." title="Permalink to this headline">¶</a></h3>
<p><code class="docutils literal notranslate"><span class="pre">ddl$pp$reannotate_genes</span></code> uses <a class="reference external" href="https://changeo.readthedocs.io/en/stable/examples/10x.html">changeo</a>’s scripts to call <em>igblastn</em> to reannotate the fasta files. Depending on the <code class="docutils literal notranslate"><span class="pre">fileformat</span></code> option, it will parse out as either an <code class="docutils literal notranslate"><span class="pre">airr</span></code> (default) or <code class="docutils literal notranslate"><span class="pre">changeo</span></code>-legacy TSV file. Importantly, with the recent update to changeo v1.0.0, all the column headers, including <em>changeo</em> format, are now adhereing to the <a class="reference external" href="http://docs.airr-community.org/">AIRR</a> standard (lowercase and some column
name changes). Specifying <code class="docutils literal notranslate"><span class="pre">extended</span> <span class="pre">=</span> <span class="pre">True</span></code> will return the additional 10x annotation of V/D/J genes but they are unnecessary at this stage.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[5]:
</pre></div>
</div>
<div class="input_area highlight-r notranslate"><div class="highlight"><pre>
<span></span><span class="n">ddl</span><span class="o">$</span><span class="n">pp</span><span class="o">$</span><span class="nf">reannotate_genes</span><span class="p">(</span><span class="n">samples</span><span class="p">)</span>
</pre></div>
</div>
</div>
<p>you should see something like this in the terminal:</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="go">Assigning genes :   0%|                                        | 0/4 [00:00&lt;?, ?it/s]</span>
<span class="go">START&gt; AssignGenes</span>
<span class="go"> COMMAND&gt; igblast</span>
<span class="go"> VERSION&gt; 1.15.0</span>
<span class="go">    FILE&gt; sc5p_v2_hs_PBMC_10k_b_filtered_contig.fasta</span>
<span class="go">ORGANISM&gt; human</span>
<span class="go">    LOCI&gt; ig</span>
<span class="go">   NPROC&gt; 4</span>

<span class="go">PROGRESS&gt; 12:30:37 |Running IgBLAST          | 0.0 min</span>
</pre></div>
</div>
</div>
</div>
<div class="section" id="Step-3-(optional-but-recommended):">
<h2>Step 3 <em>(optional but recommended)</em>:<a class="headerlink" href="#Step-3-(optional-but-recommended):" title="Permalink to this headline">¶</a></h2>
<p><em>note that prior to v0.0.25, this was step 4</em> #### Reassigning heavy chain V gene alleles.</p>
<p>Next, we use <em>immcantation’s TIgGER</em> method to reassign allelic calls for heavy chain V genes with <code class="docutils literal notranslate"><span class="pre">pp.reassign_alleles</span></code>. As stated in TIgGER’s <a class="reference external" href="https://tigger.readthedocs.io/en/stable/">website</a> and <a class="reference external" href="https://pubmed.ncbi.nlm.nih.gov/25675496/">manuscript</a>, <em>‘TIgGER is a computational method that significantly improves V(D)J allele assignments by first determining the complete set of gene segments carried by an individual (including novel alleles) from V(D)J-rearrange sequences. TIgGER
can then infer a subject’s genotype from these sequences, and use this genotype to correct the initial V(D)J allele assignments.’</em></p>
<p>This impact’s on how contigs are chosen for finding clones later so it is highly recommended to run it. It is also important when considering to do mutational analysis.</p>
<p>However, the main caveat is that this needs to be run on multiple samples from the same subject to allow for more information to be used to confidently assign a genotyped <em>v_call</em>. In this tutorial, I’m assuming the four samples can be split into two sets where sets of two corresponds to a different/single individual. So while important, this step can be skipped if you don’t have the samples to do this.</p>
<p><code class="docutils literal notranslate"><span class="pre">pp.reassign_alleles</span></code> requires the <code class="docutils literal notranslate"><span class="pre">combined_folder</span></code> option to be specified so that a merged/concatenated file can be produced for running TIgGER. The function also runs <code class="docutils literal notranslate"><span class="pre">pp.create_germlines</span></code> using the germline database updated with the germline corrections from TIgGER. The default behavior is that it will return a <code class="docutils literal notranslate"><span class="pre">germline_alignment_d_mask</span></code> column in the final output. This can be changed by specifying <code class="docutils literal notranslate"><span class="pre">germ_types</span></code> option; see
<a class="reference external" href="https://changeo.readthedocs.io/en/stable/tools/CreateGermlines.html#creategermlines">here</a> for other options.</p>
<p>Specifying <code class="docutils literal notranslate"><span class="pre">fileformat</span> <span class="pre">=</span> <span class="pre">'changeo'</span></code> will run on changeo formatted files if this was run earlier; but it’s probably best to stick to airr’s standard format.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-r notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># reassigning alleles on the first set of samples</span>
<span class="n">ddl</span><span class="o">$</span><span class="n">pp</span><span class="o">$</span><span class="nf">reassign_alleles</span><span class="p">(</span><span class="n">samples</span><span class="p">[</span><span class="m">1</span><span class="o">:</span><span class="m">2</span><span class="p">],</span> <span class="n">combined_folder</span> <span class="o">=</span> <span class="s">&#39;tutorial_scgp1&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-r notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># reassigning alleles on the first set of samples</span>
<span class="n">ddl</span><span class="o">$</span><span class="n">pp</span><span class="o">$</span><span class="nf">reassign_alleles</span><span class="p">(</span><span class="n">samples</span><span class="p">[</span><span class="m">3</span><span class="o">:</span><span class="m">4</span><span class="p">],</span> <span class="n">combined_folder</span> <span class="o">=</span> <span class="s">&#39;tutorial_scgp2&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<p>We can see that most of the original ambiguous V calls have now been corrected and only a few remain. These will be flagged as multi later on and can probably be excluded from detailed analyses. For now, leaving them in the data will not impact on subsequent analyses.</p>
</div>
<div class="section" id="Step-4:">
<h2>Step 4:<a class="headerlink" href="#Step-4:" title="Permalink to this headline">¶</a></h2>
<div class="section" id="Assigning-constant-region-calls">
<h3>Assigning constant region calls<a class="headerlink" href="#Assigning-constant-region-calls" title="Permalink to this headline">¶</a></h3>
<p>10X’s annotation.csv files provides a <em>c_gene</em> column, but rather than simply relying on 10x’s annotation, <a class="reference external" href="https://twitter.com/hamish_king">hk6</a> recommended using <a class="reference external" href="https://presto.readthedocs.io/en/version-0.5.3---license-change/tools/MaskPrimers.html">immcantation-presto’s MaskPrimers.py</a> with his custom primer list and I tested that; worked well but it took <strong>20 min</strong> for the first file (~6k contigs). It also only calls the constant region for the heavy chains. The processing speed for
MaskPrimers can be sped up with using the filtered file.</p>
<p>Anyway, as an alternative, I wrote a pre-processing function, <code class="docutils literal notranslate"><span class="pre">ddl$pp$assign_isotypes</span></code>, to use <em>blast</em> to annotate constant region calls for all contigs and retrieves the call, merging it with the tsv files. This function will simply overwrite the output from previous steps and add a <em>c_call</em> column at the end, or replace the existing column if it already exists.</p>
<p>To deal with incorrect constant gene calls due to insufficient length, an internal subfunction will run a pairwise alignment against <a class="reference external" href="https://twitter.com/hamish_king">hk6</a>’s curated sequences that were deemed to be highly specific in distinguishing IGHA1-2, IGHG1-4. I have also curated sets of sequences that should help deal with <code class="docutils literal notranslate"><span class="pre">IGLC3/6/7</span></code> as these are problematic too. If there’s insufficient info, the <code class="docutils literal notranslate"><span class="pre">c_call</span></code> will be returned as a combination of the most aligned sets of sequences.
Because of how similar the light chains are, extremely ambiguous calls (only able to map to a common sequence across the light chains) will be returned as <code class="docutils literal notranslate"><span class="pre">IGLC</span></code>. This typically occurs when the constant sequence is very short. Those that have equal alignment scores between <code class="docutils literal notranslate"><span class="pre">IGLC3/6/7</span></code> sequences and the common sequence will be returned as a concatenated call; for example with a contig initially annotated as <code class="docutils literal notranslate"><span class="pre">IGLC3</span></code> will be returned as <code class="docutils literal notranslate"><span class="pre">IGLC,IGLC3</span></code>. If you do not want this subfunction to
run, toggle:</p>
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">correct_c_call</span> <span class="o">=</span> <span class="kc">FALSE</span>
</pre></div>
</div>
<hr class="docutils" />
<p>Before running, there is a need to set up a database with IMGT constant gene fasta sequences using <em>makeblastdb</em>, basically following the instructions from <a class="reference external" href="https://www.ncbi.nlm.nih.gov/books/NBK279688/">https://www.ncbi.nlm.nih.gov/books/NBK279688/</a>. <strong>This only needs to be done once.</strong></p>
<p>The fasta files were downloaded from IMGT and only sequences corresponding to <em>CH1</em> region for each constant gene/allele were retained. The headers were trimmed to only keep the gene and allele information. Links to find the sequences can be found here : <a class="reference external" href="http://www.imgt.org/genedb/GENElect?query=7.2+IGHC&amp;species=Homo+sapiens">human</a> and <a class="reference external" href="http://www.imgt.org/genedb/GENElect?query=7.2+IGHC&amp;species=Mus">mouse</a>.</p>
<p>The database file is provided in the repository and I’ve written a utility function <code class="docutils literal notranslate"><span class="pre">ddl$utl$makeblastdb</span></code> to prep new fasta files/databases if you need to.</p>
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">ddl</span><span class="o">$</span><span class="n">utl</span><span class="o">$</span><span class="nf">makeblastdb</span><span class="p">(</span><span class="s">&#39;/path/to/folder/containing/database/blast/human/human_BCR_C.fasta&#39;</span><span class="p">)</span>
</pre></div>
</div>
<p><strong>Again, we really only need to do it once</strong>; the file path can be added as an environmental variable after running:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span><span class="nb">echo</span> <span class="s2">&quot;export BLASTDB=/path/to/folder/containing/database/blast/&quot;</span> &gt;&gt; ~/.bash_profile
<span class="nb">source</span> ~/.bash_profile
</pre></div>
</div>
<p>I’ve set it up so that if the default option for <code class="docutils literal notranslate"><span class="pre">blastdb</span></code> is left as <code class="docutils literal notranslate"><span class="pre">None</span></code>, the function will retrieve a relative path from the environmental variable <code class="docutils literal notranslate"><span class="pre">$BLASTDB</span></code> and then, depending on which organism was specified (default = human), point to the correct fasta file. If you choose not to add it to environment, you can provide a string specifying a path to the fasta file for the <code class="docutils literal notranslate"><span class="pre">blastdb</span></code> option. The string has to point directly to the fasta file, i.e. end with <code class="docutils literal notranslate"><span class="pre">.fasta</span></code>.</p>
<hr class="docutils" />
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[6]:
</pre></div>
</div>
<div class="input_area highlight-r notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># ddl$utl$makeblastdb(&#39;/Users/kt16/Documents/Github/dandelion/database/blast/human/human_BCR_C.fasta&#39;)</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-r notranslate"><div class="highlight"><pre>
<span></span><span class="n">ddl</span><span class="o">$</span><span class="n">pp</span><span class="o">$</span><span class="nf">assign_isotypes</span><span class="p">(</span><span class="n">samples</span><span class="p">)</span>
</pre></div>
</div>
</div>
<p>This still takes a while when dealing with large files; the number of cpus to size of file isn’t exactly linear. Nevertheless, I have enabled parallelization as default because there were noticeable improvements in processing speeds with the smaller files. Maybe it will work better on a cluster with more cpus, rather than just a standard laptop. Other than a couple of samples that took about <strong>~10-40min</strong>, most ran within <strong>2-5min</strong>. I expect that this should run faster with filtered files too.</p>
<p>The default option will return a summary plot that can be disabled with <code class="docutils literal notranslate"><span class="pre">plot</span> <span class="pre">=</span> <span class="pre">FALSE</span></code>. In R, if you leave the <code class="docutils literal notranslate"><span class="pre">plot</span> <span class="pre">=</span> <span class="pre">TRUE</span></code>, it will wait for you to close the plots before continuing.</p>
<p>Also, the function can be run with <code class="docutils literal notranslate"><span class="pre">fileformat</span> <span class="pre">=</span> <span class="pre">'changeo'</span></code> if preferred.</p>
<p>It’s worthwhile to manually check the the sequences for constant calls returned as IGHA1-2, IGHG1-4 and the light chains and manually correct them if necessary.</p>
</div>
</div>
<div class="section" id="Step-5:-Quantify-mutations-(optional).">
<h2>Step 5: Quantify mutations <em>(optional)</em>.<a class="headerlink" href="#Step-5:-Quantify-mutations-(optional)." title="Permalink to this headline">¶</a></h2>
<p>In my original notebook, at this stage, I quantified the basic mutational load with <code class="docutils literal notranslate"><span class="pre">ddl.pp.quantify_mutations</span></code> before subsequent analyses. This would not run properly within this R session due to rpy2/reticulate conflict. Instead, i’d recommend you to run this separately on the. <code class="docutils literal notranslate"><span class="pre">*_igblast_gap_genotyped.tsv</span></code> file that was generated in the previous step, by following <a class="reference external" href="https://shazam.readthedocs.io/en/stable/vignettes/Mutation-Vignette/">Shazam’s tutorial</a> on basic mutational analysis,
before continuing to step 6.</p>
</div>
<div class="section" id="Step-6:-Marking-BCR-doublets-and-poor-quality-BCRs">
<h2>Step 6: Marking BCR doublets and poor quality BCRs<a class="headerlink" href="#Step-6:-Marking-BCR-doublets-and-poor-quality-BCRs" title="Permalink to this headline">¶</a></h2>
<div class="section" id="Create-a-Seurat-object-from-the-transcriptome-data">
<h3>Create a Seurat object from the transcriptome data<a class="headerlink" href="#Create-a-Seurat-object-from-the-transcriptome-data" title="Permalink to this headline">¶</a></h3>
<p>Let’s first import the gene expression data. Let’s try from Seurat object.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-r notranslate"><div class="highlight"><pre>
<span></span><span class="nf">setwd</span><span class="p">(</span><span class="s">&#39;/Users/kt16/Downloads/dandelion_tutorial_R/&#39;</span><span class="p">)</span>
<span class="nf">library</span><span class="p">(</span><span class="n">reticulate</span><span class="p">)</span>
<span class="n">ddl</span> <span class="o">=</span> <span class="nf">import_from_path</span><span class="p">(</span><span class="s">&#39;dandelion&#39;</span><span class="p">,</span> <span class="n">path</span> <span class="o">=</span> <span class="s">&#39;/Users/kt16/Documents/Github/dandelion&#39;</span><span class="p">)</span>
<span class="c1"># ddl = import(&#39;dandelion&#39;)</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-r notranslate"><div class="highlight"><pre>
<span></span><span class="nf">library</span><span class="p">(</span><span class="n">Seurat</span><span class="p">)</span>
<span class="n">samples</span> <span class="o">=</span> <span class="nf">c</span><span class="p">(</span><span class="s">&#39;sc5p_v2_hs_PBMC_1k&#39;</span><span class="p">,</span> <span class="s">&#39;sc5p_v2_hs_PBMC_10k&#39;</span><span class="p">,</span> <span class="s">&#39;vdj_v1_hs_pbmc3&#39;</span><span class="p">,</span> <span class="s">&#39;vdj_nextgem_hs_pbmc3&#39;</span><span class="p">)</span>
<span class="n">seurat_objects</span> <span class="o">=</span> <span class="nf">list</span><span class="p">()</span>
<span class="nf">for </span><span class="p">(</span><span class="n">i</span> <span class="n">in</span> <span class="m">1</span><span class="o">:</span><span class="nf">length</span><span class="p">(</span><span class="n">samples</span><span class="p">)){</span>
    <span class="n">filename</span> <span class="o">=</span> <span class="nf">paste0</span><span class="p">(</span><span class="n">samples</span><span class="p">[</span><span class="n">i</span><span class="p">],</span> <span class="s">&#39;/&#39;</span><span class="p">,</span> <span class="n">samples</span><span class="p">[</span><span class="n">i</span><span class="p">],</span> <span class="s">&quot;_filtered_feature_bc_matrix.h5&quot;</span><span class="p">)</span>
    <span class="n">data</span> <span class="o">&lt;-</span> <span class="nf">Read10X_h5</span><span class="p">(</span><span class="n">filename</span><span class="p">)</span>
    <span class="n">seurat_objects</span><span class="p">[[</span><span class="n">i</span><span class="p">]]</span> <span class="o">&lt;-</span> <span class="nf">CreateSeuratObject</span><span class="p">(</span><span class="n">counts</span> <span class="o">=</span> <span class="n">data</span><span class="o">$</span><span class="n">`Gene Expression`</span><span class="p">)</span>
<span class="p">}</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-r notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># merge them, there&#39;s probably better ways</span>
<span class="n">merged1</span> <span class="o">=</span> <span class="nf">merge</span><span class="p">(</span><span class="n">seurat_objects</span><span class="p">[[</span><span class="m">1</span><span class="p">]],</span> <span class="n">seurat_objects</span><span class="p">[[</span><span class="m">2</span><span class="p">]],</span> <span class="n">add.cell.ids</span> <span class="o">=</span> <span class="n">samples</span><span class="p">[</span><span class="m">1</span><span class="o">:</span><span class="m">2</span><span class="p">])</span>
<span class="n">merged2</span> <span class="o">=</span> <span class="nf">merge</span><span class="p">(</span><span class="n">seurat_objects</span><span class="p">[[</span><span class="m">3</span><span class="p">]],</span> <span class="n">seurat_objects</span><span class="p">[[</span><span class="m">4</span><span class="p">]],</span> <span class="n">add.cell.ids</span> <span class="o">=</span> <span class="n">samples</span><span class="p">[</span><span class="m">3</span><span class="o">:</span><span class="m">4</span><span class="p">])</span>
<span class="n">merged</span> <span class="o">=</span> <span class="nf">merge</span><span class="p">(</span><span class="n">merged1</span><span class="p">,</span> <span class="n">merged2</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-r notranslate"><div class="highlight"><pre>
<span></span><span class="nf">head</span><span class="p">(</span><span class="n">merged</span><span class="o">@</span><span class="n">meta.data</span><span class="p">)</span>
</pre></div>
</div>
</div>
<p><code class="docutils literal notranslate"><span class="pre">dandelion</span></code> removes hyphen from the cell barcodes as a default behaviour, so we will do the same for the <code class="docutils literal notranslate"><span class="pre">Seurat</span></code> object.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-r notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># remove the -# from the end of each cell name</span>
<span class="n">merged</span> <span class="o">&lt;-</span> <span class="nf">RenameCells</span><span class="p">(</span><span class="n">merged</span><span class="p">,</span> <span class="n">new.names</span> <span class="o">=</span> <span class="nf">gsub</span><span class="p">(</span><span class="s">&#39;-.*&#39;</span><span class="p">,</span> <span class="s">&#39;&#39;</span><span class="p">,</span> <span class="nf">row.names</span><span class="p">(</span><span class="n">merged</span><span class="o">@</span><span class="n">meta.data</span><span class="p">)))</span>
<span class="n">merged</span>
</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="Standard-Seurat-pre-processing-workflow">
<h2>Standard Seurat pre-processing workflow<a class="headerlink" href="#Standard-Seurat-pre-processing-workflow" title="Permalink to this headline">¶</a></h2>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-r notranslate"><div class="highlight"><pre>
<span></span><span class="n">merged</span><span class="p">[[</span><span class="s">&quot;percent.mt&quot;</span><span class="p">]]</span> <span class="o">&lt;-</span> <span class="nf">PercentageFeatureSet</span><span class="p">(</span><span class="n">merged</span><span class="p">,</span> <span class="n">pattern</span> <span class="o">=</span> <span class="s">&quot;^MT-&quot;</span><span class="p">)</span>
<span class="n">merged</span> <span class="o">&lt;-</span> <span class="nf">subset</span><span class="p">(</span><span class="n">merged</span><span class="p">,</span> <span class="n">subset</span> <span class="o">=</span> <span class="n">nFeature_RNA</span> <span class="o">&gt;</span> <span class="m">200</span> <span class="o">&amp;</span> <span class="n">nFeature_RNA</span> <span class="o">&lt;</span> <span class="m">2500</span> <span class="o">&amp;</span> <span class="n">percent.mt</span> <span class="o">&lt;</span> <span class="m">5</span><span class="p">)</span>
<span class="n">merged</span> <span class="o">&lt;-</span> <span class="nf">NormalizeData</span><span class="p">(</span><span class="n">merged</span><span class="p">)</span>
<span class="n">merged</span> <span class="o">&lt;-</span> <span class="nf">FindVariableFeatures</span><span class="p">(</span><span class="n">merged</span><span class="p">,</span> <span class="n">selection.method</span> <span class="o">=</span> <span class="s">&quot;vst&quot;</span><span class="p">,</span> <span class="n">nfeatures</span> <span class="o">=</span> <span class="m">2000</span><span class="p">)</span>
<span class="n">merged</span> <span class="o">&lt;-</span> <span class="nf">ScaleData</span><span class="p">(</span><span class="n">merged</span><span class="p">)</span>
<span class="n">merged</span> <span class="o">&lt;-</span> <span class="nf">RunPCA</span><span class="p">(</span><span class="n">merged</span><span class="p">,</span> <span class="n">features</span> <span class="o">=</span> <span class="nf">VariableFeatures</span><span class="p">(</span><span class="n">object</span> <span class="o">=</span> <span class="n">merged</span><span class="p">))</span>
<span class="n">merged</span> <span class="o">&lt;-</span> <span class="nf">FindNeighbors</span><span class="p">(</span><span class="n">merged</span><span class="p">,</span> <span class="n">dims</span> <span class="o">=</span> <span class="m">1</span><span class="o">:</span><span class="m">50</span><span class="p">)</span>
<span class="n">merged</span> <span class="o">&lt;-</span> <span class="nf">FindClusters</span><span class="p">(</span><span class="n">merged</span><span class="p">)</span>
<span class="n">merged</span> <span class="o">&lt;-</span> <span class="nf">RunUMAP</span><span class="p">(</span><span class="n">merged</span><span class="p">,</span> <span class="n">dims</span> <span class="o">=</span> <span class="m">1</span><span class="o">:</span><span class="m">50</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-r notranslate"><div class="highlight"><pre>
<span></span><span class="nf">DimPlot</span><span class="p">(</span><span class="n">merged</span><span class="p">,</span> <span class="n">reduction</span> <span class="o">=</span> <span class="s">&quot;umap&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-r notranslate"><div class="highlight"><pre>
<span></span><span class="n">merged</span>
</pre></div>
</div>
</div>
<p>In order for <code class="docutils literal notranslate"><span class="pre">dandelion</span></code> to do the next step (marking/filtering of poor quality BCRs and BCR doublets), we need to include a column in the metadata, <code class="docutils literal notranslate"><span class="pre">filter_rna</span></code>. This column will tell <code class="docutils literal notranslate"><span class="pre">dandelion</span></code> whether or not the cell has passed transcriptomic QCs. So, we will set the column to <code class="docutils literal notranslate"><span class="pre">FALSE</span></code> i.e. every cell passed QC. This is important because we want to remove as many doublets and poor quality contigs to save time on computation and construction of the final data tables.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-r notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># create a column called filter_rna and set it to FALSE</span>
<span class="n">merged</span><span class="o">@</span><span class="n">meta.data</span><span class="o">$</span><span class="n">filter_rna</span> <span class="o">=</span> <span class="kc">FALSE</span>
</pre></div>
</div>
</div>
</div>
<div class="section" id="Convert-to-AnnData">
<h2>Convert to AnnData<a class="headerlink" href="#Convert-to-AnnData" title="Permalink to this headline">¶</a></h2>
<p>We will need to convert the <code class="docutils literal notranslate"><span class="pre">Seurat</span></code> object to <code class="docutils literal notranslate"><span class="pre">AnnData</span></code> to be able to continue. <code class="docutils literal notranslate"><span class="pre">AnnData</span></code> <code class="docutils literal notranslate"><span class="pre">.obs</span></code> slot is essentially the same as <code class="docutils literal notranslate"><span class="pre">&#64;meta.data</span></code> in seurat.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-r notranslate"><div class="highlight"><pre>
<span></span><span class="n">sc</span> <span class="o">=</span> <span class="nf">import</span><span class="p">(</span><span class="s">&quot;scanpy&quot;</span><span class="p">)</span>
<span class="c1"># convert the meta.data slot to a python friendly object</span>
<span class="n">obs</span> <span class="o">=</span> <span class="nf">r_to_py</span><span class="p">(</span><span class="n">merged</span><span class="o">@</span><span class="n">meta.data</span><span class="p">)</span>
<span class="n">normcounts</span> <span class="o">=</span> <span class="nf">r_to_py</span><span class="p">(</span><span class="n">Matrix</span><span class="o">::</span><span class="nf">t</span><span class="p">(</span><span class="nf">GetAssayData</span><span class="p">(</span><span class="n">merged</span><span class="p">)))</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-r notranslate"><div class="highlight"><pre>
<span></span><span class="n">adata</span> <span class="o">=</span> <span class="n">sc</span><span class="o">$</span><span class="nf">AnnData</span><span class="p">(</span><span class="n">X</span> <span class="o">=</span> <span class="n">normcounts</span><span class="p">,</span> <span class="n">obs</span> <span class="o">=</span> <span class="n">obs</span><span class="p">)</span>
<span class="n">adata</span>
</pre></div>
</div>
</div>
<p>We need to populate the <code class="docutils literal notranslate"><span class="pre">.neighbors</span></code> slots via scanpy smooth transfer later</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-r notranslate"><div class="highlight"><pre>
<span></span><span class="n">sc</span><span class="o">$</span><span class="n">pp</span><span class="o">$</span><span class="nf">neighbors</span><span class="p">(</span><span class="n">adata</span><span class="p">)</span>
<span class="n">adata</span>
</pre></div>
</div>
</div>
</div>
<div class="section" id="Read-in-the-BCR-files-and-merge-them">
<h2>Read in the BCR files and merge them<a class="headerlink" href="#Read-in-the-BCR-files-and-merge-them" title="Permalink to this headline">¶</a></h2>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-r notranslate"><div class="highlight"><pre>
<span></span><span class="n">files</span> <span class="o">=</span> <span class="nf">list</span><span class="p">()</span>
<span class="nf">for </span><span class="p">(</span><span class="n">i</span> <span class="n">in</span> <span class="m">1</span><span class="o">:</span><span class="nf">length</span><span class="p">(</span><span class="n">samples</span><span class="p">)){</span>
    <span class="n">filename</span> <span class="o">=</span> <span class="nf">paste0</span><span class="p">(</span><span class="n">samples</span><span class="p">[</span><span class="n">i</span><span class="p">],</span> <span class="s">&#39;/dandelion/data/&#39;</span><span class="p">,</span> <span class="n">samples</span><span class="p">[</span><span class="n">i</span><span class="p">],</span> <span class="s">&quot;_b_filtered_contig_igblast_gap_genotyped.tsv&quot;</span><span class="p">)</span>
    <span class="n">files</span><span class="p">[[</span><span class="n">i</span><span class="p">]]</span> <span class="o">&lt;-</span> <span class="n">readr</span><span class="o">::</span><span class="nf">read_tsv</span><span class="p">(</span><span class="n">filename</span><span class="p">)</span>
<span class="p">}</span>
<span class="n">combined_bcr</span> <span class="o">=</span> <span class="nf">do.call</span><span class="p">(</span><span class="n">rbind</span><span class="p">,</span> <span class="n">files</span><span class="p">)</span>
<span class="nf">head</span><span class="p">(</span><span class="n">combined_bcr</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-r notranslate"><div class="highlight"><pre>
<span></span><span class="nf">head</span><span class="p">(</span><span class="n">combined_bcr</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="section" id="Step-7:-Run-ddl$pp$filter_bcr">
<h2>Step 7: Run <code class="docutils literal notranslate"><span class="pre">ddl$pp$filter_bcr</span></code><a class="headerlink" href="#Step-7:-Run-ddl$pp$filter_bcr" title="Permalink to this headline">¶</a></h2>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-r notranslate"><div class="highlight"><pre>
<span></span><span class="n">vdj_results_list</span> <span class="o">=</span> <span class="n">ddl</span><span class="o">$</span><span class="n">pp</span><span class="o">$</span><span class="nf">filter_bcr</span><span class="p">(</span><span class="n">combined_bcr</span><span class="p">,</span> <span class="n">adata</span><span class="p">)</span>
<span class="n">vdj_results_list</span>
</pre></div>
</div>
</div>
<p>This returns a two level list in R. The first level is the VDJ results, stored as a <code class="docutils literal notranslate"><span class="pre">Dandelion</span></code> python-class object and the second level is the accompanying <code class="docutils literal notranslate"><span class="pre">AnnData</span></code> object.</p>
<p>The <code class="docutils literal notranslate"><span class="pre">Dandelion</span></code> class is structured like a multi-slot object and the two data frames below are linked: 1) data &lt;- BCR table with row names as individual vdj contigs</p>
<ol class="arabic simple" start="2">
<li><p>metadata &lt;- BCR table collapsed to cell barcodes as row names</p></li>
</ol>
<p>More details on the <code class="docutils literal notranslate"><span class="pre">Dandelion</span></code> class are in my python notebooks. The most important slot for now, is the <code class="docutils literal notranslate"><span class="pre">metadata</span></code> slot within <code class="docutils literal notranslate"><span class="pre">Dandelion</span></code>.</p>
<p>In order for the <code class="docutils literal notranslate"><span class="pre">metadata</span></code> to form properly, there must not be any duplicate barcodes, or incorrectly retrieved information from the <code class="docutils literal notranslate"><span class="pre">data</span></code> slot. If you end up with a <code class="docutils literal notranslate"><span class="pre">Dandelion</span></code> object that only contains the <code class="docutils literal notranslate"><span class="pre">data</span></code> slot filled, it means one of the two conditions happened. In those situations, I would recommend you to send me a copy of the file so I can check why it’s failing; it is usually due to coding eror that arise from string and float incompatibilities when constructing the
object.</p>
<p>To save the <code class="docutils literal notranslate"><span class="pre">Dandelion</span></code> object, you can do the following:</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-r notranslate"><div class="highlight"><pre>
<span></span><span class="n">vdj_results_list</span><span class="p">[[</span><span class="m">1</span><span class="p">]]</span><span class="o">$</span><span class="nf">write</span><span class="p">(</span><span class="s">&#39;vdj_save.pkl.pbz2&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<p>The <code class="docutils literal notranslate"><span class="pre">.pkl.pbz2</span></code> extension is basically a bzip2-compressed pickle file format from python.</p>
<p>To read the file back into R, you can do the following:</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-r notranslate"><div class="highlight"><pre>
<span></span><span class="n">vdj_data</span> <span class="o">&lt;-</span> <span class="n">ddl</span><span class="o">$</span><span class="nf">read</span><span class="p">(</span><span class="s">&#39;vdj_save.pkl.pbz2&#39;</span><span class="p">)</span>
<span class="n">vdj_data</span>
</pre></div>
</div>
</div>
</div>
<div class="section" id="Step-8-:-Finding-clones">
<h2>Step 8 : Finding clones<a class="headerlink" href="#Step-8-:-Finding-clones" title="Permalink to this headline">¶</a></h2>
<p><code class="docutils literal notranslate"><span class="pre">Dandelion</span></code> comes with a method to define clones based on V-J gene usuage and CDR3 junction similarity but you can always run Immcantation/Change-O’s DefineClones with the filtered file from earlier using their <a class="reference external" href="https://changeo.readthedocs.io/en/stable/examples/10x.html">tutorial</a>. To use dandelion’s you just need to do the following:</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-r notranslate"><div class="highlight"><pre>
<span></span><span class="n">ddl</span><span class="o">$</span><span class="n">tl</span><span class="o">$</span><span class="nf">find_clones</span><span class="p">(</span><span class="n">vdj_data</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="section" id="Calculating-size-of-clones">
<h2>Calculating size of clones<a class="headerlink" href="#Calculating-size-of-clones" title="Permalink to this headline">¶</a></h2>
<p>Sometimes it’s useful to evaluate the size of the clone. Here <code class="docutils literal notranslate"><span class="pre">ddl$tl$clone_size</span></code> does a simple calculation to enable that.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-r notranslate"><div class="highlight"><pre>
<span></span><span class="n">ddl</span><span class="o">$</span><span class="n">tl</span><span class="o">$</span><span class="nf">clone_size</span><span class="p">(</span><span class="n">vdj_data</span><span class="p">)</span>
</pre></div>
</div>
</div>
<p>You can also specify <code class="docutils literal notranslate"><span class="pre">max_size</span></code> to clip off the calculation at a fixed value.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-r notranslate"><div class="highlight"><pre>
<span></span><span class="n">ddl</span><span class="o">$</span><span class="n">tl</span><span class="o">$</span><span class="nf">clone_size</span><span class="p">(</span><span class="n">vdj_data</span><span class="p">,</span> <span class="n">max_size</span> <span class="o">=</span> <span class="m">3</span><span class="p">)</span>
</pre></div>
</div>
</div>
<p>I have blitz through the last 3 functions without showing you the output but don’t worry, they are all stashed in the <code class="docutils literal notranslate"><span class="pre">Dandelion</span></code> object.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-r notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># compare the column names in the metadata slot of the object below with the one above.</span>
<span class="n">vdj_data</span>
</pre></div>
</div>
</div>
</div>
<div class="section" id="Step-9-:-Generate-BCR-network-visualization">
<h2>Step 9 : Generate BCR network visualization<a class="headerlink" href="#Step-9-:-Generate-BCR-network-visualization" title="Permalink to this headline">¶</a></h2>
<p>The name of <code class="docutils literal notranslate"><span class="pre">Dandelion</span></code> came from the way it visualizes the BCR as networks, that look like dandelions in the field. We need to first generate the network. We will hopewfully visualize in Seurat later.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-r notranslate"><div class="highlight"><pre>
<span></span><span class="n">ddl</span><span class="o">$</span><span class="n">tl</span><span class="o">$</span><span class="nf">generate_network</span><span class="p">(</span><span class="n">vdj_data</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="section" id="Step-10-:-Integrating-with-Seurat">
<h2>Step 10 : Integrating with Seurat<a class="headerlink" href="#Step-10-:-Integrating-with-Seurat" title="Permalink to this headline">¶</a></h2>
<p>At this point, you might want to transfer the metadata slot back to Seurat so you can visualise some things. You can do that column by column directly from <code class="docutils literal notranslate"><span class="pre">Dandelion</span></code> object like as follows:</p>
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">isotype</span> <span class="o">=</span> <span class="nf">unlist</span><span class="p">(</span><span class="n">vdj_data</span><span class="o">$</span><span class="n">metadata</span><span class="o">$</span><span class="n">isotype</span><span class="p">)</span> <span class="c1"># because of the python to R conversion, it thinks it&#39;s a list rather than a vector. we can correct this with unlist</span>
<span class="nf">names</span><span class="p">(</span><span class="n">isotype</span><span class="p">)</span> <span class="o">&lt;-</span> <span class="nf">row.names</span><span class="p">(</span><span class="n">vdj_data</span><span class="o">$</span><span class="n">metadata</span><span class="p">)</span>
<span class="n">merged</span> <span class="o">&lt;-</span> <span class="nf">AddMetaData</span><span class="p">(</span><span class="n">merged</span><span class="p">,</span> <span class="n">isotype</span><span class="p">,</span> <span class="s">&#39;isotype&#39;</span><span class="p">)</span>
<span class="nf">DimPlot</span><span class="p">(</span><span class="n">merged</span><span class="p">,</span> <span class="n">group.by</span> <span class="o">=</span> <span class="s">&#39;isotype&#39;</span><span class="p">)</span>
</pre></div>
</div>
<p>I will show you how to do via the <code class="docutils literal notranslate"><span class="pre">AnnData</span></code> object.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-r notranslate"><div class="highlight"><pre>
<span></span><span class="n">adata2</span> <span class="o">=</span> <span class="n">vdj_results_list</span><span class="p">[[</span><span class="m">2</span><span class="p">]]</span>
<span class="n">adata2</span>
</pre></div>
</div>
</div>
<div class="section" id="Transfer-Dandelion-to-AnnData">
<h3>Transfer <code class="docutils literal notranslate"><span class="pre">Dandelion</span></code> to <code class="docutils literal notranslate"><span class="pre">AnnData</span></code><a class="headerlink" href="#Transfer-Dandelion-to-AnnData" title="Permalink to this headline">¶</a></h3>
<p><code class="docutils literal notranslate"><span class="pre">ddl$tl$transfer</span></code> will act to transfer the metadata and graph slots from <code class="docutils literal notranslate"><span class="pre">Dandelion</span></code> object to <code class="docutils literal notranslate"><span class="pre">AnnData</span></code>.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-r notranslate"><div class="highlight"><pre>
<span></span><span class="n">ddl</span><span class="o">$</span><span class="n">tl</span><span class="o">$</span><span class="nf">transfer</span><span class="p">(</span><span class="n">adata2</span><span class="p">,</span> <span class="n">vdj_data</span><span class="p">,</span> <span class="n">full_graph</span> <span class="o">=</span> <span class="kc">TRUE</span><span class="p">)</span> <span class="c1"># switch full_graph to FALSE if you only want to get the coordinates for expanded clones</span>
</pre></div>
</div>
</div>
<p>This will populate the adata <code class="docutils literal notranslate"><span class="pre">.obs</span></code> slot with the metadata from the <code class="docutils literal notranslate"><span class="pre">Dandelion</span></code> object.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-r notranslate"><div class="highlight"><pre>
<span></span><span class="n">adata2</span>
</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="Save">
<h2>Save<a class="headerlink" href="#Save" title="Permalink to this headline">¶</a></h2>
<p>There may be some issues with conversions between python and R and vice versa. So my recommendation at this stage is to save the three objects separately and load them up in a fresh session. There’s a high chance your session will crash if you ignore this.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-r notranslate"><div class="highlight"><pre>
<span></span><span class="n">adata2</span><span class="o">$</span><span class="nf">write</span><span class="p">(</span><span class="s">&#39;adata_test.h5ad&#39;</span><span class="p">,</span> <span class="n">compression</span> <span class="o">=</span> <span class="s">&#39;gzip&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-r notranslate"><div class="highlight"><pre>
<span></span><span class="nf">saveRDS</span><span class="p">(</span><span class="n">merged</span><span class="p">,</span> <span class="s">&#39;merged.RDS&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-r notranslate"><div class="highlight"><pre>
<span></span><span class="n">vdj_data</span><span class="o">$</span><span class="nf">write</span><span class="p">(</span><span class="s">&#39;vdj_save.pkl.pbz2&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="section" id="New-Session:-Transfer-AnnData-to-Seurat">
<h2>New Session: Transfer <code class="docutils literal notranslate"><span class="pre">AnnData</span></code> to <code class="docutils literal notranslate"><span class="pre">Seurat</span></code><a class="headerlink" href="#New-Session:-Transfer-AnnData-to-Seurat" title="Permalink to this headline">¶</a></h2>
<p>We start a new session and read in the files.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[1]:
</pre></div>
</div>
<div class="input_area highlight-r notranslate"><div class="highlight"><pre>
<span></span><span class="nf">setwd</span><span class="p">(</span><span class="s">&#39;/Users/kt16/Downloads/dandelion_tutorial_R/&#39;</span><span class="p">)</span>
<span class="nf">library</span><span class="p">(</span><span class="n">reticulate</span><span class="p">)</span>
<span class="n">ddl</span> <span class="o">=</span> <span class="nf">import_from_path</span><span class="p">(</span><span class="s">&#39;dandelion&#39;</span><span class="p">,</span> <span class="n">path</span> <span class="o">=</span> <span class="s">&#39;/Users/kt16/Documents/Github/dandelion&#39;</span><span class="p">)</span>
<span class="c1"># ddl = import(&#39;dandelion&#39;)</span>
<span class="n">sc</span> <span class="o">=</span> <span class="nf">import</span><span class="p">(</span><span class="s">&#39;scanpy&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[2]:
</pre></div>
</div>
<div class="input_area highlight-r notranslate"><div class="highlight"><pre>
<span></span><span class="nf">library</span><span class="p">(</span><span class="n">Seurat</span><span class="p">)</span>
<span class="n">samples</span> <span class="o">=</span> <span class="nf">c</span><span class="p">(</span><span class="s">&#39;sc5p_v2_hs_PBMC_1k&#39;</span><span class="p">,</span> <span class="s">&#39;sc5p_v2_hs_PBMC_10k&#39;</span><span class="p">,</span> <span class="s">&#39;vdj_v1_hs_pbmc3&#39;</span><span class="p">,</span> <span class="s">&#39;vdj_nextgem_hs_pbmc3&#39;</span><span class="p">)</span>
<span class="n">adata</span> <span class="o">=</span> <span class="n">sc</span><span class="o">$</span><span class="nf">read_h5ad</span><span class="p">(</span><span class="s">&#39;adata_test.h5ad&#39;</span><span class="p">)</span>
<span class="c1"># vdj = ddl$read(&#39;vdj_save.pkl.pbz2&#39;) # don&#39;t need this at this stage</span>
<span class="n">merged</span> <span class="o">=</span> <span class="nf">readRDS</span><span class="p">(</span><span class="s">&#39;merged.RDS&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
Warning message:
“package ‘Seurat’ was built under R version 3.6.2”
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[3]:
</pre></div>
</div>
<div class="input_area highlight-r notranslate"><div class="highlight"><pre>
<span></span><span class="n">adata</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
AnnData object with n_obs × n_vars = 18413 × 38224
    obs: &#39;orig.ident&#39;, &#39;nCount_RNA&#39;, &#39;nFeature_RNA&#39;, &#39;percent.mt&#39;, &#39;RNA_snn_res.0.8&#39;, &#39;seurat_clusters&#39;, &#39;filter_rna&#39;, &#39;filter_bcr_quality&#39;, &#39;filter_bcr_heavy&#39;, &#39;filter_bcr_light&#39;, &#39;filter_bcr&#39;, &#39;has_bcr&#39;, &#39;sample_id&#39;, &#39;clone_id&#39;, &#39;clone_id_group&#39;, &#39;clone_id_by_size&#39;, &#39;isotype&#39;, &#39;lightchain&#39;, &#39;status&#39;, &#39;vdj_status&#39;, &#39;productive&#39;, &#39;umi_counts_heavy&#39;, &#39;umi_counts_light&#39;, &#39;v_call_heavy&#39;, &#39;v_call_light&#39;, &#39;j_call_heavy&#39;, &#39;j_call_light&#39;, &#39;c_call_heavy&#39;, &#39;c_call_light&#39;, &#39;clone_id_size&#39;, &#39;clone_id_group_size&#39;
    uns: &#39;neighbors&#39;, &#39;rna_neighbors&#39;
    obsm: &#39;X_bcr&#39;, &#39;X_pca&#39;
    obsp: &#39;bcr_connectivities&#39;, &#39;bcr_distances&#39;, &#39;connectivities&#39;, &#39;distances&#39;, &#39;rna_connectivities&#39;, &#39;rna_distances&#39;
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[4]:
</pre></div>
</div>
<div class="input_area highlight-r notranslate"><div class="highlight"><pre>
<span></span><span class="n">merged</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
An object of class Seurat
38224 features across 18634 samples within 1 assay
Active assay: RNA (38224 features, 2000 variable features)
 2 dimensional reductions calculated: pca, umap
</pre></div></div>
</div>
<p>So there are a few cells missing from the <code class="docutils literal notranslate"><span class="pre">AnnData</span></code> object because they were filtered away. Let’s do a simple merge to populate the <code class="docutils literal notranslate"><span class="pre">Seurat</span></code> object’s <code class="docutils literal notranslate"><span class="pre">meta.data</span></code> slot.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[5]:
</pre></div>
</div>
<div class="input_area highlight-r notranslate"><div class="highlight"><pre>
<span></span><span class="n">merged_meta</span> <span class="o">=</span> <span class="n">merged</span><span class="o">@</span><span class="n">meta.data</span>
<span class="nf">head</span><span class="p">(</span><span class="n">merged_meta</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area rendered_html docutils container">
<table>
<caption>A data.frame: 6 × 7</caption>
<thead>
    <tr><th></th><th scope=col>orig.ident</th><th scope=col>nCount_RNA</th><th scope=col>nFeature_RNA</th><th scope=col>percent.mt</th><th scope=col>RNA_snn_res.0.8</th><th scope=col>seurat_clusters</th><th scope=col>filter_rna</th></tr>
    <tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;lgl&gt;</th></tr>
</thead>
<tbody>
    <tr><th scope=row>sc5p_v2_hs_PBMC_1k_AAACCTGCAGCCTGTG</th><td>SeuratProject</td><td>5029</td><td>1819</td><td>3.579240</td><td>19</td><td>19</td><td>FALSE</td></tr>
    <tr><th scope=row>sc5p_v2_hs_PBMC_1k_AAACGGGTCGCTGATA</th><td>SeuratProject</td><td>4343</td><td>1738</td><td>2.440709</td><td>5 </td><td>5 </td><td>FALSE</td></tr>
    <tr><th scope=row>sc5p_v2_hs_PBMC_1k_AAAGCAAAGTATGACA</th><td>SeuratProject</td><td>4999</td><td>1796</td><td>2.880576</td><td>5 </td><td>5 </td><td>FALSE</td></tr>
    <tr><th scope=row>sc5p_v2_hs_PBMC_1k_AAATGCCCACTTAAGC</th><td>SeuratProject</td><td>4998</td><td>1530</td><td>3.281313</td><td>2 </td><td>2 </td><td>FALSE</td></tr>
    <tr><th scope=row>sc5p_v2_hs_PBMC_1k_AACACGTCAACAACCT</th><td>SeuratProject</td><td>4818</td><td>1579</td><td>2.283105</td><td>2 </td><td>2 </td><td>FALSE</td></tr>
    <tr><th scope=row>sc5p_v2_hs_PBMC_1k_AACACGTGTTATCCGA</th><td>SeuratProject</td><td>7564</td><td>2220</td><td>2.525119</td><td>5 </td><td>5 </td><td>FALSE</td></tr>
</tbody>
</table></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[6]:
</pre></div>
</div>
<div class="input_area highlight-r notranslate"><div class="highlight"><pre>
<span></span><span class="n">adata_meta</span> <span class="o">=</span> <span class="n">adata</span><span class="o">$</span><span class="n">obs</span>
<span class="nf">head</span><span class="p">(</span><span class="n">adata_meta</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area rendered_html docutils container">
<table>
<caption>A data.frame: 6 × 31</caption>
<thead>
    <tr><th></th><th scope=col>orig.ident</th><th scope=col>nCount_RNA</th><th scope=col>nFeature_RNA</th><th scope=col>percent.mt</th><th scope=col>RNA_snn_res.0.8</th><th scope=col>seurat_clusters</th><th scope=col>filter_rna</th><th scope=col>filter_bcr_quality</th><th scope=col>filter_bcr_heavy</th><th scope=col>filter_bcr_light</th><th scope=col>⋯</th><th scope=col>umi_counts_heavy</th><th scope=col>umi_counts_light</th><th scope=col>v_call_heavy</th><th scope=col>v_call_light</th><th scope=col>j_call_heavy</th><th scope=col>j_call_light</th><th scope=col>c_call_heavy</th><th scope=col>c_call_light</th><th scope=col>clone_id_size</th><th scope=col>clone_id_group_size</th></tr>
    <tr><th></th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>⋯</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;fct&gt;</th></tr>
</thead>
<tbody>
    <tr><th scope=row>sc5p_v2_hs_PBMC_1k_AAACCTGCAGCCTGTG</th><td>SeuratProject</td><td>5029</td><td>1819</td><td>3.579240</td><td>19</td><td>19</td><td>FALSE</td><td>False</td><td>False</td><td>False</td><td>⋯</td><td>NaN</td><td>nan</td><td>nan</td><td>nan</td><td>nan</td><td>nan</td><td>nan</td><td>nan</td><td>nan</td><td>nan</td></tr>
    <tr><th scope=row>sc5p_v2_hs_PBMC_1k_AAACGGGTCGCTGATA</th><td>SeuratProject</td><td>4343</td><td>1738</td><td>2.440709</td><td>5 </td><td>5 </td><td>FALSE</td><td>False</td><td>False</td><td>False</td><td>⋯</td><td>NaN</td><td>nan</td><td>nan</td><td>nan</td><td>nan</td><td>nan</td><td>nan</td><td>nan</td><td>nan</td><td>nan</td></tr>
    <tr><th scope=row>sc5p_v2_hs_PBMC_1k_AAAGCAAAGTATGACA</th><td>SeuratProject</td><td>4999</td><td>1796</td><td>2.880576</td><td>5 </td><td>5 </td><td>FALSE</td><td>False</td><td>False</td><td>False</td><td>⋯</td><td>NaN</td><td>nan</td><td>nan</td><td>nan</td><td>nan</td><td>nan</td><td>nan</td><td>nan</td><td>nan</td><td>nan</td></tr>
    <tr><th scope=row>sc5p_v2_hs_PBMC_1k_AAATGCCCACTTAAGC</th><td>SeuratProject</td><td>4998</td><td>1530</td><td>3.281313</td><td>2 </td><td>2 </td><td>FALSE</td><td>False</td><td>False</td><td>False</td><td>⋯</td><td>NaN</td><td>nan</td><td>nan</td><td>nan</td><td>nan</td><td>nan</td><td>nan</td><td>nan</td><td>nan</td><td>nan</td></tr>
    <tr><th scope=row>sc5p_v2_hs_PBMC_1k_AACACGTCAACAACCT</th><td>SeuratProject</td><td>4818</td><td>1579</td><td>2.283105</td><td>2 </td><td>2 </td><td>FALSE</td><td>False</td><td>False</td><td>False</td><td>⋯</td><td>NaN</td><td>nan</td><td>nan</td><td>nan</td><td>nan</td><td>nan</td><td>nan</td><td>nan</td><td>nan</td><td>nan</td></tr>
    <tr><th scope=row>sc5p_v2_hs_PBMC_1k_AACACGTGTTATCCGA</th><td>SeuratProject</td><td>7564</td><td>2220</td><td>2.525119</td><td>5 </td><td>5 </td><td>FALSE</td><td>False</td><td>False</td><td>False</td><td>⋯</td><td>NaN</td><td>nan</td><td>nan</td><td>nan</td><td>nan</td><td>nan</td><td>nan</td><td>nan</td><td>nan</td><td>nan</td></tr>
</tbody>
</table></div>
</div>
<p>If you run into issues with the conversion, unfortunately there’s not much I can do about it. One alternative is to just transfer the <code class="docutils literal notranslate"><span class="pre">.obs</span></code> slot from the dandelion object one by one as above into the seurat object.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[7]:
</pre></div>
</div>
<div class="input_area highlight-r notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># merge into a data frame and then make sure the format is correct</span>
<span class="n">merged_data</span> <span class="o">&lt;-</span> <span class="nf">as.data.frame</span><span class="p">(</span><span class="nf">merge</span><span class="p">(</span><span class="n">merged_meta</span><span class="p">,</span> <span class="n">adata_meta</span><span class="p">,</span> <span class="n">by</span><span class="o">=</span><span class="m">0</span><span class="p">,</span> <span class="n">all</span><span class="o">=</span><span class="kc">TRUE</span><span class="p">))</span>
<span class="nf">rownames</span><span class="p">(</span><span class="n">merged_data</span><span class="p">)</span> <span class="o">&lt;-</span> <span class="n">merged_data</span><span class="p">[,</span><span class="m">1</span><span class="p">]</span>
<span class="n">merged_data</span><span class="p">[,</span><span class="m">1</span><span class="p">]</span> <span class="o">&lt;-</span> <span class="kc">NULL</span>
<span class="n">merged_data</span> <span class="o">&lt;-</span> <span class="nf">as.data.frame</span><span class="p">(</span><span class="nf">apply</span><span class="p">(</span><span class="n">merged_data</span><span class="p">,</span> <span class="m">2</span><span class="p">,</span> <span class="nf">function</span><span class="p">(</span><span class="n">x</span><span class="p">)</span> <span class="n">plyr</span><span class="o">::</span><span class="nf">revalue</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="nf">c</span><span class="p">(</span><span class="s">&quot;NaN&quot;</span><span class="o">=</span><span class="kc">NA</span><span class="p">,</span> <span class="s">&#39;nan&#39;</span> <span class="o">=</span> <span class="kc">NA</span><span class="p">))))</span> <span class="c1"># just replace the NAs manually</span>
<span class="nf">head</span><span class="p">(</span><span class="n">merged_data</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
The following `from` values were not present in `x`: NaN, nan

The following `from` values were not present in `x`: NaN, nan

The following `from` values were not present in `x`: NaN, nan

The following `from` values were not present in `x`: NaN, nan

The following `from` values were not present in `x`: NaN, nan

The following `from` values were not present in `x`: NaN, nan

The following `from` values were not present in `x`: NaN, nan

The following `from` values were not present in `x`: NaN, nan

The following `from` values were not present in `x`: NaN, nan

The following `from` values were not present in `x`: NaN, nan

The following `from` values were not present in `x`: NaN, nan

The following `from` values were not present in `x`: NaN, nan

The following `from` values were not present in `x`: NaN, nan

The following `from` values were not present in `x`: NaN, nan

The following `from` values were not present in `x`: NaN, nan

The following `from` values were not present in `x`: NaN, nan

The following `from` values were not present in `x`: NaN, nan

The following `from` values were not present in `x`: NaN, nan

The following `from` values were not present in `x`: NaN, nan

The following `from` values were not present in `x`: NaN

The following `from` values were not present in `x`: NaN

The following `from` values were not present in `x`: NaN

The following `from` values were not present in `x`: NaN, nan

The following `from` values were not present in `x`: NaN

The following `from` values were not present in `x`: NaN

The following `from` values were not present in `x`: NaN

The following `from` values were not present in `x`: NaN

The following `from` values were not present in `x`: NaN

The following `from` values were not present in `x`: NaN, nan

The following `from` values were not present in `x`: NaN

The following `from` values were not present in `x`: NaN

The following `from` values were not present in `x`: NaN

The following `from` values were not present in `x`: NaN

The following `from` values were not present in `x`: NaN

The following `from` values were not present in `x`: NaN

The following `from` values were not present in `x`: NaN

The following `from` values were not present in `x`: NaN

The following `from` values were not present in `x`: NaN

</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area rendered_html docutils container">
<table>
<caption>A data.frame: 6 × 38</caption>
<thead>
    <tr><th></th><th scope=col>orig.ident.x</th><th scope=col>nCount_RNA.x</th><th scope=col>nFeature_RNA.x</th><th scope=col>percent.mt.x</th><th scope=col>RNA_snn_res.0.8.x</th><th scope=col>seurat_clusters.x</th><th scope=col>filter_rna.x</th><th scope=col>orig.ident.y</th><th scope=col>nCount_RNA.y</th><th scope=col>nFeature_RNA.y</th><th scope=col>⋯</th><th scope=col>umi_counts_heavy</th><th scope=col>umi_counts_light</th><th scope=col>v_call_heavy</th><th scope=col>v_call_light</th><th scope=col>j_call_heavy</th><th scope=col>j_call_light</th><th scope=col>c_call_heavy</th><th scope=col>c_call_light</th><th scope=col>clone_id_size</th><th scope=col>clone_id_group_size</th></tr>
    <tr><th></th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>⋯</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;fct&gt;</th></tr>
</thead>
<tbody>
    <tr><th scope=row>sc5p_v2_hs_PBMC_10k_AAACCTGAGACAGACC</th><td>SeuratProject</td><td> 4751</td><td>1982</td><td>4.25173648</td><td>3 </td><td>3 </td><td>FALSE</td><td>SeuratProject</td><td> 4751</td><td>1982</td><td>⋯</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td></tr>
    <tr><th scope=row>sc5p_v2_hs_PBMC_10k_AAACCTGAGCGATAGC</th><td>SeuratProject</td><td> 5767</td><td>1761</td><td>1.92474423</td><td>2 </td><td>2 </td><td>FALSE</td><td>SeuratProject</td><td> 5767</td><td>1761</td><td>⋯</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td></tr>
    <tr><th scope=row>sc5p_v2_hs_PBMC_10k_AAACCTGAGCGGCTTC</th><td>SeuratProject</td><td> 4991</td><td>1957</td><td>2.50450811</td><td>3 </td><td>3 </td><td>FALSE</td><td>SeuratProject</td><td> 4991</td><td>1957</td><td>⋯</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td></tr>
    <tr><th scope=row>sc5p_v2_hs_PBMC_10k_AAACCTGAGGATCGCA</th><td>SeuratProject</td><td> 6012</td><td>2310</td><td>3.69261477</td><td>3 </td><td>3 </td><td>FALSE</td><td>SeuratProject</td><td> 6012</td><td>2310</td><td>⋯</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td></tr>
    <tr><th scope=row>sc5p_v2_hs_PBMC_10k_AAACCTGAGTCACGCC</th><td>SeuratProject</td><td> 7267</td><td>2315</td><td>2.17421219</td><td>5 </td><td>5 </td><td>FALSE</td><td>SeuratProject</td><td> 7267</td><td>2315</td><td>⋯</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td></tr>
    <tr><th scope=row>sc5p_v2_hs_PBMC_10k_AAACCTGCACGTCAGC</th><td>SeuratProject</td><td> 5671</td><td>1950</td><td>3.82648563</td><td>16</td><td>16</td><td>FALSE</td><td>SeuratProject</td><td> 5671</td><td>1950</td><td>⋯</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td></tr>
</tbody>
</table></div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[8]:
</pre></div>
</div>
<div class="input_area highlight-r notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># now just replace the current Seurat@meta.data</span>
<span class="n">merged</span><span class="o">@</span><span class="n">meta.data</span> <span class="o">=</span> <span class="n">merged_data</span>
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[9]:
</pre></div>
</div>
<div class="input_area highlight-r notranslate"><div class="highlight"><pre>
<span></span><span class="nf">options</span><span class="p">(</span><span class="n">repr.plot.width</span> <span class="o">=</span> <span class="m">20</span><span class="p">,</span> <span class="n">repr.plot.height</span> <span class="o">=</span> <span class="m">5</span><span class="p">,</span> <span class="n">repr.plot.res</span> <span class="o">=</span> <span class="m">200</span><span class="p">)</span>
<span class="nf">DimPlot</span><span class="p">(</span><span class="n">merged</span><span class="p">,</span> <span class="n">group.by</span> <span class="o">=</span> <span class="nf">c</span><span class="p">(</span><span class="s">&#39;has_bcr&#39;</span><span class="p">,</span> <span class="s">&#39;isotype&#39;</span><span class="p">,</span> <span class="s">&#39;status&#39;</span><span class="p">))</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_5_dandelion_running_from_R_85_0.png" class="no-scaled-link" src="../_images/notebooks_5_dandelion_running_from_R_85_0.png" style="width: 2000px; height: 500px;" />
</div>
</div>
<p>If you want to visualise the BCR network, you will have to subset to cells that contain BCR.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[10]:
</pre></div>
</div>
<div class="input_area highlight-r notranslate"><div class="highlight"><pre>
<span></span><span class="n">merged_bcr</span> <span class="o">=</span> <span class="nf">subset</span><span class="p">(</span><span class="n">merged</span><span class="p">,</span> <span class="n">subset</span> <span class="o">=</span> <span class="n">has_bcr</span> <span class="o">==</span> <span class="s">&#39;True&#39;</span><span class="p">)</span>
<span class="n">merged_bcr</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
An object of class Seurat
38224 features across 972 samples within 1 assay
Active assay: RNA (38224 features, 2000 variable features)
 2 dimensional reductions calculated: pca, umap
</pre></div></div>
</div>
<div class="section" id="Extract-the-coordinates-from-AnnData-to-a-R-friendly-version">
<h3>Extract the coordinates from <code class="docutils literal notranslate"><span class="pre">AnnData</span></code> to a R friendly version<a class="headerlink" href="#Extract-the-coordinates-from-AnnData-to-a-R-friendly-version" title="Permalink to this headline">¶</a></h3>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[11]:
</pre></div>
</div>
<div class="input_area highlight-r notranslate"><div class="highlight"><pre>
<span></span><span class="n">X_bcr</span> <span class="o">=</span> <span class="n">adata</span><span class="o">$</span><span class="n">obsm</span><span class="p">[</span><span class="s">&#39;X_bcr&#39;</span><span class="p">]</span>
<span class="n">X_bcr</span> <span class="o">&lt;-</span> <span class="nf">apply</span><span class="p">(</span><span class="n">X_bcr</span><span class="p">,</span> <span class="m">2</span><span class="p">,</span> <span class="nf">function</span><span class="p">(</span><span class="n">x</span><span class="p">)</span> <span class="nf">gsub</span><span class="p">(</span><span class="s">&#39;NaN&#39;</span><span class="p">,</span> <span class="kc">NA</span><span class="p">,</span> <span class="n">x</span><span class="p">))</span> <span class="c1"># convert python NAs to R NAs</span>
<span class="n">X_bcr</span> <span class="o">&lt;-</span> <span class="nf">apply</span><span class="p">(</span><span class="n">X_bcr</span><span class="p">,</span> <span class="m">2</span><span class="p">,</span> <span class="nf">function</span><span class="p">(</span><span class="n">x</span><span class="p">)</span> <span class="nf">as.numeric</span><span class="p">(</span><span class="n">x</span><span class="p">))</span> <span class="c1"># Make sure they are actually numbers</span>
<span class="n">X_bcr</span> <span class="o">&lt;-</span> <span class="nf">as.matrix</span><span class="p">(</span><span class="n">X_bcr</span><span class="p">)</span>
<span class="nf">row.names</span><span class="p">(</span><span class="n">X_bcr</span><span class="p">)</span> <span class="o">&lt;-</span> <span class="nf">row.names</span><span class="p">(</span><span class="n">adata</span><span class="o">$</span><span class="n">obs</span><span class="p">)</span> <span class="c1"># will not work if the anndata .obs slot is malformed. the row.names for adata$obs and row.names(merged_bcr@meta.data) should be identical anyway, so just replace with row.names(merged_bcr@meta.data)</span>
<span class="n">X_bcr</span> <span class="o">&lt;-</span> <span class="n">X_bcr</span><span class="p">[</span><span class="o">!</span><span class="nf">is.na</span><span class="p">(</span><span class="n">X_bcr</span><span class="p">[,</span><span class="m">1</span><span class="p">]),</span> <span class="p">]</span>
<span class="nf">colnames</span><span class="p">(</span><span class="n">X_bcr</span><span class="p">)</span> <span class="o">&lt;-</span> <span class="nf">c</span><span class="p">(</span><span class="s">&#39;BCR_1&#39;</span><span class="p">,</span> <span class="s">&#39;BCR_2&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[12]:
</pre></div>
</div>
<div class="input_area highlight-r notranslate"><div class="highlight"><pre>
<span></span><span class="n">merged_bcr</span><span class="p">[[</span><span class="s">&quot;bcr&quot;</span><span class="p">]]</span> <span class="o">&lt;-</span> <span class="nf">CreateDimReducObject</span><span class="p">(</span><span class="n">embeddings</span> <span class="o">=</span> <span class="n">X_bcr</span><span class="p">,</span> <span class="n">key</span> <span class="o">=</span> <span class="s">&quot;BCR_&quot;</span><span class="p">,</span> <span class="n">assay</span> <span class="o">=</span> <span class="nf">DefaultAssay</span><span class="p">(</span><span class="n">merged_bcr</span><span class="p">))</span>
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[13]:
</pre></div>
</div>
<div class="input_area highlight-r notranslate"><div class="highlight"><pre>
<span></span><span class="nf">options</span><span class="p">(</span><span class="n">repr.plot.width</span> <span class="o">=</span> <span class="m">6</span><span class="p">,</span> <span class="n">repr.plot.height</span> <span class="o">=</span> <span class="m">5</span><span class="p">,</span> <span class="n">repr.plot.res</span> <span class="o">=</span> <span class="m">200</span><span class="p">)</span>
<span class="nf">DimPlot</span><span class="p">(</span><span class="n">merged_bcr</span><span class="p">,</span> <span class="n">reduction</span> <span class="o">=</span> <span class="s">&#39;bcr&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_5_dandelion_running_from_R_91_0.png" class="no-scaled-link" src="../_images/notebooks_5_dandelion_running_from_R_91_0.png" style="width: 600px; height: 500px;" />
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[14]:
</pre></div>
</div>
<div class="input_area highlight-r notranslate"><div class="highlight"><pre>
<span></span><span class="nf">DimPlot</span><span class="p">(</span><span class="n">merged_bcr</span><span class="p">,</span> <span class="n">reduction</span> <span class="o">=</span> <span class="s">&#39;bcr&#39;</span><span class="p">,</span> <span class="n">group.by</span> <span class="o">=</span> <span class="s">&#39;isotype&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_5_dandelion_running_from_R_92_0.png" class="no-scaled-link" src="../_images/notebooks_5_dandelion_running_from_R_92_0.png" style="width: 600px; height: 500px;" />
</div>
</div>
<p>This concludes a quick primer on how to use <code class="docutils literal notranslate"><span class="pre">dandelion</span></code> from R. It may be a bit buggy from time to time due to how reticulate works but hopefully it will be overall alright.</p>
<p>The rest of the functions could be potentially run from R (just be changing <code class="docutils literal notranslate"><span class="pre">.</span></code> to <code class="docutils literal notranslate"><span class="pre">$</span></code> for example), but I haven’t tested it. Might be easier to run it through python, or maybe with the new RStudio 4?</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-r notranslate"><div class="highlight"><pre>
<span></span><span class="nf">sessionInfo</span><span class="p">()</span>
</pre></div>
</div>
</div>
</div>
</div>
</div>


           </div>
           
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
      
        <a href="3_dandelion_findingclones_and_analysis-10x%20data.html" class="btn btn-neutral float-left" title="dandelion Notebook-3" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        
        &copy; Copyright 2020, zktuong

    </p>
  </div>
    
    
    
    Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a
    
    <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a>
    
    provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  

  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>